{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv\n",
    "import pandas as pd\n",
    "from civicpy import civic"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Normalized Queries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['2489', 'NC_000003.11:g.10191648_10191649insC', 'genomic', 'True']\n"
     ]
    }
   ],
   "source": [
    "with open(\"./able_to_normalize_queries.csv\", \"r\") as f:\n",
    "    csv_reader = csv.reader(f, delimiter=\"\\t\")\n",
    "\n",
    "    next(csv_reader)\n",
    "\n",
    "    for row in csv_reader:\n",
    "        print(row)\n",
    "        break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>query</th>\n",
       "      <th>query_type</th>\n",
       "      <th>variant_accepted</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2489</td>\n",
       "      <td>NC_000003.11:g.10191648_10191649insC</td>\n",
       "      <td>genomic</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1988</td>\n",
       "      <td>NC_000003.11:g.10191649A&gt;T</td>\n",
       "      <td>genomic</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2488</td>\n",
       "      <td>3-10191647-T-G</td>\n",
       "      <td>genomic</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1986</td>\n",
       "      <td>NC_000003.11:g.10191648G&gt;T</td>\n",
       "      <td>genomic</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1987</td>\n",
       "      <td>NC_000003.11:g.10191649A&gt;G</td>\n",
       "      <td>genomic</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1853</th>\n",
       "      <td>3161</td>\n",
       "      <td>3-10183878-G-A</td>\n",
       "      <td>genomic</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1854</th>\n",
       "      <td>877</td>\n",
       "      <td>NC_000020.11:g.58903752C&gt;T</td>\n",
       "      <td>genomic</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1855</th>\n",
       "      <td>731</td>\n",
       "      <td>NC_000003.11:g.37056036G&gt;A</td>\n",
       "      <td>genomic</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1856</th>\n",
       "      <td>3045</td>\n",
       "      <td>VHL p.F76del</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1857</th>\n",
       "      <td>3310</td>\n",
       "      <td>HDAC9 p.L33R</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1858 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      variant_id                                 query query_type   \n",
       "0           2489  NC_000003.11:g.10191648_10191649insC    genomic  \\\n",
       "1           1988            NC_000003.11:g.10191649A>T    genomic   \n",
       "2           2488                        3-10191647-T-G    genomic   \n",
       "3           1986            NC_000003.11:g.10191648G>T    genomic   \n",
       "4           1987            NC_000003.11:g.10191649A>G    genomic   \n",
       "...          ...                                   ...        ...   \n",
       "1853        3161                        3-10183878-G-A    genomic   \n",
       "1854         877            NC_000020.11:g.58903752C>T    genomic   \n",
       "1855         731            NC_000003.11:g.37056036G>A    genomic   \n",
       "1856        3045                          VHL p.F76del    protein   \n",
       "1857        3310                          HDAC9 p.L33R    protein   \n",
       "\n",
       "      variant_accepted  \n",
       "0                 True  \n",
       "1                 True  \n",
       "2                 True  \n",
       "3                 True  \n",
       "4                 True  \n",
       "...                ...  \n",
       "1853             False  \n",
       "1854              True  \n",
       "1855              True  \n",
       "1856             False  \n",
       "1857             False  \n",
       "\n",
       "[1858 rows x 4 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "norm_quer_df = pd.read_csv(\"./able_to_normalize_queries.csv\", sep= \"\\t\")\n",
    "norm_quer_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "civic_variant_ids = civic.get_all_variants()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_variant_id_list = list(norm_quer_df[\"variant_id\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0       2489\n",
       "1       1988\n",
       "2       2488\n",
       "3       1986\n",
       "4       1987\n",
       "        ... \n",
       "1853    3161\n",
       "1854     877\n",
       "1855     731\n",
       "1856    3045\n",
       "1857    3310\n",
       "Name: variant_id, Length: 1858, dtype: int64"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "norm_quer_df.variant_id.astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>query</th>\n",
       "      <th>query_type</th>\n",
       "      <th>variant_accepted</th>\n",
       "      <th>evidence_ids</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2489</td>\n",
       "      <td>NC_000003.11:g.10191648_10191649insC</td>\n",
       "      <td>genomic</td>\n",
       "      <td>True</td>\n",
       "      <td>[9347, 6724]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1988</td>\n",
       "      <td>NC_000003.11:g.10191649A&gt;T</td>\n",
       "      <td>genomic</td>\n",
       "      <td>True</td>\n",
       "      <td>[5336]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2488</td>\n",
       "      <td>3-10191647-T-G</td>\n",
       "      <td>genomic</td>\n",
       "      <td>True</td>\n",
       "      <td>[10779, 6723, 8258]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1986</td>\n",
       "      <td>NC_000003.11:g.10191648G&gt;T</td>\n",
       "      <td>genomic</td>\n",
       "      <td>True</td>\n",
       "      <td>[5334]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1987</td>\n",
       "      <td>NC_000003.11:g.10191649A&gt;G</td>\n",
       "      <td>genomic</td>\n",
       "      <td>True</td>\n",
       "      <td>[5335]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1853</th>\n",
       "      <td>3161</td>\n",
       "      <td>3-10183878-G-A</td>\n",
       "      <td>genomic</td>\n",
       "      <td>False</td>\n",
       "      <td>[8659]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1854</th>\n",
       "      <td>877</td>\n",
       "      <td>NC_000020.11:g.58903752C&gt;T</td>\n",
       "      <td>genomic</td>\n",
       "      <td>True</td>\n",
       "      <td>[1997, 2892, 2893, 2895, 2896, 2897, 2898, 199...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1855</th>\n",
       "      <td>731</td>\n",
       "      <td>NC_000003.11:g.37056036G&gt;A</td>\n",
       "      <td>genomic</td>\n",
       "      <td>True</td>\n",
       "      <td>[1794]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1856</th>\n",
       "      <td>3045</td>\n",
       "      <td>VHL p.F76del</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>[8240]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1857</th>\n",
       "      <td>3310</td>\n",
       "      <td>HDAC9 p.L33R</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>[9240]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1858 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      variant_id                                 query query_type   \n",
       "0           2489  NC_000003.11:g.10191648_10191649insC    genomic  \\\n",
       "1           1988            NC_000003.11:g.10191649A>T    genomic   \n",
       "2           2488                        3-10191647-T-G    genomic   \n",
       "3           1986            NC_000003.11:g.10191648G>T    genomic   \n",
       "4           1987            NC_000003.11:g.10191649A>G    genomic   \n",
       "...          ...                                   ...        ...   \n",
       "1853        3161                        3-10183878-G-A    genomic   \n",
       "1854         877            NC_000020.11:g.58903752C>T    genomic   \n",
       "1855         731            NC_000003.11:g.37056036G>A    genomic   \n",
       "1856        3045                          VHL p.F76del    protein   \n",
       "1857        3310                          HDAC9 p.L33R    protein   \n",
       "\n",
       "      variant_accepted                                       evidence_ids  \n",
       "0                 True                                       [9347, 6724]  \n",
       "1                 True                                             [5336]  \n",
       "2                 True                                [10779, 6723, 8258]  \n",
       "3                 True                                             [5334]  \n",
       "4                 True                                             [5335]  \n",
       "...                ...                                                ...  \n",
       "1853             False                                             [8659]  \n",
       "1854              True  [1997, 2892, 2893, 2895, 2896, 2897, 2898, 199...  \n",
       "1855              True                                             [1794]  \n",
       "1856             False                                             [8240]  \n",
       "1857             False                                             [9240]  \n",
       "\n",
       "[1858 rows x 5 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "n_e_ids = []\n",
    "for v in n_variant_id_list:\n",
    "    n_civic_e_ids = []\n",
    "    for variant in civic_variant_ids: \n",
    "        # print(civic_id, type(civic_id))\n",
    "        # print(   = =variant.gene_id, type(variant.gene_id))\n",
    "        if int(v) == variant.id:\n",
    "            for mp in variant.molecular_profiles:\n",
    "                for e in mp.evidence_items:\n",
    "                    if e.id not in n_civic_e_ids:\n",
    "                        n_civic_e_ids.append(e.id)\n",
    "            n_civic_e_ids = n_civic_e_ids or \"\"\n",
    "    n_e_ids.append(n_civic_e_ids)\n",
    "norm_quer_df[\"evidence_ids\"] = n_e_ids\n",
    "norm_quer_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>query</th>\n",
       "      <th>query_type</th>\n",
       "      <th>variant_accepted</th>\n",
       "      <th>evidence_ids</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2489</td>\n",
       "      <td>NC_000003.11:g.10191648_10191649insC</td>\n",
       "      <td>genomic</td>\n",
       "      <td>True</td>\n",
       "      <td>9347</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2489</td>\n",
       "      <td>NC_000003.11:g.10191648_10191649insC</td>\n",
       "      <td>genomic</td>\n",
       "      <td>True</td>\n",
       "      <td>6724</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1988</td>\n",
       "      <td>NC_000003.11:g.10191649A&gt;T</td>\n",
       "      <td>genomic</td>\n",
       "      <td>True</td>\n",
       "      <td>5336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2488</td>\n",
       "      <td>3-10191647-T-G</td>\n",
       "      <td>genomic</td>\n",
       "      <td>True</td>\n",
       "      <td>10779</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2488</td>\n",
       "      <td>3-10191647-T-G</td>\n",
       "      <td>genomic</td>\n",
       "      <td>True</td>\n",
       "      <td>6723</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1854</th>\n",
       "      <td>877</td>\n",
       "      <td>NC_000020.11:g.58903752C&gt;T</td>\n",
       "      <td>genomic</td>\n",
       "      <td>True</td>\n",
       "      <td>1996</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1854</th>\n",
       "      <td>877</td>\n",
       "      <td>NC_000020.11:g.58903752C&gt;T</td>\n",
       "      <td>genomic</td>\n",
       "      <td>True</td>\n",
       "      <td>1995</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1855</th>\n",
       "      <td>731</td>\n",
       "      <td>NC_000003.11:g.37056036G&gt;A</td>\n",
       "      <td>genomic</td>\n",
       "      <td>True</td>\n",
       "      <td>1794</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1856</th>\n",
       "      <td>3045</td>\n",
       "      <td>VHL p.F76del</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>8240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1857</th>\n",
       "      <td>3310</td>\n",
       "      <td>HDAC9 p.L33R</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>9240</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5896 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      variant_id                                 query query_type   \n",
       "0           2489  NC_000003.11:g.10191648_10191649insC    genomic  \\\n",
       "0           2489  NC_000003.11:g.10191648_10191649insC    genomic   \n",
       "1           1988            NC_000003.11:g.10191649A>T    genomic   \n",
       "2           2488                        3-10191647-T-G    genomic   \n",
       "2           2488                        3-10191647-T-G    genomic   \n",
       "...          ...                                   ...        ...   \n",
       "1854         877            NC_000020.11:g.58903752C>T    genomic   \n",
       "1854         877            NC_000020.11:g.58903752C>T    genomic   \n",
       "1855         731            NC_000003.11:g.37056036G>A    genomic   \n",
       "1856        3045                          VHL p.F76del    protein   \n",
       "1857        3310                          HDAC9 p.L33R    protein   \n",
       "\n",
       "      variant_accepted evidence_ids  \n",
       "0                 True         9347  \n",
       "0                 True         6724  \n",
       "1                 True         5336  \n",
       "2                 True        10779  \n",
       "2                 True         6723  \n",
       "...                ...          ...  \n",
       "1854              True         1996  \n",
       "1854              True         1995  \n",
       "1855              True         1794  \n",
       "1856             False         8240  \n",
       "1857             False         9240  \n",
       "\n",
       "[5896 rows x 5 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "norm_quer_df = norm_quer_df.explode(column=\"evidence_ids\")\n",
    "norm_quer_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "civic_evidence_ids = civic.get_all_evidence()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_evidence_id_list = list(norm_quer_df[\"evidence_ids\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>query</th>\n",
       "      <th>query_type</th>\n",
       "      <th>variant_accepted</th>\n",
       "      <th>evidence_ids</th>\n",
       "      <th>evidence_status</th>\n",
       "      <th>evidence_rating</th>\n",
       "      <th>evidence_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2489</td>\n",
       "      <td>NC_000003.11:g.10191648_10191649insC</td>\n",
       "      <td>genomic</td>\n",
       "      <td>True</td>\n",
       "      <td>9347</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2489</td>\n",
       "      <td>NC_000003.11:g.10191648_10191649insC</td>\n",
       "      <td>genomic</td>\n",
       "      <td>True</td>\n",
       "      <td>6724</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1988</td>\n",
       "      <td>NC_000003.11:g.10191649A&gt;T</td>\n",
       "      <td>genomic</td>\n",
       "      <td>True</td>\n",
       "      <td>5336</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2488</td>\n",
       "      <td>3-10191647-T-G</td>\n",
       "      <td>genomic</td>\n",
       "      <td>True</td>\n",
       "      <td>10779</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2488</td>\n",
       "      <td>3-10191647-T-G</td>\n",
       "      <td>genomic</td>\n",
       "      <td>True</td>\n",
       "      <td>6723</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1854</th>\n",
       "      <td>877</td>\n",
       "      <td>NC_000020.11:g.58903752C&gt;T</td>\n",
       "      <td>genomic</td>\n",
       "      <td>True</td>\n",
       "      <td>1996</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1854</th>\n",
       "      <td>877</td>\n",
       "      <td>NC_000020.11:g.58903752C&gt;T</td>\n",
       "      <td>genomic</td>\n",
       "      <td>True</td>\n",
       "      <td>1995</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1855</th>\n",
       "      <td>731</td>\n",
       "      <td>NC_000003.11:g.37056036G&gt;A</td>\n",
       "      <td>genomic</td>\n",
       "      <td>True</td>\n",
       "      <td>1794</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1856</th>\n",
       "      <td>3045</td>\n",
       "      <td>VHL p.F76del</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>8240</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1857</th>\n",
       "      <td>3310</td>\n",
       "      <td>HDAC9 p.L33R</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>9240</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5896 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      variant_id                                 query query_type   \n",
       "0           2489  NC_000003.11:g.10191648_10191649insC    genomic  \\\n",
       "0           2489  NC_000003.11:g.10191648_10191649insC    genomic   \n",
       "1           1988            NC_000003.11:g.10191649A>T    genomic   \n",
       "2           2488                        3-10191647-T-G    genomic   \n",
       "2           2488                        3-10191647-T-G    genomic   \n",
       "...          ...                                   ...        ...   \n",
       "1854         877            NC_000020.11:g.58903752C>T    genomic   \n",
       "1854         877            NC_000020.11:g.58903752C>T    genomic   \n",
       "1855         731            NC_000003.11:g.37056036G>A    genomic   \n",
       "1856        3045                          VHL p.F76del    protein   \n",
       "1857        3310                          HDAC9 p.L33R    protein   \n",
       "\n",
       "      variant_accepted evidence_ids evidence_status evidence_rating   \n",
       "0                 True         9347     [submitted]             [3]  \\\n",
       "0                 True         6724      [accepted]             [2]   \n",
       "1                 True         5336      [accepted]             [2]   \n",
       "2                 True        10779     [submitted]             [3]   \n",
       "2                 True         6723      [accepted]             [2]   \n",
       "...                ...          ...             ...             ...   \n",
       "1854              True         1996      [accepted]             [3]   \n",
       "1854              True         1995      [accepted]             [2]   \n",
       "1855              True         1794      [accepted]             [1]   \n",
       "1856             False         8240     [submitted]             [2]   \n",
       "1857             False         9240     [submitted]             [4]   \n",
       "\n",
       "     evidence_level  \n",
       "0               [C]  \n",
       "0               [C]  \n",
       "1               [C]  \n",
       "2               [C]  \n",
       "2               [C]  \n",
       "...             ...  \n",
       "1854            [B]  \n",
       "1854            [B]  \n",
       "1855            [C]  \n",
       "1856            [C]  \n",
       "1857            [D]  \n",
       "\n",
       "[5896 rows x 8 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "n_e_status = []\n",
    "n_e_rating = []\n",
    "n_e_level = []\n",
    "for e in n_evidence_id_list:\n",
    "    n_civic_e_status = []\n",
    "    n_civic_e_rating = []\n",
    "    n_civic_e_level = []\n",
    "    for evidence in civic_evidence_ids: \n",
    "        # print(civic_id, type(civic_id))\n",
    "        # print(   = =variant.gene_id, type(variant.gene_id))\n",
    "        if int(e) == evidence.id:\n",
    "            if evidence.status not in n_civic_e_status:\n",
    "                n_civic_e_status.append(evidence.status)\n",
    "            n_civic_e_status = n_civic_e_status or \"\"\n",
    "            if evidence.rating not in n_civic_e_rating:\n",
    "                n_civic_e_rating.append(evidence.rating)\n",
    "            n_civic_e_rating = n_civic_e_rating or \"\"\n",
    "            if evidence.evidence_level not in n_civic_e_level:\n",
    "                n_civic_e_level.append(evidence.evidence_level)\n",
    "            n_civic_e_level = n_civic_e_level or \"\"\n",
    "    n_e_status.append(n_civic_e_status)\n",
    "    n_e_rating.append(n_civic_e_rating)\n",
    "    n_e_level.append(n_civic_e_level)\n",
    "norm_quer_df[\"evidence_status\"] = n_e_status\n",
    "norm_quer_df[\"evidence_rating\"] = n_e_rating\n",
    "norm_quer_df[\"evidence_level\"] = n_e_level\n",
    "norm_quer_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Unable to Normalize Queries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>query</th>\n",
       "      <th>query_type</th>\n",
       "      <th>variant_accepted</th>\n",
       "      <th>exception_raised</th>\n",
       "      <th>message</th>\n",
       "      <th>warnings</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>748</td>\n",
       "      <td>MLH1 *757L</td>\n",
       "      <td>protein</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3718</td>\n",
       "      <td>AR A748V</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3725</td>\n",
       "      <td>AR A765T</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>248</td>\n",
       "      <td>TERT C228T</td>\n",
       "      <td>protein</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4004</td>\n",
       "      <td>TERT C250T</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>3454</td>\n",
       "      <td>BRAF V599R</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>2825</td>\n",
       "      <td>BRAF V601E</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>3721</td>\n",
       "      <td>AR V757A</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>3722</td>\n",
       "      <td>AR V757I</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>3724</td>\n",
       "      <td>AR Y763C</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>63 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    variant_id       query query_type  variant_accepted  exception_raised   \n",
       "0          748  MLH1 *757L    protein              True             False  \\\n",
       "1         3718    AR A748V    protein             False             False   \n",
       "2         3725    AR A765T    protein             False             False   \n",
       "3          248  TERT C228T    protein              True             False   \n",
       "4         4004  TERT C250T    protein             False             False   \n",
       "..         ...         ...        ...               ...               ...   \n",
       "58        3454  BRAF V599R    protein             False             False   \n",
       "59        2825  BRAF V601E    protein             False             False   \n",
       "60        3721    AR V757A    protein             False             False   \n",
       "61        3722    AR V757I    protein             False             False   \n",
       "62        3724    AR Y763C    protein             False             False   \n",
       "\n",
       "                message                                           warnings  \n",
       "0   unable to normalize  [\"Unable to find valid result for classificati...  \n",
       "1   unable to normalize  [\"Unable to find valid result for classificati...  \n",
       "2   unable to normalize  [\"Unable to find valid result for classificati...  \n",
       "3   unable to normalize  [\"Unable to find valid result for classificati...  \n",
       "4   unable to normalize  [\"Unable to find valid result for classificati...  \n",
       "..                  ...                                                ...  \n",
       "58  unable to normalize  [\"Unable to find valid result for classificati...  \n",
       "59  unable to normalize  [\"Unable to find valid result for classificati...  \n",
       "60  unable to normalize  [\"Unable to find valid result for classificati...  \n",
       "61  unable to normalize  [\"Unable to find valid result for classificati...  \n",
       "62  unable to normalize  [\"Unable to find valid result for classificati...  \n",
       "\n",
       "[63 rows x 7 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "non_norm_quer_df = pd.read_csv(\"./unable_to_normalize_queries.csv\", sep= \"\\t\")\n",
    "non_norm_quer_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "nn_variant_id_list = list(non_norm_quer_df[\"variant_id\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>query</th>\n",
       "      <th>query_type</th>\n",
       "      <th>variant_accepted</th>\n",
       "      <th>exception_raised</th>\n",
       "      <th>message</th>\n",
       "      <th>warnings</th>\n",
       "      <th>evidence_ids</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>748</td>\n",
       "      <td>MLH1 *757L</td>\n",
       "      <td>protein</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "      <td>[1812]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3718</td>\n",
       "      <td>AR A748V</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "      <td>[10128]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3725</td>\n",
       "      <td>AR A765T</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "      <td>[10135]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>248</td>\n",
       "      <td>TERT C228T</td>\n",
       "      <td>protein</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "      <td>[655, 1646, 6934, 6935]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4004</td>\n",
       "      <td>TERT C250T</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "      <td>[10331]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>3454</td>\n",
       "      <td>BRAF V599R</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "      <td>[9571]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>2825</td>\n",
       "      <td>BRAF V601E</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "      <td>[7633, 7627]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>3721</td>\n",
       "      <td>AR V757A</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "      <td>[10131]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>3722</td>\n",
       "      <td>AR V757I</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "      <td>[10132]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>3724</td>\n",
       "      <td>AR Y763C</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "      <td>[10134]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>63 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    variant_id       query query_type  variant_accepted  exception_raised   \n",
       "0          748  MLH1 *757L    protein              True             False  \\\n",
       "1         3718    AR A748V    protein             False             False   \n",
       "2         3725    AR A765T    protein             False             False   \n",
       "3          248  TERT C228T    protein              True             False   \n",
       "4         4004  TERT C250T    protein             False             False   \n",
       "..         ...         ...        ...               ...               ...   \n",
       "58        3454  BRAF V599R    protein             False             False   \n",
       "59        2825  BRAF V601E    protein             False             False   \n",
       "60        3721    AR V757A    protein             False             False   \n",
       "61        3722    AR V757I    protein             False             False   \n",
       "62        3724    AR Y763C    protein             False             False   \n",
       "\n",
       "                message                                           warnings   \n",
       "0   unable to normalize  [\"Unable to find valid result for classificati...  \\\n",
       "1   unable to normalize  [\"Unable to find valid result for classificati...   \n",
       "2   unable to normalize  [\"Unable to find valid result for classificati...   \n",
       "3   unable to normalize  [\"Unable to find valid result for classificati...   \n",
       "4   unable to normalize  [\"Unable to find valid result for classificati...   \n",
       "..                  ...                                                ...   \n",
       "58  unable to normalize  [\"Unable to find valid result for classificati...   \n",
       "59  unable to normalize  [\"Unable to find valid result for classificati...   \n",
       "60  unable to normalize  [\"Unable to find valid result for classificati...   \n",
       "61  unable to normalize  [\"Unable to find valid result for classificati...   \n",
       "62  unable to normalize  [\"Unable to find valid result for classificati...   \n",
       "\n",
       "               evidence_ids  \n",
       "0                    [1812]  \n",
       "1                   [10128]  \n",
       "2                   [10135]  \n",
       "3   [655, 1646, 6934, 6935]  \n",
       "4                   [10331]  \n",
       "..                      ...  \n",
       "58                   [9571]  \n",
       "59             [7633, 7627]  \n",
       "60                  [10131]  \n",
       "61                  [10132]  \n",
       "62                  [10134]  \n",
       "\n",
       "[63 rows x 8 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nn_e_ids = []\n",
    "for v in nn_variant_id_list:\n",
    "    nn_civic_e_ids = []\n",
    "    for variant in civic_variant_ids: \n",
    "        if int(v) == variant.id:\n",
    "            for mp in variant.molecular_profiles:\n",
    "                for e in mp.evidence_items:\n",
    "                    if e.id not in nn_civic_e_ids:\n",
    "                        nn_civic_e_ids.append(e.id)\n",
    "            nn_civic_e_ids = nn_civic_e_ids or \"\"\n",
    "    nn_e_ids.append(nn_civic_e_ids)\n",
    "non_norm_quer_df[\"evidence_ids\"] = nn_e_ids\n",
    "non_norm_quer_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>query</th>\n",
       "      <th>query_type</th>\n",
       "      <th>variant_accepted</th>\n",
       "      <th>exception_raised</th>\n",
       "      <th>message</th>\n",
       "      <th>warnings</th>\n",
       "      <th>evidence_ids</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>748</td>\n",
       "      <td>MLH1 *757L</td>\n",
       "      <td>protein</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "      <td>1812</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3718</td>\n",
       "      <td>AR A748V</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "      <td>10128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3725</td>\n",
       "      <td>AR A765T</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "      <td>10135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>248</td>\n",
       "      <td>TERT C228T</td>\n",
       "      <td>protein</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "      <td>655</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>248</td>\n",
       "      <td>TERT C228T</td>\n",
       "      <td>protein</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "      <td>1646</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>2825</td>\n",
       "      <td>BRAF V601E</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "      <td>7633</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>2825</td>\n",
       "      <td>BRAF V601E</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "      <td>7627</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>3721</td>\n",
       "      <td>AR V757A</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "      <td>10131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>3722</td>\n",
       "      <td>AR V757I</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "      <td>10132</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>3724</td>\n",
       "      <td>AR Y763C</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "      <td>10134</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>80 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    variant_id       query query_type  variant_accepted  exception_raised   \n",
       "0          748  MLH1 *757L    protein              True             False  \\\n",
       "1         3718    AR A748V    protein             False             False   \n",
       "2         3725    AR A765T    protein             False             False   \n",
       "3          248  TERT C228T    protein              True             False   \n",
       "3          248  TERT C228T    protein              True             False   \n",
       "..         ...         ...        ...               ...               ...   \n",
       "59        2825  BRAF V601E    protein             False             False   \n",
       "59        2825  BRAF V601E    protein             False             False   \n",
       "60        3721    AR V757A    protein             False             False   \n",
       "61        3722    AR V757I    protein             False             False   \n",
       "62        3724    AR Y763C    protein             False             False   \n",
       "\n",
       "                message                                           warnings   \n",
       "0   unable to normalize  [\"Unable to find valid result for classificati...  \\\n",
       "1   unable to normalize  [\"Unable to find valid result for classificati...   \n",
       "2   unable to normalize  [\"Unable to find valid result for classificati...   \n",
       "3   unable to normalize  [\"Unable to find valid result for classificati...   \n",
       "3   unable to normalize  [\"Unable to find valid result for classificati...   \n",
       "..                  ...                                                ...   \n",
       "59  unable to normalize  [\"Unable to find valid result for classificati...   \n",
       "59  unable to normalize  [\"Unable to find valid result for classificati...   \n",
       "60  unable to normalize  [\"Unable to find valid result for classificati...   \n",
       "61  unable to normalize  [\"Unable to find valid result for classificati...   \n",
       "62  unable to normalize  [\"Unable to find valid result for classificati...   \n",
       "\n",
       "   evidence_ids  \n",
       "0          1812  \n",
       "1         10128  \n",
       "2         10135  \n",
       "3           655  \n",
       "3          1646  \n",
       "..          ...  \n",
       "59         7633  \n",
       "59         7627  \n",
       "60        10131  \n",
       "61        10132  \n",
       "62        10134  \n",
       "\n",
       "[80 rows x 8 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "non_norm_quer_df = non_norm_quer_df.explode(column=\"evidence_ids\")\n",
    "non_norm_quer_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "nn_evidence_id_list = list(non_norm_quer_df[\"evidence_ids\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>query</th>\n",
       "      <th>query_type</th>\n",
       "      <th>variant_accepted</th>\n",
       "      <th>exception_raised</th>\n",
       "      <th>message</th>\n",
       "      <th>warnings</th>\n",
       "      <th>evidence_ids</th>\n",
       "      <th>evidence_status</th>\n",
       "      <th>evidence_rating</th>\n",
       "      <th>evidence_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>748</td>\n",
       "      <td>MLH1 *757L</td>\n",
       "      <td>protein</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "      <td>1812</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3718</td>\n",
       "      <td>AR A748V</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "      <td>10128</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3725</td>\n",
       "      <td>AR A765T</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "      <td>10135</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>248</td>\n",
       "      <td>TERT C228T</td>\n",
       "      <td>protein</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "      <td>655</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>248</td>\n",
       "      <td>TERT C228T</td>\n",
       "      <td>protein</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "      <td>1646</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>2825</td>\n",
       "      <td>BRAF V601E</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "      <td>7633</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>2825</td>\n",
       "      <td>BRAF V601E</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "      <td>7627</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>3721</td>\n",
       "      <td>AR V757A</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "      <td>10131</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>3722</td>\n",
       "      <td>AR V757I</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "      <td>10132</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>3724</td>\n",
       "      <td>AR Y763C</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "      <td>10134</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>80 rows Ã— 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    variant_id       query query_type  variant_accepted  exception_raised   \n",
       "0          748  MLH1 *757L    protein              True             False  \\\n",
       "1         3718    AR A748V    protein             False             False   \n",
       "2         3725    AR A765T    protein             False             False   \n",
       "3          248  TERT C228T    protein              True             False   \n",
       "3          248  TERT C228T    protein              True             False   \n",
       "..         ...         ...        ...               ...               ...   \n",
       "59        2825  BRAF V601E    protein             False             False   \n",
       "59        2825  BRAF V601E    protein             False             False   \n",
       "60        3721    AR V757A    protein             False             False   \n",
       "61        3722    AR V757I    protein             False             False   \n",
       "62        3724    AR Y763C    protein             False             False   \n",
       "\n",
       "                message                                           warnings   \n",
       "0   unable to normalize  [\"Unable to find valid result for classificati...  \\\n",
       "1   unable to normalize  [\"Unable to find valid result for classificati...   \n",
       "2   unable to normalize  [\"Unable to find valid result for classificati...   \n",
       "3   unable to normalize  [\"Unable to find valid result for classificati...   \n",
       "3   unable to normalize  [\"Unable to find valid result for classificati...   \n",
       "..                  ...                                                ...   \n",
       "59  unable to normalize  [\"Unable to find valid result for classificati...   \n",
       "59  unable to normalize  [\"Unable to find valid result for classificati...   \n",
       "60  unable to normalize  [\"Unable to find valid result for classificati...   \n",
       "61  unable to normalize  [\"Unable to find valid result for classificati...   \n",
       "62  unable to normalize  [\"Unable to find valid result for classificati...   \n",
       "\n",
       "   evidence_ids evidence_status evidence_rating evidence_level  \n",
       "0          1812      [accepted]             [1]            [C]  \n",
       "1         10128     [submitted]             [3]            [D]  \n",
       "2         10135     [submitted]             [3]            [D]  \n",
       "3           655      [accepted]             [5]            [B]  \n",
       "3          1646      [accepted]             [3]            [B]  \n",
       "..          ...             ...             ...            ...  \n",
       "59         7633     [submitted]             [3]            [B]  \n",
       "59         7627     [submitted]             [3]            [C]  \n",
       "60        10131     [submitted]             [3]            [D]  \n",
       "61        10132     [submitted]             [3]            [D]  \n",
       "62        10134     [submitted]             [3]            [D]  \n",
       "\n",
       "[80 rows x 11 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nn_e_status = []\n",
    "nn_e_rating = []\n",
    "nn_e_level = []\n",
    "for e in nn_evidence_id_list:\n",
    "    nn_civic_e_status = []\n",
    "    nn_civic_e_rating = []\n",
    "    nn_civic_e_level = []\n",
    "    for evidence in civic_evidence_ids: \n",
    "        # print(civic_id, type(civic_id))\n",
    "        # print(   = =variant.gene_id, type(variant.gene_id))\n",
    "        if int(e) == evidence.id:\n",
    "            if evidence.status not in nn_civic_e_status:\n",
    "                nn_civic_e_status.append(evidence.status)\n",
    "            nn_civic_e_status = nn_civic_e_status or \"\"\n",
    "            if evidence.rating not in nn_civic_e_rating:\n",
    "                nn_civic_e_rating.append(evidence.rating)\n",
    "            nn_civic_e_rating = nn_civic_e_rating or \"\"\n",
    "            if evidence.evidence_level not in nn_civic_e_level:\n",
    "                nn_civic_e_level.append(evidence.evidence_level)\n",
    "            nn_civic_e_level = nn_civic_e_level or \"\"\n",
    "    nn_e_status.append(nn_civic_e_status)\n",
    "    nn_e_rating.append(nn_civic_e_rating)\n",
    "    nn_e_level.append(nn_civic_e_level)\n",
    "non_norm_quer_df[\"evidence_status\"] = nn_e_status\n",
    "non_norm_quer_df[\"evidence_rating\"] = nn_e_rating\n",
    "non_norm_quer_df[\"evidence_level\"] = nn_e_level\n",
    "non_norm_quer_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Not Supported Variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>variant_name</th>\n",
       "      <th>category</th>\n",
       "      <th>variant_accepted</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4417</td>\n",
       "      <td>ALK</td>\n",
       "      <td>FBXO11::ALK</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4188</td>\n",
       "      <td>VHL</td>\n",
       "      <td>10kb Deletion</td>\n",
       "      <td>Region Defined Variant</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>785</td>\n",
       "      <td>CHEK2</td>\n",
       "      <td>1100DELC</td>\n",
       "      <td>Protein Consequence</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2438</td>\n",
       "      <td>VHL</td>\n",
       "      <td>235 (CAG-TAG)</td>\n",
       "      <td>Other</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4186</td>\n",
       "      <td>VHL</td>\n",
       "      <td>2kb Deletion</td>\n",
       "      <td>Region Defined Variant</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1199</th>\n",
       "      <td>4008</td>\n",
       "      <td>VHL</td>\n",
       "      <td>t(3;6)(p13;q25.1)</td>\n",
       "      <td>Rearrangements</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1200</th>\n",
       "      <td>3351</td>\n",
       "      <td>VHL</td>\n",
       "      <td>t(3;8)(p13;q24.1)</td>\n",
       "      <td>Rearrangements</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1201</th>\n",
       "      <td>3478</td>\n",
       "      <td>ESR2</td>\n",
       "      <td>underexpression beta-1</td>\n",
       "      <td>Other</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1202</th>\n",
       "      <td>3508</td>\n",
       "      <td>CD274</td>\n",
       "      <td>v242</td>\n",
       "      <td>Protein Consequence</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1203</th>\n",
       "      <td>2422</td>\n",
       "      <td>NTRK3</td>\n",
       "      <td>~DEPRECATED-ETV6-NTRK3</td>\n",
       "      <td>Other</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1204 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      variant_id gene_name            variant_name                category   \n",
       "0           4417       ALK             FBXO11::ALK                  Fusion  \\\n",
       "1           4188       VHL           10kb Deletion  Region Defined Variant   \n",
       "2            785     CHEK2                1100DELC     Protein Consequence   \n",
       "3           2438       VHL           235 (CAG-TAG)                   Other   \n",
       "4           4186       VHL            2kb Deletion  Region Defined Variant   \n",
       "...          ...       ...                     ...                     ...   \n",
       "1199        4008       VHL       t(3;6)(p13;q25.1)          Rearrangements   \n",
       "1200        3351       VHL       t(3;8)(p13;q24.1)          Rearrangements   \n",
       "1201        3478      ESR2  underexpression beta-1                   Other   \n",
       "1202        3508     CD274                    v242     Protein Consequence   \n",
       "1203        2422     NTRK3  ~DEPRECATED-ETV6-NTRK3                   Other   \n",
       "\n",
       "      variant_accepted  \n",
       "0                False  \n",
       "1                False  \n",
       "2                 True  \n",
       "3                False  \n",
       "4                False  \n",
       "...                ...  \n",
       "1199             False  \n",
       "1200             False  \n",
       "1201             False  \n",
       "1202             False  \n",
       "1203             False  \n",
       "\n",
       "[1204 rows x 5 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "non_supp_var_df = pd.read_csv(\"./not_supported_variants.csv\", sep= \"\\t\")\n",
    "non_supp_var_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "ns_variant_id_list = list(non_supp_var_df[\"variant_id\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>variant_name</th>\n",
       "      <th>category</th>\n",
       "      <th>variant_accepted</th>\n",
       "      <th>evidence_ids</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4417</td>\n",
       "      <td>ALK</td>\n",
       "      <td>FBXO11::ALK</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>False</td>\n",
       "      <td>[7428]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4188</td>\n",
       "      <td>VHL</td>\n",
       "      <td>10kb Deletion</td>\n",
       "      <td>Region Defined Variant</td>\n",
       "      <td>False</td>\n",
       "      <td>[10678]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>785</td>\n",
       "      <td>CHEK2</td>\n",
       "      <td>1100DELC</td>\n",
       "      <td>Protein Consequence</td>\n",
       "      <td>True</td>\n",
       "      <td>[7235, 1849, 1850]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2438</td>\n",
       "      <td>VHL</td>\n",
       "      <td>235 (CAG-TAG)</td>\n",
       "      <td>Other</td>\n",
       "      <td>False</td>\n",
       "      <td>[6524]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4186</td>\n",
       "      <td>VHL</td>\n",
       "      <td>2kb Deletion</td>\n",
       "      <td>Region Defined Variant</td>\n",
       "      <td>False</td>\n",
       "      <td>[10791, 10676]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1199</th>\n",
       "      <td>4008</td>\n",
       "      <td>VHL</td>\n",
       "      <td>t(3;6)(p13;q25.1)</td>\n",
       "      <td>Rearrangements</td>\n",
       "      <td>False</td>\n",
       "      <td>[10357, 10350]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1200</th>\n",
       "      <td>3351</td>\n",
       "      <td>VHL</td>\n",
       "      <td>t(3;8)(p13;q24.1)</td>\n",
       "      <td>Rearrangements</td>\n",
       "      <td>False</td>\n",
       "      <td>[9340]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1201</th>\n",
       "      <td>3478</td>\n",
       "      <td>ESR2</td>\n",
       "      <td>underexpression beta-1</td>\n",
       "      <td>Other</td>\n",
       "      <td>False</td>\n",
       "      <td>[9612, 9613, 9618, 9619]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1202</th>\n",
       "      <td>3508</td>\n",
       "      <td>CD274</td>\n",
       "      <td>v242</td>\n",
       "      <td>Protein Consequence</td>\n",
       "      <td>False</td>\n",
       "      <td>[9695]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1203</th>\n",
       "      <td>2422</td>\n",
       "      <td>NTRK3</td>\n",
       "      <td>~DEPRECATED-ETV6-NTRK3</td>\n",
       "      <td>Other</td>\n",
       "      <td>False</td>\n",
       "      <td>[10692, 10717]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1204 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      variant_id gene_name            variant_name                category   \n",
       "0           4417       ALK             FBXO11::ALK                  Fusion  \\\n",
       "1           4188       VHL           10kb Deletion  Region Defined Variant   \n",
       "2            785     CHEK2                1100DELC     Protein Consequence   \n",
       "3           2438       VHL           235 (CAG-TAG)                   Other   \n",
       "4           4186       VHL            2kb Deletion  Region Defined Variant   \n",
       "...          ...       ...                     ...                     ...   \n",
       "1199        4008       VHL       t(3;6)(p13;q25.1)          Rearrangements   \n",
       "1200        3351       VHL       t(3;8)(p13;q24.1)          Rearrangements   \n",
       "1201        3478      ESR2  underexpression beta-1                   Other   \n",
       "1202        3508     CD274                    v242     Protein Consequence   \n",
       "1203        2422     NTRK3  ~DEPRECATED-ETV6-NTRK3                   Other   \n",
       "\n",
       "      variant_accepted              evidence_ids  \n",
       "0                False                    [7428]  \n",
       "1                False                   [10678]  \n",
       "2                 True        [7235, 1849, 1850]  \n",
       "3                False                    [6524]  \n",
       "4                False            [10791, 10676]  \n",
       "...                ...                       ...  \n",
       "1199             False            [10357, 10350]  \n",
       "1200             False                    [9340]  \n",
       "1201             False  [9612, 9613, 9618, 9619]  \n",
       "1202             False                    [9695]  \n",
       "1203             False            [10692, 10717]  \n",
       "\n",
       "[1204 rows x 6 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ns_e_ids = []\n",
    "for v in ns_variant_id_list:\n",
    "    ns_civic_e_ids = []\n",
    "    for variant in civic_variant_ids: \n",
    "        # print(civic_id, type(civic_id))\n",
    "        # print(   = =variant.gene_id, type(variant.gene_id))\n",
    "        if int(v) == variant.id:\n",
    "            for mp in variant.molecular_profiles:\n",
    "                for e in mp.evidence_items:\n",
    "                    if e.id not in ns_civic_e_ids:\n",
    "                        ns_civic_e_ids.append(e.id)\n",
    "            ns_civic_e_ids = ns_civic_e_ids or \"\"\n",
    "    ns_e_ids.append(ns_civic_e_ids)\n",
    "non_supp_var_df[\"evidence_ids\"] = ns_e_ids\n",
    "non_supp_var_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>variant_name</th>\n",
       "      <th>category</th>\n",
       "      <th>variant_accepted</th>\n",
       "      <th>evidence_ids</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4417</td>\n",
       "      <td>ALK</td>\n",
       "      <td>FBXO11::ALK</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>False</td>\n",
       "      <td>7428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4188</td>\n",
       "      <td>VHL</td>\n",
       "      <td>10kb Deletion</td>\n",
       "      <td>Region Defined Variant</td>\n",
       "      <td>False</td>\n",
       "      <td>10678</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>785</td>\n",
       "      <td>CHEK2</td>\n",
       "      <td>1100DELC</td>\n",
       "      <td>Protein Consequence</td>\n",
       "      <td>True</td>\n",
       "      <td>7235</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>785</td>\n",
       "      <td>CHEK2</td>\n",
       "      <td>1100DELC</td>\n",
       "      <td>Protein Consequence</td>\n",
       "      <td>True</td>\n",
       "      <td>1849</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>785</td>\n",
       "      <td>CHEK2</td>\n",
       "      <td>1100DELC</td>\n",
       "      <td>Protein Consequence</td>\n",
       "      <td>True</td>\n",
       "      <td>1850</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1201</th>\n",
       "      <td>3478</td>\n",
       "      <td>ESR2</td>\n",
       "      <td>underexpression beta-1</td>\n",
       "      <td>Other</td>\n",
       "      <td>False</td>\n",
       "      <td>9618</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1201</th>\n",
       "      <td>3478</td>\n",
       "      <td>ESR2</td>\n",
       "      <td>underexpression beta-1</td>\n",
       "      <td>Other</td>\n",
       "      <td>False</td>\n",
       "      <td>9619</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1202</th>\n",
       "      <td>3508</td>\n",
       "      <td>CD274</td>\n",
       "      <td>v242</td>\n",
       "      <td>Protein Consequence</td>\n",
       "      <td>False</td>\n",
       "      <td>9695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1203</th>\n",
       "      <td>2422</td>\n",
       "      <td>NTRK3</td>\n",
       "      <td>~DEPRECATED-ETV6-NTRK3</td>\n",
       "      <td>Other</td>\n",
       "      <td>False</td>\n",
       "      <td>10692</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1203</th>\n",
       "      <td>2422</td>\n",
       "      <td>NTRK3</td>\n",
       "      <td>~DEPRECATED-ETV6-NTRK3</td>\n",
       "      <td>Other</td>\n",
       "      <td>False</td>\n",
       "      <td>10717</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3764 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      variant_id gene_name            variant_name                category   \n",
       "0           4417       ALK             FBXO11::ALK                  Fusion  \\\n",
       "1           4188       VHL           10kb Deletion  Region Defined Variant   \n",
       "2            785     CHEK2                1100DELC     Protein Consequence   \n",
       "2            785     CHEK2                1100DELC     Protein Consequence   \n",
       "2            785     CHEK2                1100DELC     Protein Consequence   \n",
       "...          ...       ...                     ...                     ...   \n",
       "1201        3478      ESR2  underexpression beta-1                   Other   \n",
       "1201        3478      ESR2  underexpression beta-1                   Other   \n",
       "1202        3508     CD274                    v242     Protein Consequence   \n",
       "1203        2422     NTRK3  ~DEPRECATED-ETV6-NTRK3                   Other   \n",
       "1203        2422     NTRK3  ~DEPRECATED-ETV6-NTRK3                   Other   \n",
       "\n",
       "      variant_accepted evidence_ids  \n",
       "0                False         7428  \n",
       "1                False        10678  \n",
       "2                 True         7235  \n",
       "2                 True         1849  \n",
       "2                 True         1850  \n",
       "...                ...          ...  \n",
       "1201             False         9618  \n",
       "1201             False         9619  \n",
       "1202             False         9695  \n",
       "1203             False        10692  \n",
       "1203             False        10717  \n",
       "\n",
       "[3764 rows x 6 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "non_supp_var_df = non_supp_var_df.explode(column=\"evidence_ids\")\n",
    "non_supp_var_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "ns_evidence_id_list = list(non_supp_var_df[\"evidence_ids\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>variant_name</th>\n",
       "      <th>category</th>\n",
       "      <th>variant_accepted</th>\n",
       "      <th>evidence_ids</th>\n",
       "      <th>evidence_status</th>\n",
       "      <th>evidence_rating</th>\n",
       "      <th>evidence_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4417</td>\n",
       "      <td>ALK</td>\n",
       "      <td>FBXO11::ALK</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>False</td>\n",
       "      <td>7428</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4188</td>\n",
       "      <td>VHL</td>\n",
       "      <td>10kb Deletion</td>\n",
       "      <td>Region Defined Variant</td>\n",
       "      <td>False</td>\n",
       "      <td>10678</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>785</td>\n",
       "      <td>CHEK2</td>\n",
       "      <td>1100DELC</td>\n",
       "      <td>Protein Consequence</td>\n",
       "      <td>True</td>\n",
       "      <td>7235</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>785</td>\n",
       "      <td>CHEK2</td>\n",
       "      <td>1100DELC</td>\n",
       "      <td>Protein Consequence</td>\n",
       "      <td>True</td>\n",
       "      <td>1849</td>\n",
       "      <td>[rejected]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>785</td>\n",
       "      <td>CHEK2</td>\n",
       "      <td>1100DELC</td>\n",
       "      <td>Protein Consequence</td>\n",
       "      <td>True</td>\n",
       "      <td>1850</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1201</th>\n",
       "      <td>3478</td>\n",
       "      <td>ESR2</td>\n",
       "      <td>underexpression beta-1</td>\n",
       "      <td>Other</td>\n",
       "      <td>False</td>\n",
       "      <td>9618</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1201</th>\n",
       "      <td>3478</td>\n",
       "      <td>ESR2</td>\n",
       "      <td>underexpression beta-1</td>\n",
       "      <td>Other</td>\n",
       "      <td>False</td>\n",
       "      <td>9619</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1202</th>\n",
       "      <td>3508</td>\n",
       "      <td>CD274</td>\n",
       "      <td>v242</td>\n",
       "      <td>Protein Consequence</td>\n",
       "      <td>False</td>\n",
       "      <td>9695</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[E]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1203</th>\n",
       "      <td>2422</td>\n",
       "      <td>NTRK3</td>\n",
       "      <td>~DEPRECATED-ETV6-NTRK3</td>\n",
       "      <td>Other</td>\n",
       "      <td>False</td>\n",
       "      <td>10692</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1203</th>\n",
       "      <td>2422</td>\n",
       "      <td>NTRK3</td>\n",
       "      <td>~DEPRECATED-ETV6-NTRK3</td>\n",
       "      <td>Other</td>\n",
       "      <td>False</td>\n",
       "      <td>10717</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3764 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      variant_id gene_name            variant_name                category   \n",
       "0           4417       ALK             FBXO11::ALK                  Fusion  \\\n",
       "1           4188       VHL           10kb Deletion  Region Defined Variant   \n",
       "2            785     CHEK2                1100DELC     Protein Consequence   \n",
       "2            785     CHEK2                1100DELC     Protein Consequence   \n",
       "2            785     CHEK2                1100DELC     Protein Consequence   \n",
       "...          ...       ...                     ...                     ...   \n",
       "1201        3478      ESR2  underexpression beta-1                   Other   \n",
       "1201        3478      ESR2  underexpression beta-1                   Other   \n",
       "1202        3508     CD274                    v242     Protein Consequence   \n",
       "1203        2422     NTRK3  ~DEPRECATED-ETV6-NTRK3                   Other   \n",
       "1203        2422     NTRK3  ~DEPRECATED-ETV6-NTRK3                   Other   \n",
       "\n",
       "      variant_accepted evidence_ids evidence_status evidence_rating   \n",
       "0                False         7428     [submitted]             [3]  \\\n",
       "1                False        10678     [submitted]             [3]   \n",
       "2                 True         7235     [submitted]             [4]   \n",
       "2                 True         1849      [rejected]             [3]   \n",
       "2                 True         1850      [accepted]             [3]   \n",
       "...                ...          ...             ...             ...   \n",
       "1201             False         9618     [submitted]             [4]   \n",
       "1201             False         9619     [submitted]             [4]   \n",
       "1202             False         9695     [submitted]             [4]   \n",
       "1203             False        10692     [submitted]             [3]   \n",
       "1203             False        10717     [submitted]             [3]   \n",
       "\n",
       "     evidence_level  \n",
       "0               [C]  \n",
       "1               [C]  \n",
       "2               [B]  \n",
       "2               [B]  \n",
       "2               [B]  \n",
       "...             ...  \n",
       "1201            [B]  \n",
       "1201            [B]  \n",
       "1202            [E]  \n",
       "1203            [C]  \n",
       "1203            [C]  \n",
       "\n",
       "[3764 rows x 9 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ns_e_status = []\n",
    "ns_e_rating = []\n",
    "ns_e_level = []\n",
    "for e in ns_evidence_id_list:\n",
    "    ns_civic_e_status = []\n",
    "    ns_civic_e_rating = []\n",
    "    ns_civic_e_level = []\n",
    "    for evidence in civic_evidence_ids: \n",
    "        # print(civic_id, type(civic_id))\n",
    "        # print(   = =variant.gene_id, type(variant.gene_id))\n",
    "        if int(e) == evidence.id:\n",
    "            if evidence.status not in ns_civic_e_status:\n",
    "                ns_civic_e_status.append(evidence.status)\n",
    "            ns_civic_e_status = ns_civic_e_status or \"\"\n",
    "            if evidence.rating not in ns_civic_e_rating:\n",
    "                ns_civic_e_rating.append(evidence.rating)\n",
    "            ns_civic_e_rating = ns_civic_e_rating or \"\"\n",
    "            if evidence.evidence_level not in ns_civic_e_level:\n",
    "                ns_civic_e_level.append(evidence.evidence_level)\n",
    "            ns_civic_e_level = ns_civic_e_level or \"\"\n",
    "    ns_e_status.append(ns_civic_e_status)\n",
    "    ns_e_rating.append(ns_civic_e_rating)\n",
    "    ns_e_level.append(ns_civic_e_level)\n",
    "non_supp_var_df[\"evidence_status\"] = ns_e_status\n",
    "non_supp_var_df[\"evidence_rating\"] = ns_e_rating\n",
    "non_supp_var_df[\"evidence_level\"] = ns_e_level\n",
    "non_supp_var_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "expression_non_supp_var_df = non_supp_var_df[non_supp_var_df.category == 'Expression']\n",
    "epigenetic_modification_non_supp_var_df = non_supp_var_df[non_supp_var_df.category == 'Epigenetic Modification']\n",
    "fusion_non_supp_var_df = non_supp_var_df[non_supp_var_df.category == 'Fusion']\n",
    "protein_consequence_non_supp_var_df = non_supp_var_df[non_supp_var_df.category == 'Protein Consequence']\n",
    "gene_function_non_supp_var_df = non_supp_var_df[non_supp_var_df.category == 'Gene Function']\n",
    "rearrangements_non_supp_var_df = non_supp_var_df[non_supp_var_df.category == 'Rearrangements']\n",
    "copy_number_non_supp_var_df = non_supp_var_df[non_supp_var_df.category == 'Copy Number']\n",
    "other_non_supp_var_df = non_supp_var_df[non_supp_var_df.category == 'Other']\n",
    "genotypes_easy_non_supp_var_df = non_supp_var_df[non_supp_var_df.category == 'Genotypes Easy']\n",
    "genotypes_compound_non_supp_var_df = non_supp_var_df[non_supp_var_df.category == 'Genotypes Compound']\n",
    "region_defined_variant_non_supp_var_df = non_supp_var_df[non_supp_var_df.category == 'Region Defined Variant']\n",
    "transcript_variant_non_supp_var_df = non_supp_var_df[non_supp_var_df.category == 'Transcript Variant']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>variant_name</th>\n",
       "      <th>category</th>\n",
       "      <th>variant_accepted</th>\n",
       "      <th>evidence_ids</th>\n",
       "      <th>evidence_status</th>\n",
       "      <th>evidence_rating</th>\n",
       "      <th>evidence_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>309</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>AUTOCRINE ACTIVATION</td>\n",
       "      <td>Expression</td>\n",
       "      <td>False</td>\n",
       "      <td>769</td>\n",
       "      <td>[rejected]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>309</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>AUTOCRINE ACTIVATION</td>\n",
       "      <td>Expression</td>\n",
       "      <td>False</td>\n",
       "      <td>768</td>\n",
       "      <td>[rejected]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>3385</td>\n",
       "      <td>CD44</td>\n",
       "      <td>CD44s</td>\n",
       "      <td>Expression</td>\n",
       "      <td>False</td>\n",
       "      <td>9426</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>341</td>\n",
       "      <td>CD44</td>\n",
       "      <td>CD44s Expression</td>\n",
       "      <td>Expression</td>\n",
       "      <td>True</td>\n",
       "      <td>825</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>3386</td>\n",
       "      <td>CD44</td>\n",
       "      <td>CD44v5 and CD44v6</td>\n",
       "      <td>Expression</td>\n",
       "      <td>False</td>\n",
       "      <td>9432</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1183</th>\n",
       "      <td>272</td>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>p16 Expression</td>\n",
       "      <td>Expression</td>\n",
       "      <td>True</td>\n",
       "      <td>753</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1183</th>\n",
       "      <td>272</td>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>p16 Expression</td>\n",
       "      <td>Expression</td>\n",
       "      <td>True</td>\n",
       "      <td>804</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1183</th>\n",
       "      <td>272</td>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>p16 Expression</td>\n",
       "      <td>Expression</td>\n",
       "      <td>True</td>\n",
       "      <td>1155</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1183</th>\n",
       "      <td>272</td>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>p16 Expression</td>\n",
       "      <td>Expression</td>\n",
       "      <td>True</td>\n",
       "      <td>1314</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1183</th>\n",
       "      <td>272</td>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>p16 Expression</td>\n",
       "      <td>Expression</td>\n",
       "      <td>True</td>\n",
       "      <td>4873</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>619 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      variant_id gene_name          variant_name    category   \n",
       "33           309      EGFR  AUTOCRINE ACTIVATION  Expression  \\\n",
       "33           309      EGFR  AUTOCRINE ACTIVATION  Expression   \n",
       "126         3385      CD44                 CD44s  Expression   \n",
       "127          341      CD44      CD44s Expression  Expression   \n",
       "128         3386      CD44     CD44v5 and CD44v6  Expression   \n",
       "...          ...       ...                   ...         ...   \n",
       "1183         272    CDKN2A        p16 Expression  Expression   \n",
       "1183         272    CDKN2A        p16 Expression  Expression   \n",
       "1183         272    CDKN2A        p16 Expression  Expression   \n",
       "1183         272    CDKN2A        p16 Expression  Expression   \n",
       "1183         272    CDKN2A        p16 Expression  Expression   \n",
       "\n",
       "      variant_accepted evidence_ids evidence_status evidence_rating   \n",
       "33               False          769      [rejected]             [2]  \\\n",
       "33               False          768      [rejected]             [3]   \n",
       "126              False         9426     [submitted]             [4]   \n",
       "127               True          825      [accepted]             [3]   \n",
       "128              False         9432     [submitted]             [4]   \n",
       "...                ...          ...             ...             ...   \n",
       "1183              True          753      [accepted]             [3]   \n",
       "1183              True          804      [accepted]             [3]   \n",
       "1183              True         1155      [accepted]             [2]   \n",
       "1183              True         1314      [accepted]             [2]   \n",
       "1183              True         4873     [submitted]             [4]   \n",
       "\n",
       "     evidence_level  \n",
       "33              [B]  \n",
       "33              [D]  \n",
       "126             [B]  \n",
       "127             [D]  \n",
       "128             [B]  \n",
       "...             ...  \n",
       "1183            [B]  \n",
       "1183            [B]  \n",
       "1183            [B]  \n",
       "1183            [B]  \n",
       "1183            [D]  \n",
       "\n",
       "[619 rows x 9 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expression_non_supp_var_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>variant_name</th>\n",
       "      <th>category</th>\n",
       "      <th>variant_accepted</th>\n",
       "      <th>evidence_ids</th>\n",
       "      <th>evidence_status</th>\n",
       "      <th>evidence_rating</th>\n",
       "      <th>evidence_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>613</th>\n",
       "      <td>538</td>\n",
       "      <td>MLH1</td>\n",
       "      <td>METHYLATION</td>\n",
       "      <td>Epigenetic Modification</td>\n",
       "      <td>True</td>\n",
       "      <td>1315</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>614</th>\n",
       "      <td>784</td>\n",
       "      <td>CDKN2B</td>\n",
       "      <td>METHYLATION</td>\n",
       "      <td>Epigenetic Modification</td>\n",
       "      <td>True</td>\n",
       "      <td>1848</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>908</th>\n",
       "      <td>387</td>\n",
       "      <td>EIF4EBP1</td>\n",
       "      <td>PHOSPHORYLATION</td>\n",
       "      <td>Epigenetic Modification</td>\n",
       "      <td>True</td>\n",
       "      <td>918</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>908</th>\n",
       "      <td>387</td>\n",
       "      <td>EIF4EBP1</td>\n",
       "      <td>PHOSPHORYLATION</td>\n",
       "      <td>Epigenetic Modification</td>\n",
       "      <td>True</td>\n",
       "      <td>932</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>909</th>\n",
       "      <td>395</td>\n",
       "      <td>RPS6</td>\n",
       "      <td>PHOSPHORYLATION</td>\n",
       "      <td>Epigenetic Modification</td>\n",
       "      <td>True</td>\n",
       "      <td>930</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>910</th>\n",
       "      <td>632</td>\n",
       "      <td>RB1</td>\n",
       "      <td>PHOSPHORYLATION</td>\n",
       "      <td>Epigenetic Modification</td>\n",
       "      <td>True</td>\n",
       "      <td>1609</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>919</th>\n",
       "      <td>22</td>\n",
       "      <td>CCND2</td>\n",
       "      <td>PROMOTER DEMETHYLATION</td>\n",
       "      <td>Epigenetic Modification</td>\n",
       "      <td>True</td>\n",
       "      <td>219</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>920</th>\n",
       "      <td>2254</td>\n",
       "      <td>DBI</td>\n",
       "      <td>PROMOTER DEMETHYLATION</td>\n",
       "      <td>Epigenetic Modification</td>\n",
       "      <td>False</td>\n",
       "      <td>6033</td>\n",
       "      <td>[rejected]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>921</th>\n",
       "      <td>711</td>\n",
       "      <td>KLLN</td>\n",
       "      <td>PROMOTER METHYLATION</td>\n",
       "      <td>Epigenetic Modification</td>\n",
       "      <td>True</td>\n",
       "      <td>1761</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>921</th>\n",
       "      <td>711</td>\n",
       "      <td>KLLN</td>\n",
       "      <td>PROMOTER METHYLATION</td>\n",
       "      <td>Epigenetic Modification</td>\n",
       "      <td>True</td>\n",
       "      <td>1762</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>942</th>\n",
       "      <td>27</td>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>Promoter Hypermethylation</td>\n",
       "      <td>Epigenetic Modification</td>\n",
       "      <td>True</td>\n",
       "      <td>367</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>942</th>\n",
       "      <td>27</td>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>Promoter Hypermethylation</td>\n",
       "      <td>Epigenetic Modification</td>\n",
       "      <td>True</td>\n",
       "      <td>368</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>943</th>\n",
       "      <td>333</td>\n",
       "      <td>PTPRT</td>\n",
       "      <td>Promoter Hypermethylation</td>\n",
       "      <td>Epigenetic Modification</td>\n",
       "      <td>True</td>\n",
       "      <td>817</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>944</th>\n",
       "      <td>374</td>\n",
       "      <td>NAPRT</td>\n",
       "      <td>Promoter Hypermethylation</td>\n",
       "      <td>Epigenetic Modification</td>\n",
       "      <td>True</td>\n",
       "      <td>897</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>945</th>\n",
       "      <td>85</td>\n",
       "      <td>MGMT</td>\n",
       "      <td>Promoter Methylation</td>\n",
       "      <td>Epigenetic Modification</td>\n",
       "      <td>True</td>\n",
       "      <td>307</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>[A]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>945</th>\n",
       "      <td>85</td>\n",
       "      <td>MGMT</td>\n",
       "      <td>Promoter Methylation</td>\n",
       "      <td>Epigenetic Modification</td>\n",
       "      <td>True</td>\n",
       "      <td>7303</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[A]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>945</th>\n",
       "      <td>85</td>\n",
       "      <td>MGMT</td>\n",
       "      <td>Promoter Methylation</td>\n",
       "      <td>Epigenetic Modification</td>\n",
       "      <td>True</td>\n",
       "      <td>491</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>945</th>\n",
       "      <td>85</td>\n",
       "      <td>MGMT</td>\n",
       "      <td>Promoter Methylation</td>\n",
       "      <td>Epigenetic Modification</td>\n",
       "      <td>True</td>\n",
       "      <td>172</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>945</th>\n",
       "      <td>85</td>\n",
       "      <td>MGMT</td>\n",
       "      <td>Promoter Methylation</td>\n",
       "      <td>Epigenetic Modification</td>\n",
       "      <td>True</td>\n",
       "      <td>308</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>945</th>\n",
       "      <td>85</td>\n",
       "      <td>MGMT</td>\n",
       "      <td>Promoter Methylation</td>\n",
       "      <td>Epigenetic Modification</td>\n",
       "      <td>True</td>\n",
       "      <td>490</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>945</th>\n",
       "      <td>85</td>\n",
       "      <td>MGMT</td>\n",
       "      <td>Promoter Methylation</td>\n",
       "      <td>Epigenetic Modification</td>\n",
       "      <td>True</td>\n",
       "      <td>309</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[E]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1013</th>\n",
       "      <td>3257</td>\n",
       "      <td>AKT1</td>\n",
       "      <td>S473 Phosphorylation</td>\n",
       "      <td>Epigenetic Modification</td>\n",
       "      <td>True</td>\n",
       "      <td>4051</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1050</th>\n",
       "      <td>385</td>\n",
       "      <td>PRKAA2</td>\n",
       "      <td>T172 PHOSPHORYLATION</td>\n",
       "      <td>Epigenetic Modification</td>\n",
       "      <td>True</td>\n",
       "      <td>913</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[E]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1149</th>\n",
       "      <td>390</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>Y1092 PHOSPHORYLATION</td>\n",
       "      <td>Epigenetic Modification</td>\n",
       "      <td>True</td>\n",
       "      <td>923</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      variant_id gene_name               variant_name   \n",
       "613          538      MLH1                METHYLATION  \\\n",
       "614          784    CDKN2B                METHYLATION   \n",
       "908          387  EIF4EBP1            PHOSPHORYLATION   \n",
       "908          387  EIF4EBP1            PHOSPHORYLATION   \n",
       "909          395      RPS6            PHOSPHORYLATION   \n",
       "910          632       RB1            PHOSPHORYLATION   \n",
       "919           22     CCND2     PROMOTER DEMETHYLATION   \n",
       "920         2254       DBI     PROMOTER DEMETHYLATION   \n",
       "921          711      KLLN       PROMOTER METHYLATION   \n",
       "921          711      KLLN       PROMOTER METHYLATION   \n",
       "942           27    CDKN2A  Promoter Hypermethylation   \n",
       "942           27    CDKN2A  Promoter Hypermethylation   \n",
       "943          333     PTPRT  Promoter Hypermethylation   \n",
       "944          374     NAPRT  Promoter Hypermethylation   \n",
       "945           85      MGMT       Promoter Methylation   \n",
       "945           85      MGMT       Promoter Methylation   \n",
       "945           85      MGMT       Promoter Methylation   \n",
       "945           85      MGMT       Promoter Methylation   \n",
       "945           85      MGMT       Promoter Methylation   \n",
       "945           85      MGMT       Promoter Methylation   \n",
       "945           85      MGMT       Promoter Methylation   \n",
       "1013        3257      AKT1       S473 Phosphorylation   \n",
       "1050         385    PRKAA2       T172 PHOSPHORYLATION   \n",
       "1149         390      EGFR      Y1092 PHOSPHORYLATION   \n",
       "\n",
       "                     category  variant_accepted evidence_ids evidence_status   \n",
       "613   Epigenetic Modification              True         1315      [accepted]  \\\n",
       "614   Epigenetic Modification              True         1848      [accepted]   \n",
       "908   Epigenetic Modification              True          918      [accepted]   \n",
       "908   Epigenetic Modification              True          932      [accepted]   \n",
       "909   Epigenetic Modification              True          930      [accepted]   \n",
       "910   Epigenetic Modification              True         1609      [accepted]   \n",
       "919   Epigenetic Modification              True          219      [accepted]   \n",
       "920   Epigenetic Modification             False         6033      [rejected]   \n",
       "921   Epigenetic Modification              True         1761      [accepted]   \n",
       "921   Epigenetic Modification              True         1762      [accepted]   \n",
       "942   Epigenetic Modification              True          367      [accepted]   \n",
       "942   Epigenetic Modification              True          368      [accepted]   \n",
       "943   Epigenetic Modification              True          817      [accepted]   \n",
       "944   Epigenetic Modification              True          897      [accepted]   \n",
       "945   Epigenetic Modification              True          307      [accepted]   \n",
       "945   Epigenetic Modification              True         7303     [submitted]   \n",
       "945   Epigenetic Modification              True          491      [accepted]   \n",
       "945   Epigenetic Modification              True          172      [accepted]   \n",
       "945   Epigenetic Modification              True          308      [accepted]   \n",
       "945   Epigenetic Modification              True          490      [accepted]   \n",
       "945   Epigenetic Modification              True          309      [accepted]   \n",
       "1013  Epigenetic Modification              True         4051      [accepted]   \n",
       "1050  Epigenetic Modification              True          913      [accepted]   \n",
       "1149  Epigenetic Modification              True          923      [accepted]   \n",
       "\n",
       "     evidence_rating evidence_level  \n",
       "613              [2]            [B]  \n",
       "614              [4]            [B]  \n",
       "908              [2]            [D]  \n",
       "908              [2]            [D]  \n",
       "909              [2]            [D]  \n",
       "910              [3]            [B]  \n",
       "919              [3]            [D]  \n",
       "920              [5]            [B]  \n",
       "921              [3]            [B]  \n",
       "921              [3]            [B]  \n",
       "942              [3]            [B]  \n",
       "942              [3]            [B]  \n",
       "943              [2]            [D]  \n",
       "944              [3]            [D]  \n",
       "945              [5]            [A]  \n",
       "945              [3]            [A]  \n",
       "945              [5]            [B]  \n",
       "945              [4]            [B]  \n",
       "945              [4]            [B]  \n",
       "945              [4]            [B]  \n",
       "945              [4]            [E]  \n",
       "1013             [3]            [B]  \n",
       "1050             [2]            [E]  \n",
       "1149             [3]            [B]  "
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "epigenetic_modification_non_supp_var_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>variant_name</th>\n",
       "      <th>category</th>\n",
       "      <th>variant_accepted</th>\n",
       "      <th>evidence_ids</th>\n",
       "      <th>evidence_status</th>\n",
       "      <th>evidence_rating</th>\n",
       "      <th>evidence_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4417</td>\n",
       "      <td>ALK</td>\n",
       "      <td>FBXO11::ALK</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>False</td>\n",
       "      <td>7428</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>780</td>\n",
       "      <td>GLI1</td>\n",
       "      <td>ACTB::GLI1</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>True</td>\n",
       "      <td>1844</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>3702</td>\n",
       "      <td>NTRK2</td>\n",
       "      <td>AFAP1::NTRK2</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>True</td>\n",
       "      <td>10098</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>2577</td>\n",
       "      <td>PDGFRB</td>\n",
       "      <td>AGGF1::PDGFRB C843G</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>False</td>\n",
       "      <td>6999</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>285</td>\n",
       "      <td>BRAF</td>\n",
       "      <td>AGK::BRAF</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>True</td>\n",
       "      <td>723</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1161</th>\n",
       "      <td>466</td>\n",
       "      <td>FGFR1</td>\n",
       "      <td>ZNF198::FGFR1</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>True</td>\n",
       "      <td>1104</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1162</th>\n",
       "      <td>2864</td>\n",
       "      <td>JAK2</td>\n",
       "      <td>ZNF274::JAK2</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>True</td>\n",
       "      <td>7706</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1163</th>\n",
       "      <td>2977</td>\n",
       "      <td>JAK2</td>\n",
       "      <td>ZNF430::JAK2</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>True</td>\n",
       "      <td>7999</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1172</th>\n",
       "      <td>4406</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>fusion</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>False</td>\n",
       "      <td>11143</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1172</th>\n",
       "      <td>4406</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>fusion</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>False</td>\n",
       "      <td>11169</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1187 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      variant_id gene_name         variant_name category  variant_accepted   \n",
       "0           4417       ALK          FBXO11::ALK   Fusion             False  \\\n",
       "18           780      GLI1           ACTB::GLI1   Fusion              True   \n",
       "19          3702     NTRK2         AFAP1::NTRK2   Fusion              True   \n",
       "20          2577    PDGFRB  AGGF1::PDGFRB C843G   Fusion             False   \n",
       "21           285      BRAF            AGK::BRAF   Fusion              True   \n",
       "...          ...       ...                  ...      ...               ...   \n",
       "1161         466     FGFR1        ZNF198::FGFR1   Fusion              True   \n",
       "1162        2864      JAK2         ZNF274::JAK2   Fusion              True   \n",
       "1163        2977      JAK2         ZNF430::JAK2   Fusion              True   \n",
       "1172        4406      EGFR               fusion   Fusion             False   \n",
       "1172        4406      EGFR               fusion   Fusion             False   \n",
       "\n",
       "     evidence_ids evidence_status evidence_rating evidence_level  \n",
       "0            7428     [submitted]             [3]            [C]  \n",
       "18           1844      [accepted]             [3]            [B]  \n",
       "19          10098      [accepted]             [3]            [D]  \n",
       "20           6999     [submitted]             [3]            [D]  \n",
       "21            723      [accepted]             [2]            [C]  \n",
       "...           ...             ...             ...            ...  \n",
       "1161         1104      [accepted]             [4]            [C]  \n",
       "1162         7706      [accepted]             [3]            [C]  \n",
       "1163         7999      [accepted]             [3]            [C]  \n",
       "1172        11143     [submitted]             [1]            [C]  \n",
       "1172        11169     [submitted]             [1]            [C]  \n",
       "\n",
       "[1187 rows x 9 columns]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fusion_non_supp_var_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>variant_name</th>\n",
       "      <th>category</th>\n",
       "      <th>variant_accepted</th>\n",
       "      <th>evidence_ids</th>\n",
       "      <th>evidence_status</th>\n",
       "      <th>evidence_rating</th>\n",
       "      <th>evidence_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>785</td>\n",
       "      <td>CHEK2</td>\n",
       "      <td>1100DELC</td>\n",
       "      <td>Protein Consequence</td>\n",
       "      <td>True</td>\n",
       "      <td>7235</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>785</td>\n",
       "      <td>CHEK2</td>\n",
       "      <td>1100DELC</td>\n",
       "      <td>Protein Consequence</td>\n",
       "      <td>True</td>\n",
       "      <td>1849</td>\n",
       "      <td>[rejected]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>785</td>\n",
       "      <td>CHEK2</td>\n",
       "      <td>1100DELC</td>\n",
       "      <td>Protein Consequence</td>\n",
       "      <td>True</td>\n",
       "      <td>1850</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>2727</td>\n",
       "      <td>ERBB2</td>\n",
       "      <td>A129</td>\n",
       "      <td>Protein Consequence</td>\n",
       "      <td>False</td>\n",
       "      <td>7346</td>\n",
       "      <td>[rejected]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>2729</td>\n",
       "      <td>TP53</td>\n",
       "      <td>A129</td>\n",
       "      <td>Protein Consequence</td>\n",
       "      <td>False</td>\n",
       "      <td>7348</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1154</th>\n",
       "      <td>3055</td>\n",
       "      <td>FLT3</td>\n",
       "      <td>Y693</td>\n",
       "      <td>Protein Consequence</td>\n",
       "      <td>False</td>\n",
       "      <td>8285</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1155</th>\n",
       "      <td>1672</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>Y69FS*11</td>\n",
       "      <td>Protein Consequence</td>\n",
       "      <td>False</td>\n",
       "      <td>4808</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[None]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1157</th>\n",
       "      <td>3410</td>\n",
       "      <td>TEK</td>\n",
       "      <td>Y897C and R915C</td>\n",
       "      <td>Protein Consequence</td>\n",
       "      <td>False</td>\n",
       "      <td>9476</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1182</th>\n",
       "      <td>3057</td>\n",
       "      <td>GATA1</td>\n",
       "      <td>p.Ser51Alafs*86</td>\n",
       "      <td>Protein Consequence</td>\n",
       "      <td>False</td>\n",
       "      <td>8302</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1202</th>\n",
       "      <td>3508</td>\n",
       "      <td>CD274</td>\n",
       "      <td>v242</td>\n",
       "      <td>Protein Consequence</td>\n",
       "      <td>False</td>\n",
       "      <td>9695</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[E]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>299 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      variant_id gene_name     variant_name             category   \n",
       "2            785     CHEK2         1100DELC  Protein Consequence  \\\n",
       "2            785     CHEK2         1100DELC  Protein Consequence   \n",
       "2            785     CHEK2         1100DELC  Protein Consequence   \n",
       "12          2727     ERBB2             A129  Protein Consequence   \n",
       "13          2729      TP53             A129  Protein Consequence   \n",
       "...          ...       ...              ...                  ...   \n",
       "1154        3055      FLT3             Y693  Protein Consequence   \n",
       "1155        1672      EGFR         Y69FS*11  Protein Consequence   \n",
       "1157        3410       TEK  Y897C and R915C  Protein Consequence   \n",
       "1182        3057     GATA1  p.Ser51Alafs*86  Protein Consequence   \n",
       "1202        3508     CD274             v242  Protein Consequence   \n",
       "\n",
       "      variant_accepted evidence_ids evidence_status evidence_rating   \n",
       "2                 True         7235     [submitted]             [4]  \\\n",
       "2                 True         1849      [rejected]             [3]   \n",
       "2                 True         1850      [accepted]             [3]   \n",
       "12               False         7346      [rejected]             [4]   \n",
       "13               False         7348     [submitted]             [4]   \n",
       "...                ...          ...             ...             ...   \n",
       "1154             False         8285     [submitted]             [3]   \n",
       "1155             False         4808     [submitted]          [None]   \n",
       "1157             False         9476     [submitted]             [2]   \n",
       "1182             False         8302     [submitted]             [2]   \n",
       "1202             False         9695     [submitted]             [4]   \n",
       "\n",
       "     evidence_level  \n",
       "2               [B]  \n",
       "2               [B]  \n",
       "2               [B]  \n",
       "12              [D]  \n",
       "13              [D]  \n",
       "...             ...  \n",
       "1154            [D]  \n",
       "1155            [D]  \n",
       "1157            [C]  \n",
       "1182            [C]  \n",
       "1202            [E]  \n",
       "\n",
       "[299 rows x 9 columns]"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein_consequence_non_supp_var_df "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>variant_name</th>\n",
       "      <th>category</th>\n",
       "      <th>variant_accepted</th>\n",
       "      <th>evidence_ids</th>\n",
       "      <th>evidence_status</th>\n",
       "      <th>evidence_rating</th>\n",
       "      <th>evidence_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>1296</td>\n",
       "      <td>CTNNB1</td>\n",
       "      <td>Activating Mutation</td>\n",
       "      <td>Gene Function</td>\n",
       "      <td>True</td>\n",
       "      <td>2988</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>2649</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>Activating Mutation</td>\n",
       "      <td>Gene Function</td>\n",
       "      <td>False</td>\n",
       "      <td>7180</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>2649</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>Activating Mutation</td>\n",
       "      <td>Gene Function</td>\n",
       "      <td>False</td>\n",
       "      <td>7182</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>2649</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>Activating Mutation</td>\n",
       "      <td>Gene Function</td>\n",
       "      <td>False</td>\n",
       "      <td>7184</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>2657</td>\n",
       "      <td>ERBB2</td>\n",
       "      <td>Activating Mutation</td>\n",
       "      <td>Gene Function</td>\n",
       "      <td>False</td>\n",
       "      <td>7201</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1133</th>\n",
       "      <td>312</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>VIII</td>\n",
       "      <td>Gene Function</td>\n",
       "      <td>True</td>\n",
       "      <td>773</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1133</th>\n",
       "      <td>312</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>VIII</td>\n",
       "      <td>Gene Function</td>\n",
       "      <td>True</td>\n",
       "      <td>8191</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1133</th>\n",
       "      <td>312</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>VIII</td>\n",
       "      <td>Gene Function</td>\n",
       "      <td>True</td>\n",
       "      <td>4500</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1133</th>\n",
       "      <td>312</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>VIII</td>\n",
       "      <td>Gene Function</td>\n",
       "      <td>True</td>\n",
       "      <td>8192</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1133</th>\n",
       "      <td>312</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>VIII</td>\n",
       "      <td>Gene Function</td>\n",
       "      <td>True</td>\n",
       "      <td>1017</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>345 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      variant_id gene_name         variant_name       category   \n",
       "34          1296    CTNNB1  Activating Mutation  Gene Function  \\\n",
       "35          2649      KRAS  Activating Mutation  Gene Function   \n",
       "35          2649      KRAS  Activating Mutation  Gene Function   \n",
       "35          2649      KRAS  Activating Mutation  Gene Function   \n",
       "36          2657     ERBB2  Activating Mutation  Gene Function   \n",
       "...          ...       ...                  ...            ...   \n",
       "1133         312      EGFR                 VIII  Gene Function   \n",
       "1133         312      EGFR                 VIII  Gene Function   \n",
       "1133         312      EGFR                 VIII  Gene Function   \n",
       "1133         312      EGFR                 VIII  Gene Function   \n",
       "1133         312      EGFR                 VIII  Gene Function   \n",
       "\n",
       "      variant_accepted evidence_ids evidence_status evidence_rating   \n",
       "34                True         2988      [accepted]             [4]  \\\n",
       "35               False         7180     [submitted]             [3]   \n",
       "35               False         7182     [submitted]             [2]   \n",
       "35               False         7184     [submitted]             [1]   \n",
       "36               False         7201     [submitted]             [3]   \n",
       "...                ...          ...             ...             ...   \n",
       "1133              True          773      [accepted]             [2]   \n",
       "1133              True         8191     [submitted]             [4]   \n",
       "1133              True         4500      [accepted]             [3]   \n",
       "1133              True         8192     [submitted]             [3]   \n",
       "1133              True         1017      [accepted]             [1]   \n",
       "\n",
       "     evidence_level  \n",
       "34              [D]  \n",
       "35              [B]  \n",
       "35              [D]  \n",
       "35              [D]  \n",
       "36              [B]  \n",
       "...             ...  \n",
       "1133            [C]  \n",
       "1133            [D]  \n",
       "1133            [D]  \n",
       "1133            [D]  \n",
       "1133            [D]  \n",
       "\n",
       "[345 rows x 9 columns]"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_function_non_supp_var_df "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>variant_name</th>\n",
       "      <th>category</th>\n",
       "      <th>variant_accepted</th>\n",
       "      <th>evidence_ids</th>\n",
       "      <th>evidence_status</th>\n",
       "      <th>evidence_rating</th>\n",
       "      <th>evidence_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>2390</td>\n",
       "      <td>MAP2K1</td>\n",
       "      <td>56_61QKQKVG&gt;R</td>\n",
       "      <td>Rearrangements</td>\n",
       "      <td>False</td>\n",
       "      <td>6384</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>709</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>Alu insertion</td>\n",
       "      <td>Rearrangements</td>\n",
       "      <td>True</td>\n",
       "      <td>1759</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>165</th>\n",
       "      <td>943</td>\n",
       "      <td>PDGFRA</td>\n",
       "      <td>D842_H845DELDIMH</td>\n",
       "      <td>Rearrangements</td>\n",
       "      <td>True</td>\n",
       "      <td>2460</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196</th>\n",
       "      <td>3009</td>\n",
       "      <td>ABL1</td>\n",
       "      <td>Double Ph</td>\n",
       "      <td>Rearrangements</td>\n",
       "      <td>True</td>\n",
       "      <td>8093</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>207</th>\n",
       "      <td>1577</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>E709_T710&gt;D</td>\n",
       "      <td>Rearrangements</td>\n",
       "      <td>False</td>\n",
       "      <td>4663</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[None]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1197</th>\n",
       "      <td>3524</td>\n",
       "      <td>MECOM</td>\n",
       "      <td>t(3;21)(q26.2;q22)</td>\n",
       "      <td>Rearrangements</td>\n",
       "      <td>False</td>\n",
       "      <td>9719</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1198</th>\n",
       "      <td>3264</td>\n",
       "      <td>VHL</td>\n",
       "      <td>t(3;6)(p12.3;q24.3)</td>\n",
       "      <td>Rearrangements</td>\n",
       "      <td>False</td>\n",
       "      <td>9063</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1199</th>\n",
       "      <td>4008</td>\n",
       "      <td>VHL</td>\n",
       "      <td>t(3;6)(p13;q25.1)</td>\n",
       "      <td>Rearrangements</td>\n",
       "      <td>False</td>\n",
       "      <td>10357</td>\n",
       "      <td>[rejected]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1199</th>\n",
       "      <td>4008</td>\n",
       "      <td>VHL</td>\n",
       "      <td>t(3;6)(p13;q25.1)</td>\n",
       "      <td>Rearrangements</td>\n",
       "      <td>False</td>\n",
       "      <td>10350</td>\n",
       "      <td>[rejected]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1200</th>\n",
       "      <td>3351</td>\n",
       "      <td>VHL</td>\n",
       "      <td>t(3;8)(p13;q24.1)</td>\n",
       "      <td>Rearrangements</td>\n",
       "      <td>False</td>\n",
       "      <td>9340</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>138 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      variant_id gene_name         variant_name        category   \n",
       "10          2390    MAP2K1        56_61QKQKVG>R  Rearrangements  \\\n",
       "40           709     BRCA1        Alu insertion  Rearrangements   \n",
       "165          943    PDGFRA     D842_H845DELDIMH  Rearrangements   \n",
       "196         3009      ABL1            Double Ph  Rearrangements   \n",
       "207         1577      EGFR          E709_T710>D  Rearrangements   \n",
       "...          ...       ...                  ...             ...   \n",
       "1197        3524     MECOM   t(3;21)(q26.2;q22)  Rearrangements   \n",
       "1198        3264       VHL  t(3;6)(p12.3;q24.3)  Rearrangements   \n",
       "1199        4008       VHL    t(3;6)(p13;q25.1)  Rearrangements   \n",
       "1199        4008       VHL    t(3;6)(p13;q25.1)  Rearrangements   \n",
       "1200        3351       VHL    t(3;8)(p13;q24.1)  Rearrangements   \n",
       "\n",
       "      variant_accepted evidence_ids evidence_status evidence_rating   \n",
       "10               False         6384     [submitted]             [3]  \\\n",
       "40                True         1759      [accepted]             [3]   \n",
       "165               True         2460      [accepted]             [2]   \n",
       "196               True         8093      [accepted]             [4]   \n",
       "207              False         4663     [submitted]          [None]   \n",
       "...                ...          ...             ...             ...   \n",
       "1197             False         9719     [submitted]             [2]   \n",
       "1198             False         9063     [submitted]             [2]   \n",
       "1199             False        10357      [rejected]             [3]   \n",
       "1199             False        10350      [rejected]             [2]   \n",
       "1200             False         9340     [submitted]             [3]   \n",
       "\n",
       "     evidence_level  \n",
       "10              [D]  \n",
       "40              [C]  \n",
       "165             [D]  \n",
       "196             [C]  \n",
       "207             [C]  \n",
       "...             ...  \n",
       "1197            [C]  \n",
       "1198            [C]  \n",
       "1199            [C]  \n",
       "1199            [C]  \n",
       "1200            [C]  \n",
       "\n",
       "[138 rows x 9 columns]"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rearrangements_non_supp_var_df "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>variant_name</th>\n",
       "      <th>category</th>\n",
       "      <th>variant_accepted</th>\n",
       "      <th>evidence_ids</th>\n",
       "      <th>evidence_status</th>\n",
       "      <th>evidence_rating</th>\n",
       "      <th>evidence_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>3195</td>\n",
       "      <td>VHL</td>\n",
       "      <td>A149fs (c.444dup)</td>\n",
       "      <td>Copy Number</td>\n",
       "      <td>False</td>\n",
       "      <td>9982</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>3195</td>\n",
       "      <td>VHL</td>\n",
       "      <td>A149fs (c.444dup)</td>\n",
       "      <td>Copy Number</td>\n",
       "      <td>False</td>\n",
       "      <td>8822</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>3195</td>\n",
       "      <td>VHL</td>\n",
       "      <td>A149fs (c.444dup)</td>\n",
       "      <td>Copy Number</td>\n",
       "      <td>False</td>\n",
       "      <td>6156</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>2144</td>\n",
       "      <td>VHL</td>\n",
       "      <td>A56fs (c.166dup)</td>\n",
       "      <td>Copy Number</td>\n",
       "      <td>True</td>\n",
       "      <td>5742</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>1579</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>A767_V769dupASV</td>\n",
       "      <td>Copy Number</td>\n",
       "      <td>True</td>\n",
       "      <td>4665</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[None]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1156</th>\n",
       "      <td>414</td>\n",
       "      <td>ERBB2</td>\n",
       "      <td>Y772_A775DUP</td>\n",
       "      <td>Copy Number</td>\n",
       "      <td>True</td>\n",
       "      <td>4428</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[None]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1156</th>\n",
       "      <td>414</td>\n",
       "      <td>ERBB2</td>\n",
       "      <td>Y772_A775DUP</td>\n",
       "      <td>Copy Number</td>\n",
       "      <td>True</td>\n",
       "      <td>1046</td>\n",
       "      <td>[rejected]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1156</th>\n",
       "      <td>414</td>\n",
       "      <td>ERBB2</td>\n",
       "      <td>Y772_A775DUP</td>\n",
       "      <td>Copy Number</td>\n",
       "      <td>True</td>\n",
       "      <td>1047</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1156</th>\n",
       "      <td>414</td>\n",
       "      <td>ERBB2</td>\n",
       "      <td>Y772_A775DUP</td>\n",
       "      <td>Copy Number</td>\n",
       "      <td>True</td>\n",
       "      <td>960</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1181</th>\n",
       "      <td>2411</td>\n",
       "      <td>SDHB</td>\n",
       "      <td>p.193_196dupSTSC (c.577_588dupAGCACCAGCTGC)</td>\n",
       "      <td>Copy Number</td>\n",
       "      <td>True</td>\n",
       "      <td>6442</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>69 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      variant_id gene_name                                 variant_name   \n",
       "15          3195       VHL                            A149fs (c.444dup)  \\\n",
       "15          3195       VHL                            A149fs (c.444dup)   \n",
       "15          3195       VHL                            A149fs (c.444dup)   \n",
       "16          2144       VHL                             A56fs (c.166dup)   \n",
       "17          1579      EGFR                              A767_V769dupASV   \n",
       "...          ...       ...                                          ...   \n",
       "1156         414     ERBB2                                 Y772_A775DUP   \n",
       "1156         414     ERBB2                                 Y772_A775DUP   \n",
       "1156         414     ERBB2                                 Y772_A775DUP   \n",
       "1156         414     ERBB2                                 Y772_A775DUP   \n",
       "1181        2411      SDHB  p.193_196dupSTSC (c.577_588dupAGCACCAGCTGC)   \n",
       "\n",
       "         category  variant_accepted evidence_ids evidence_status   \n",
       "15    Copy Number             False         9982     [submitted]  \\\n",
       "15    Copy Number             False         8822     [submitted]   \n",
       "15    Copy Number             False         6156     [submitted]   \n",
       "16    Copy Number              True         5742      [accepted]   \n",
       "17    Copy Number              True         4665      [accepted]   \n",
       "...           ...               ...          ...             ...   \n",
       "1156  Copy Number              True         4428     [submitted]   \n",
       "1156  Copy Number              True         1046      [rejected]   \n",
       "1156  Copy Number              True         1047      [accepted]   \n",
       "1156  Copy Number              True          960      [accepted]   \n",
       "1181  Copy Number              True         6442      [accepted]   \n",
       "\n",
       "     evidence_rating evidence_level  \n",
       "15               [3]            [C]  \n",
       "15               [2]            [C]  \n",
       "15               [1]            [C]  \n",
       "16               [3]            [C]  \n",
       "17            [None]            [C]  \n",
       "...              ...            ...  \n",
       "1156          [None]            [D]  \n",
       "1156             [4]            [D]  \n",
       "1156             [4]            [D]  \n",
       "1156             [3]            [D]  \n",
       "1181             [2]            [C]  \n",
       "\n",
       "[69 rows x 9 columns]"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "copy_number_non_supp_var_df "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>variant_name</th>\n",
       "      <th>category</th>\n",
       "      <th>variant_accepted</th>\n",
       "      <th>evidence_ids</th>\n",
       "      <th>evidence_status</th>\n",
       "      <th>evidence_rating</th>\n",
       "      <th>evidence_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2438</td>\n",
       "      <td>VHL</td>\n",
       "      <td>235 (CAG-TAG)</td>\n",
       "      <td>Other</td>\n",
       "      <td>False</td>\n",
       "      <td>6524</td>\n",
       "      <td>[rejected]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>265</td>\n",
       "      <td>TYMS</td>\n",
       "      <td>5' TANDEM REPEAT</td>\n",
       "      <td>Other</td>\n",
       "      <td>True</td>\n",
       "      <td>678</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>2211</td>\n",
       "      <td>BAP1</td>\n",
       "      <td>ALTERNATIVE TRANSCRIPT (ATI)</td>\n",
       "      <td>Other</td>\n",
       "      <td>True</td>\n",
       "      <td>5929</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>842</td>\n",
       "      <td>BRAF</td>\n",
       "      <td>APC</td>\n",
       "      <td>Other</td>\n",
       "      <td>False</td>\n",
       "      <td>1941</td>\n",
       "      <td>[rejected]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>3460</td>\n",
       "      <td>AR</td>\n",
       "      <td>AR alternative transcripts</td>\n",
       "      <td>Other</td>\n",
       "      <td>False</td>\n",
       "      <td>9578</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1201</th>\n",
       "      <td>3478</td>\n",
       "      <td>ESR2</td>\n",
       "      <td>underexpression beta-1</td>\n",
       "      <td>Other</td>\n",
       "      <td>False</td>\n",
       "      <td>9613</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1201</th>\n",
       "      <td>3478</td>\n",
       "      <td>ESR2</td>\n",
       "      <td>underexpression beta-1</td>\n",
       "      <td>Other</td>\n",
       "      <td>False</td>\n",
       "      <td>9618</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1201</th>\n",
       "      <td>3478</td>\n",
       "      <td>ESR2</td>\n",
       "      <td>underexpression beta-1</td>\n",
       "      <td>Other</td>\n",
       "      <td>False</td>\n",
       "      <td>9619</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1203</th>\n",
       "      <td>2422</td>\n",
       "      <td>NTRK3</td>\n",
       "      <td>~DEPRECATED-ETV6-NTRK3</td>\n",
       "      <td>Other</td>\n",
       "      <td>False</td>\n",
       "      <td>10692</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1203</th>\n",
       "      <td>2422</td>\n",
       "      <td>NTRK3</td>\n",
       "      <td>~DEPRECATED-ETV6-NTRK3</td>\n",
       "      <td>Other</td>\n",
       "      <td>False</td>\n",
       "      <td>10717</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>160 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      variant_id gene_name                  variant_name category   \n",
       "3           2438       VHL                 235 (CAG-TAG)    Other  \\\n",
       "9            265      TYMS              5' TANDEM REPEAT    Other   \n",
       "24          2211      BAP1  ALTERNATIVE TRANSCRIPT (ATI)    Other   \n",
       "25           842      BRAF                           APC    Other   \n",
       "27          3460        AR    AR alternative transcripts    Other   \n",
       "...          ...       ...                           ...      ...   \n",
       "1201        3478      ESR2        underexpression beta-1    Other   \n",
       "1201        3478      ESR2        underexpression beta-1    Other   \n",
       "1201        3478      ESR2        underexpression beta-1    Other   \n",
       "1203        2422     NTRK3        ~DEPRECATED-ETV6-NTRK3    Other   \n",
       "1203        2422     NTRK3        ~DEPRECATED-ETV6-NTRK3    Other   \n",
       "\n",
       "      variant_accepted evidence_ids evidence_status evidence_rating   \n",
       "3                False         6524      [rejected]             [3]  \\\n",
       "9                 True          678      [accepted]             [3]   \n",
       "24                True         5929      [accepted]             [3]   \n",
       "25               False         1941      [rejected]             [3]   \n",
       "27               False         9578     [submitted]             [4]   \n",
       "...                ...          ...             ...             ...   \n",
       "1201             False         9613     [submitted]             [4]   \n",
       "1201             False         9618     [submitted]             [4]   \n",
       "1201             False         9619     [submitted]             [4]   \n",
       "1203             False        10692     [submitted]             [3]   \n",
       "1203             False        10717     [submitted]             [3]   \n",
       "\n",
       "     evidence_level  \n",
       "3               [C]  \n",
       "9               [B]  \n",
       "24              [D]  \n",
       "25              [B]  \n",
       "27              [B]  \n",
       "...             ...  \n",
       "1201            [B]  \n",
       "1201            [B]  \n",
       "1201            [B]  \n",
       "1203            [C]  \n",
       "1203            [C]  \n",
       "\n",
       "[160 rows x 9 columns]"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "other_non_supp_var_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>variant_name</th>\n",
       "      <th>category</th>\n",
       "      <th>variant_accepted</th>\n",
       "      <th>evidence_ids</th>\n",
       "      <th>evidence_status</th>\n",
       "      <th>evidence_rating</th>\n",
       "      <th>evidence_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1083</th>\n",
       "      <td>729</td>\n",
       "      <td>UGT1A1</td>\n",
       "      <td>UGT1A1*28</td>\n",
       "      <td>Genotypes Easy</td>\n",
       "      <td>True</td>\n",
       "      <td>1792</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>[A]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1084</th>\n",
       "      <td>732</td>\n",
       "      <td>UGT1A1</td>\n",
       "      <td>UGT1A1*60</td>\n",
       "      <td>Genotypes Easy</td>\n",
       "      <td>True</td>\n",
       "      <td>1795</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1141</th>\n",
       "      <td>426</td>\n",
       "      <td>BRAF</td>\n",
       "      <td>WILD TYPE</td>\n",
       "      <td>Genotypes Easy</td>\n",
       "      <td>True</td>\n",
       "      <td>995</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1142</th>\n",
       "      <td>2366</td>\n",
       "      <td>PTEN</td>\n",
       "      <td>WILD TYPE</td>\n",
       "      <td>Genotypes Easy</td>\n",
       "      <td>False</td>\n",
       "      <td>6272</td>\n",
       "      <td>[rejected]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1143</th>\n",
       "      <td>2651</td>\n",
       "      <td>KIT</td>\n",
       "      <td>WILDTYPE</td>\n",
       "      <td>Genotypes Easy</td>\n",
       "      <td>True</td>\n",
       "      <td>4144</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1145</th>\n",
       "      <td>369</td>\n",
       "      <td>TP53</td>\n",
       "      <td>Wildtype</td>\n",
       "      <td>Genotypes Easy</td>\n",
       "      <td>True</td>\n",
       "      <td>875</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1145</th>\n",
       "      <td>369</td>\n",
       "      <td>TP53</td>\n",
       "      <td>Wildtype</td>\n",
       "      <td>Genotypes Easy</td>\n",
       "      <td>True</td>\n",
       "      <td>906</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1145</th>\n",
       "      <td>369</td>\n",
       "      <td>TP53</td>\n",
       "      <td>Wildtype</td>\n",
       "      <td>Genotypes Easy</td>\n",
       "      <td>True</td>\n",
       "      <td>1149</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1145</th>\n",
       "      <td>369</td>\n",
       "      <td>TP53</td>\n",
       "      <td>Wildtype</td>\n",
       "      <td>Genotypes Easy</td>\n",
       "      <td>True</td>\n",
       "      <td>2965</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1145</th>\n",
       "      <td>369</td>\n",
       "      <td>TP53</td>\n",
       "      <td>Wildtype</td>\n",
       "      <td>Genotypes Easy</td>\n",
       "      <td>True</td>\n",
       "      <td>2963</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1145</th>\n",
       "      <td>369</td>\n",
       "      <td>TP53</td>\n",
       "      <td>Wildtype</td>\n",
       "      <td>Genotypes Easy</td>\n",
       "      <td>True</td>\n",
       "      <td>2964</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1146</th>\n",
       "      <td>2174</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>Wildtype</td>\n",
       "      <td>Genotypes Easy</td>\n",
       "      <td>True</td>\n",
       "      <td>2512</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1146</th>\n",
       "      <td>2174</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>Wildtype</td>\n",
       "      <td>Genotypes Easy</td>\n",
       "      <td>True</td>\n",
       "      <td>6185</td>\n",
       "      <td>[rejected]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1147</th>\n",
       "      <td>2590</td>\n",
       "      <td>PIK3CA</td>\n",
       "      <td>Wildtype</td>\n",
       "      <td>Genotypes Easy</td>\n",
       "      <td>True</td>\n",
       "      <td>2038</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1147</th>\n",
       "      <td>2590</td>\n",
       "      <td>PIK3CA</td>\n",
       "      <td>Wildtype</td>\n",
       "      <td>Genotypes Easy</td>\n",
       "      <td>True</td>\n",
       "      <td>7047</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1148</th>\n",
       "      <td>3200</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>Wildtype</td>\n",
       "      <td>Genotypes Easy</td>\n",
       "      <td>True</td>\n",
       "      <td>7052</td>\n",
       "      <td>[rejected]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1148</th>\n",
       "      <td>3200</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>Wildtype</td>\n",
       "      <td>Genotypes Easy</td>\n",
       "      <td>True</td>\n",
       "      <td>11057</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1148</th>\n",
       "      <td>3200</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>Wildtype</td>\n",
       "      <td>Genotypes Easy</td>\n",
       "      <td>True</td>\n",
       "      <td>8840</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1184</th>\n",
       "      <td>3512</td>\n",
       "      <td>BRAF</td>\n",
       "      <td>p61BRAF(V600E)</td>\n",
       "      <td>Genotypes Easy</td>\n",
       "      <td>False</td>\n",
       "      <td>9703</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1184</th>\n",
       "      <td>3512</td>\n",
       "      <td>BRAF</td>\n",
       "      <td>p61BRAF(V600E)</td>\n",
       "      <td>Genotypes Easy</td>\n",
       "      <td>False</td>\n",
       "      <td>9702</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      variant_id gene_name    variant_name        category  variant_accepted   \n",
       "1083         729    UGT1A1       UGT1A1*28  Genotypes Easy              True  \\\n",
       "1084         732    UGT1A1       UGT1A1*60  Genotypes Easy              True   \n",
       "1141         426      BRAF       WILD TYPE  Genotypes Easy              True   \n",
       "1142        2366      PTEN       WILD TYPE  Genotypes Easy             False   \n",
       "1143        2651       KIT        WILDTYPE  Genotypes Easy              True   \n",
       "1145         369      TP53        Wildtype  Genotypes Easy              True   \n",
       "1145         369      TP53        Wildtype  Genotypes Easy              True   \n",
       "1145         369      TP53        Wildtype  Genotypes Easy              True   \n",
       "1145         369      TP53        Wildtype  Genotypes Easy              True   \n",
       "1145         369      TP53        Wildtype  Genotypes Easy              True   \n",
       "1145         369      TP53        Wildtype  Genotypes Easy              True   \n",
       "1146        2174      EGFR        Wildtype  Genotypes Easy              True   \n",
       "1146        2174      EGFR        Wildtype  Genotypes Easy              True   \n",
       "1147        2590    PIK3CA        Wildtype  Genotypes Easy              True   \n",
       "1147        2590    PIK3CA        Wildtype  Genotypes Easy              True   \n",
       "1148        3200      KRAS        Wildtype  Genotypes Easy              True   \n",
       "1148        3200      KRAS        Wildtype  Genotypes Easy              True   \n",
       "1148        3200      KRAS        Wildtype  Genotypes Easy              True   \n",
       "1184        3512      BRAF  p61BRAF(V600E)  Genotypes Easy             False   \n",
       "1184        3512      BRAF  p61BRAF(V600E)  Genotypes Easy             False   \n",
       "\n",
       "     evidence_ids evidence_status evidence_rating evidence_level  \n",
       "1083         1792      [accepted]             [5]            [A]  \n",
       "1084         1795      [accepted]             [4]            [B]  \n",
       "1141          995      [accepted]             [2]            [B]  \n",
       "1142         6272      [rejected]             [2]            [D]  \n",
       "1143         4144      [accepted]             [2]            [B]  \n",
       "1145          875      [accepted]             [3]            [B]  \n",
       "1145          906      [accepted]             [3]            [B]  \n",
       "1145         1149      [accepted]             [3]            [B]  \n",
       "1145         2965      [accepted]             [3]            [B]  \n",
       "1145         2963      [accepted]             [4]            [D]  \n",
       "1145         2964      [accepted]             [4]            [D]  \n",
       "1146         2512      [accepted]             [4]            [B]  \n",
       "1146         6185      [rejected]             [3]            [B]  \n",
       "1147         2038      [accepted]             [3]            [B]  \n",
       "1147         7047     [submitted]             [2]            [B]  \n",
       "1148         7052      [rejected]             [5]            [B]  \n",
       "1148        11057     [submitted]             [3]            [B]  \n",
       "1148         8840      [accepted]             [2]            [D]  \n",
       "1184         9703     [submitted]             [2]            [C]  \n",
       "1184         9702     [submitted]             [4]            [D]  "
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotypes_easy_non_supp_var_df "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>variant_name</th>\n",
       "      <th>category</th>\n",
       "      <th>variant_accepted</th>\n",
       "      <th>evidence_ids</th>\n",
       "      <th>evidence_status</th>\n",
       "      <th>evidence_rating</th>\n",
       "      <th>evidence_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>821</td>\n",
       "      <td>HLA-C</td>\n",
       "      <td>COPY-NEUTRAL LOSS OF HETEROZYGOSITY</td>\n",
       "      <td>Genotypes Compound</td>\n",
       "      <td>True</td>\n",
       "      <td>1899</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>738</td>\n",
       "      <td>DPYD</td>\n",
       "      <td>DPYD*13 HOMOZYGOSITY</td>\n",
       "      <td>Genotypes Compound</td>\n",
       "      <td>True</td>\n",
       "      <td>1801</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>[A]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>176</th>\n",
       "      <td>737</td>\n",
       "      <td>DPYD</td>\n",
       "      <td>DPYD*2A HOMOZYGOSITY</td>\n",
       "      <td>Genotypes Compound</td>\n",
       "      <td>True</td>\n",
       "      <td>1800</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>[A]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>548</th>\n",
       "      <td>302</td>\n",
       "      <td>PTCH1</td>\n",
       "      <td>LOH</td>\n",
       "      <td>Genotypes Compound</td>\n",
       "      <td>True</td>\n",
       "      <td>749</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>918</th>\n",
       "      <td>3235</td>\n",
       "      <td>PRKAR1A</td>\n",
       "      <td>PRKAR1A LOH, allelic imbalance 17q</td>\n",
       "      <td>Genotypes Compound</td>\n",
       "      <td>False</td>\n",
       "      <td>8936</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1044</th>\n",
       "      <td>3108</td>\n",
       "      <td>VHL</td>\n",
       "      <td>Single Allele Deletion</td>\n",
       "      <td>Genotypes Compound</td>\n",
       "      <td>False</td>\n",
       "      <td>8499</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1044</th>\n",
       "      <td>3108</td>\n",
       "      <td>VHL</td>\n",
       "      <td>Single Allele Deletion</td>\n",
       "      <td>Genotypes Compound</td>\n",
       "      <td>False</td>\n",
       "      <td>8580</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      variant_id gene_name                         variant_name   \n",
       "149          821     HLA-C  COPY-NEUTRAL LOSS OF HETEROZYGOSITY  \\\n",
       "175          738      DPYD                 DPYD*13 HOMOZYGOSITY   \n",
       "176          737      DPYD                 DPYD*2A HOMOZYGOSITY   \n",
       "548          302     PTCH1                                  LOH   \n",
       "918         3235   PRKAR1A   PRKAR1A LOH, allelic imbalance 17q   \n",
       "1044        3108       VHL               Single Allele Deletion   \n",
       "1044        3108       VHL               Single Allele Deletion   \n",
       "\n",
       "                category  variant_accepted evidence_ids evidence_status   \n",
       "149   Genotypes Compound              True         1899      [accepted]  \\\n",
       "175   Genotypes Compound              True         1801      [accepted]   \n",
       "176   Genotypes Compound              True         1800      [accepted]   \n",
       "548   Genotypes Compound              True          749      [accepted]   \n",
       "918   Genotypes Compound             False         8936     [submitted]   \n",
       "1044  Genotypes Compound             False         8499     [submitted]   \n",
       "1044  Genotypes Compound             False         8580     [submitted]   \n",
       "\n",
       "     evidence_rating evidence_level  \n",
       "149              [3]            [C]  \n",
       "175              [5]            [A]  \n",
       "176              [5]            [A]  \n",
       "548              [2]            [B]  \n",
       "918              [3]            [B]  \n",
       "1044             [3]            [C]  \n",
       "1044             [3]            [C]  "
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotypes_compound_non_supp_var_df "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>variant_name</th>\n",
       "      <th>category</th>\n",
       "      <th>variant_accepted</th>\n",
       "      <th>evidence_ids</th>\n",
       "      <th>evidence_status</th>\n",
       "      <th>evidence_rating</th>\n",
       "      <th>evidence_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4188</td>\n",
       "      <td>VHL</td>\n",
       "      <td>10kb Deletion</td>\n",
       "      <td>Region Defined Variant</td>\n",
       "      <td>False</td>\n",
       "      <td>10678</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4186</td>\n",
       "      <td>VHL</td>\n",
       "      <td>2kb Deletion</td>\n",
       "      <td>Region Defined Variant</td>\n",
       "      <td>False</td>\n",
       "      <td>10791</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4186</td>\n",
       "      <td>VHL</td>\n",
       "      <td>2kb Deletion</td>\n",
       "      <td>Region Defined Variant</td>\n",
       "      <td>False</td>\n",
       "      <td>10676</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>823</td>\n",
       "      <td>EPCAM</td>\n",
       "      <td>3' Exon Deletion</td>\n",
       "      <td>Region Defined Variant</td>\n",
       "      <td>True</td>\n",
       "      <td>1901</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>253</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>3' UTR MUTATION</td>\n",
       "      <td>Region Defined Variant</td>\n",
       "      <td>False</td>\n",
       "      <td>663</td>\n",
       "      <td>[rejected]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1030</th>\n",
       "      <td>536</td>\n",
       "      <td>SULT1E1</td>\n",
       "      <td>SNP</td>\n",
       "      <td>Region Defined Variant</td>\n",
       "      <td>False</td>\n",
       "      <td>1312</td>\n",
       "      <td>[rejected]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1048</th>\n",
       "      <td>491</td>\n",
       "      <td>HSPH1</td>\n",
       "      <td>T17 DELETION</td>\n",
       "      <td>Region Defined Variant</td>\n",
       "      <td>True</td>\n",
       "      <td>1164</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1169</th>\n",
       "      <td>2349</td>\n",
       "      <td>IL7R</td>\n",
       "      <td>exon 6 mutations</td>\n",
       "      <td>Region Defined Variant</td>\n",
       "      <td>False</td>\n",
       "      <td>6228</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1178</th>\n",
       "      <td>2845</td>\n",
       "      <td>HOXD8</td>\n",
       "      <td>mutation</td>\n",
       "      <td>Region Defined Variant</td>\n",
       "      <td>False</td>\n",
       "      <td>7670</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1179</th>\n",
       "      <td>4436</td>\n",
       "      <td>CHEK2</td>\n",
       "      <td>mutation</td>\n",
       "      <td>Region Defined Variant</td>\n",
       "      <td>True</td>\n",
       "      <td>11205</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[A]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>896 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      variant_id gene_name      variant_name                category   \n",
       "1           4188       VHL     10kb Deletion  Region Defined Variant  \\\n",
       "4           4186       VHL      2kb Deletion  Region Defined Variant   \n",
       "4           4186       VHL      2kb Deletion  Region Defined Variant   \n",
       "5            823     EPCAM  3' Exon Deletion  Region Defined Variant   \n",
       "6            253      EGFR   3' UTR MUTATION  Region Defined Variant   \n",
       "...          ...       ...               ...                     ...   \n",
       "1030         536   SULT1E1               SNP  Region Defined Variant   \n",
       "1048         491     HSPH1      T17 DELETION  Region Defined Variant   \n",
       "1169        2349      IL7R  exon 6 mutations  Region Defined Variant   \n",
       "1178        2845     HOXD8          mutation  Region Defined Variant   \n",
       "1179        4436     CHEK2          mutation  Region Defined Variant   \n",
       "\n",
       "      variant_accepted evidence_ids evidence_status evidence_rating   \n",
       "1                False        10678     [submitted]             [3]  \\\n",
       "4                False        10791     [submitted]             [3]   \n",
       "4                False        10676     [submitted]             [2]   \n",
       "5                 True         1901      [accepted]             [4]   \n",
       "6                False          663      [rejected]             [4]   \n",
       "...                ...          ...             ...             ...   \n",
       "1030             False         1312      [rejected]             [2]   \n",
       "1048              True         1164      [accepted]             [2]   \n",
       "1169             False         6228     [submitted]             [4]   \n",
       "1178             False         7670     [submitted]             [2]   \n",
       "1179              True        11205      [accepted]             [3]   \n",
       "\n",
       "     evidence_level  \n",
       "1               [C]  \n",
       "4               [C]  \n",
       "4               [C]  \n",
       "5               [B]  \n",
       "6               [B]  \n",
       "...             ...  \n",
       "1030            [B]  \n",
       "1048            [B]  \n",
       "1169            [C]  \n",
       "1178            [C]  \n",
       "1179            [A]  \n",
       "\n",
       "[896 rows x 9 columns]"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "region_defined_variant_non_supp_var_df "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>variant_name</th>\n",
       "      <th>category</th>\n",
       "      <th>variant_accepted</th>\n",
       "      <th>evidence_ids</th>\n",
       "      <th>evidence_status</th>\n",
       "      <th>evidence_rating</th>\n",
       "      <th>evidence_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [variant_id, gene_name, variant_name, category, variant_accepted, evidence_ids, evidence_status, evidence_rating, evidence_level]\n",
       "Index: []"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transcript_variant_non_supp_var_df "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.10.9 64-bit ('3.10.9')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "1bd4a0c8bc911393f4a19827cd0d555013b2fcb9ea63c8ddc9b9ae4ac59a8883"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
