{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analysis of Normalized Queries\n",
    "\n",
    "This notebook contains an analysis on CIViC evidence data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv\n",
    "import pandas as pd\n",
    "from civicpy import civic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# civic.load_cache(on_stale=\"ignore\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import a list of all the variant ID's in CIViC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "civic_variant_ids = civic.get_all_variants()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Total Number of variants in CIViC: 3509'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "total_number_variants = len(civic_variant_ids)\n",
    "f\"Total Number of variants in CIViC: {total_number_variants}\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import a list of all the evidence ID's from CIViC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "civic_evidence_ids = civic.get_all_evidence()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Total Number of evidence items in CIViC: 9956'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "total_number_evidences = len(civic_evidence_ids)\n",
    "f\"Total Number of evidence items in CIViC: {total_number_evidences}\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import queries that were normalized"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>query</th>\n",
       "      <th>query_type</th>\n",
       "      <th>variant_accepted</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2489</td>\n",
       "      <td>NC_000003.11:g.10191648_10191649insC</td>\n",
       "      <td>genomic</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1988</td>\n",
       "      <td>NC_000003.11:g.10191649A&gt;T</td>\n",
       "      <td>genomic</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2488</td>\n",
       "      <td>3-10191647-T-G</td>\n",
       "      <td>genomic</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1986</td>\n",
       "      <td>NC_000003.11:g.10191648G&gt;T</td>\n",
       "      <td>genomic</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1987</td>\n",
       "      <td>NC_000003.11:g.10191649A&gt;G</td>\n",
       "      <td>genomic</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1853</th>\n",
       "      <td>3161</td>\n",
       "      <td>3-10183878-G-A</td>\n",
       "      <td>genomic</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1854</th>\n",
       "      <td>877</td>\n",
       "      <td>NC_000020.11:g.58903752C&gt;T</td>\n",
       "      <td>genomic</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1855</th>\n",
       "      <td>731</td>\n",
       "      <td>NC_000003.11:g.37056036G&gt;A</td>\n",
       "      <td>genomic</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1856</th>\n",
       "      <td>3045</td>\n",
       "      <td>VHL p.F76del</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1857</th>\n",
       "      <td>3310</td>\n",
       "      <td>HDAC9 p.L33R</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1858 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      variant_id                                 query query_type   \n",
       "0           2489  NC_000003.11:g.10191648_10191649insC    genomic  \\\n",
       "1           1988            NC_000003.11:g.10191649A>T    genomic   \n",
       "2           2488                        3-10191647-T-G    genomic   \n",
       "3           1986            NC_000003.11:g.10191648G>T    genomic   \n",
       "4           1987            NC_000003.11:g.10191649A>G    genomic   \n",
       "...          ...                                   ...        ...   \n",
       "1853        3161                        3-10183878-G-A    genomic   \n",
       "1854         877            NC_000020.11:g.58903752C>T    genomic   \n",
       "1855         731            NC_000003.11:g.37056036G>A    genomic   \n",
       "1856        3045                          VHL p.F76del    protein   \n",
       "1857        3310                          HDAC9 p.L33R    protein   \n",
       "\n",
       "      variant_accepted  \n",
       "0                 True  \n",
       "1                 True  \n",
       "2                 True  \n",
       "3                 True  \n",
       "4                 True  \n",
       "...                ...  \n",
       "1853             False  \n",
       "1854              True  \n",
       "1855              True  \n",
       "1856             False  \n",
       "1857             False  \n",
       "\n",
       "[1858 rows x 4 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "normalized_queries_df = pd.read_csv(\"./able_to_normalize_queries.csv\", sep= \"\\t\")\n",
    "normalized_queries_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create a list of the variant ID's of the normalized variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "normalized_variant_id_list = list(normalized_queries_df[\"variant_id\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Total Number of variants that were normalized: 1858'"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "total_number_normalized_variants = len(normalized_variant_id_list)\n",
    "f\"Total Number of variants that were normalized: {total_number_normalized_variants}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Percentage of variants that were normalized: 52.94955827871188%'"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "percentage_normalized_variants = total_number_normalized_variants/total_number_variants*100\n",
    "f\"Percentage of variants that were normalized: {percentage_normalized_variants}%\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import evidence ID's associated with the normalized variants using variant ID\n",
    "\n",
    "Evidence items are linked to the molecular profiles associated with variant items. To pull evidence IDs, need to use molecular profiles. For more information on the structure of a variant item"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>query</th>\n",
       "      <th>query_type</th>\n",
       "      <th>variant_accepted</th>\n",
       "      <th>evidence_ids</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2489</td>\n",
       "      <td>NC_000003.11:g.10191648_10191649insC</td>\n",
       "      <td>genomic</td>\n",
       "      <td>True</td>\n",
       "      <td>[9347, 6724]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1988</td>\n",
       "      <td>NC_000003.11:g.10191649A&gt;T</td>\n",
       "      <td>genomic</td>\n",
       "      <td>True</td>\n",
       "      <td>[5336]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2488</td>\n",
       "      <td>3-10191647-T-G</td>\n",
       "      <td>genomic</td>\n",
       "      <td>True</td>\n",
       "      <td>[10779, 6723, 8258]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1986</td>\n",
       "      <td>NC_000003.11:g.10191648G&gt;T</td>\n",
       "      <td>genomic</td>\n",
       "      <td>True</td>\n",
       "      <td>[5334]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1987</td>\n",
       "      <td>NC_000003.11:g.10191649A&gt;G</td>\n",
       "      <td>genomic</td>\n",
       "      <td>True</td>\n",
       "      <td>[5335]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1853</th>\n",
       "      <td>3161</td>\n",
       "      <td>3-10183878-G-A</td>\n",
       "      <td>genomic</td>\n",
       "      <td>False</td>\n",
       "      <td>[8659]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1854</th>\n",
       "      <td>877</td>\n",
       "      <td>NC_000020.11:g.58903752C&gt;T</td>\n",
       "      <td>genomic</td>\n",
       "      <td>True</td>\n",
       "      <td>[1997, 2892, 2893, 2895, 2896, 2897, 2898, 199...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1855</th>\n",
       "      <td>731</td>\n",
       "      <td>NC_000003.11:g.37056036G&gt;A</td>\n",
       "      <td>genomic</td>\n",
       "      <td>True</td>\n",
       "      <td>[1794]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1856</th>\n",
       "      <td>3045</td>\n",
       "      <td>VHL p.F76del</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>[8240]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1857</th>\n",
       "      <td>3310</td>\n",
       "      <td>HDAC9 p.L33R</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>[9240]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1858 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      variant_id                                 query query_type   \n",
       "0           2489  NC_000003.11:g.10191648_10191649insC    genomic  \\\n",
       "1           1988            NC_000003.11:g.10191649A>T    genomic   \n",
       "2           2488                        3-10191647-T-G    genomic   \n",
       "3           1986            NC_000003.11:g.10191648G>T    genomic   \n",
       "4           1987            NC_000003.11:g.10191649A>G    genomic   \n",
       "...          ...                                   ...        ...   \n",
       "1853        3161                        3-10183878-G-A    genomic   \n",
       "1854         877            NC_000020.11:g.58903752C>T    genomic   \n",
       "1855         731            NC_000003.11:g.37056036G>A    genomic   \n",
       "1856        3045                          VHL p.F76del    protein   \n",
       "1857        3310                          HDAC9 p.L33R    protein   \n",
       "\n",
       "      variant_accepted                                       evidence_ids  \n",
       "0                 True                                       [9347, 6724]  \n",
       "1                 True                                             [5336]  \n",
       "2                 True                                [10779, 6723, 8258]  \n",
       "3                 True                                             [5334]  \n",
       "4                 True                                             [5335]  \n",
       "...                ...                                                ...  \n",
       "1853             False                                             [8659]  \n",
       "1854              True  [1997, 2892, 2893, 2895, 2896, 2897, 2898, 199...  \n",
       "1855              True                                             [1794]  \n",
       "1856             False                                             [8240]  \n",
       "1857             False                                             [9240]  \n",
       "\n",
       "[1858 rows x 5 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "normalized_variants_evidence_ids = []\n",
    "for v in normalized_variant_id_list:\n",
    "    variant_evidence_id_list = []\n",
    "    for variant in civic_variant_ids:\n",
    "        if int(v) == variant.id:\n",
    "            for mp in variant.molecular_profiles:\n",
    "                for e in mp.evidence_items:\n",
    "                    if e.id not in variant_evidence_id_list:\n",
    "                        variant_evidence_id_list.append(e.id)\n",
    "            variant_evidence_id_list = variant_evidence_id_list or \"\"\n",
    "    normalized_variants_evidence_ids.append(variant_evidence_id_list)\n",
    "normalized_queries_df[\"evidence_ids\"] = normalized_variants_evidence_ids\n",
    "normalized_queries_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Split the list of evidence ID's of each variant to be one per line"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>query</th>\n",
       "      <th>query_type</th>\n",
       "      <th>variant_accepted</th>\n",
       "      <th>evidence_ids</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2489</td>\n",
       "      <td>NC_000003.11:g.10191648_10191649insC</td>\n",
       "      <td>genomic</td>\n",
       "      <td>True</td>\n",
       "      <td>9347</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2489</td>\n",
       "      <td>NC_000003.11:g.10191648_10191649insC</td>\n",
       "      <td>genomic</td>\n",
       "      <td>True</td>\n",
       "      <td>6724</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1988</td>\n",
       "      <td>NC_000003.11:g.10191649A&gt;T</td>\n",
       "      <td>genomic</td>\n",
       "      <td>True</td>\n",
       "      <td>5336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2488</td>\n",
       "      <td>3-10191647-T-G</td>\n",
       "      <td>genomic</td>\n",
       "      <td>True</td>\n",
       "      <td>10779</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2488</td>\n",
       "      <td>3-10191647-T-G</td>\n",
       "      <td>genomic</td>\n",
       "      <td>True</td>\n",
       "      <td>6723</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1854</th>\n",
       "      <td>877</td>\n",
       "      <td>NC_000020.11:g.58903752C&gt;T</td>\n",
       "      <td>genomic</td>\n",
       "      <td>True</td>\n",
       "      <td>1996</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1854</th>\n",
       "      <td>877</td>\n",
       "      <td>NC_000020.11:g.58903752C&gt;T</td>\n",
       "      <td>genomic</td>\n",
       "      <td>True</td>\n",
       "      <td>1995</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1855</th>\n",
       "      <td>731</td>\n",
       "      <td>NC_000003.11:g.37056036G&gt;A</td>\n",
       "      <td>genomic</td>\n",
       "      <td>True</td>\n",
       "      <td>1794</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1856</th>\n",
       "      <td>3045</td>\n",
       "      <td>VHL p.F76del</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>8240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1857</th>\n",
       "      <td>3310</td>\n",
       "      <td>HDAC9 p.L33R</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>9240</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5896 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      variant_id                                 query query_type   \n",
       "0           2489  NC_000003.11:g.10191648_10191649insC    genomic  \\\n",
       "0           2489  NC_000003.11:g.10191648_10191649insC    genomic   \n",
       "1           1988            NC_000003.11:g.10191649A>T    genomic   \n",
       "2           2488                        3-10191647-T-G    genomic   \n",
       "2           2488                        3-10191647-T-G    genomic   \n",
       "...          ...                                   ...        ...   \n",
       "1854         877            NC_000020.11:g.58903752C>T    genomic   \n",
       "1854         877            NC_000020.11:g.58903752C>T    genomic   \n",
       "1855         731            NC_000003.11:g.37056036G>A    genomic   \n",
       "1856        3045                          VHL p.F76del    protein   \n",
       "1857        3310                          HDAC9 p.L33R    protein   \n",
       "\n",
       "      variant_accepted evidence_ids  \n",
       "0                 True         9347  \n",
       "0                 True         6724  \n",
       "1                 True         5336  \n",
       "2                 True        10779  \n",
       "2                 True         6723  \n",
       "...                ...          ...  \n",
       "1854              True         1996  \n",
       "1854              True         1995  \n",
       "1855              True         1794  \n",
       "1856             False         8240  \n",
       "1857             False         9240  \n",
       "\n",
       "[5896 rows x 5 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "normalized_queries_df = normalized_queries_df.explode(column=\"evidence_ids\")\n",
    "normalized_queries_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create a list of the evidence ID's of the normalized variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "normalized_variant_evidence_id_list = list(normalized_queries_df[\"evidence_ids\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Total Number of times evidence items were associated with a normalized variant: 5896'"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "total_number_normalized_variant_evidence_items = len(normalized_variant_evidence_id_list)\n",
    "f\"Total Number of times evidence items were associated with a normalized variant: {total_number_normalized_variant_evidence_items}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Total Number of unique evidence items associated with normalized variants: 5845'"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "total_number_normalized_variant_unique_evidence_items = len(set(normalized_queries_df.evidence_ids))\n",
    "f\"Total Number of unique evidence items associated with normalized variants: {total_number_normalized_variant_unique_evidence_items}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Percentage of all CIViC evidence items in Normalized variants: 58.708316593009236%'"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "normalized_percentage_evidence_items = total_number_normalized_variant_unique_evidence_items/total_number_evidences*100\n",
    "f\"Percentage of all CIViC evidence items in Normalized variants: {normalized_percentage_evidence_items}%\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import evidence status, rating, and level associated with a specific evidence ID\n",
    "    please see the CIViC documentation for evidence item attribute decriptions (https://civic.readthedocs.io/en/latest/model/evidence.html)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>query</th>\n",
       "      <th>query_type</th>\n",
       "      <th>variant_accepted</th>\n",
       "      <th>evidence_ids</th>\n",
       "      <th>evidence_status</th>\n",
       "      <th>evidence_rating</th>\n",
       "      <th>evidence_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2489</td>\n",
       "      <td>NC_000003.11:g.10191648_10191649insC</td>\n",
       "      <td>genomic</td>\n",
       "      <td>True</td>\n",
       "      <td>9347</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2489</td>\n",
       "      <td>NC_000003.11:g.10191648_10191649insC</td>\n",
       "      <td>genomic</td>\n",
       "      <td>True</td>\n",
       "      <td>6724</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1988</td>\n",
       "      <td>NC_000003.11:g.10191649A&gt;T</td>\n",
       "      <td>genomic</td>\n",
       "      <td>True</td>\n",
       "      <td>5336</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2488</td>\n",
       "      <td>3-10191647-T-G</td>\n",
       "      <td>genomic</td>\n",
       "      <td>True</td>\n",
       "      <td>10779</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2488</td>\n",
       "      <td>3-10191647-T-G</td>\n",
       "      <td>genomic</td>\n",
       "      <td>True</td>\n",
       "      <td>6723</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1854</th>\n",
       "      <td>877</td>\n",
       "      <td>NC_000020.11:g.58903752C&gt;T</td>\n",
       "      <td>genomic</td>\n",
       "      <td>True</td>\n",
       "      <td>1996</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1854</th>\n",
       "      <td>877</td>\n",
       "      <td>NC_000020.11:g.58903752C&gt;T</td>\n",
       "      <td>genomic</td>\n",
       "      <td>True</td>\n",
       "      <td>1995</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1855</th>\n",
       "      <td>731</td>\n",
       "      <td>NC_000003.11:g.37056036G&gt;A</td>\n",
       "      <td>genomic</td>\n",
       "      <td>True</td>\n",
       "      <td>1794</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1856</th>\n",
       "      <td>3045</td>\n",
       "      <td>VHL p.F76del</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>8240</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1857</th>\n",
       "      <td>3310</td>\n",
       "      <td>HDAC9 p.L33R</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>9240</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5896 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      variant_id                                 query query_type   \n",
       "0           2489  NC_000003.11:g.10191648_10191649insC    genomic  \\\n",
       "0           2489  NC_000003.11:g.10191648_10191649insC    genomic   \n",
       "1           1988            NC_000003.11:g.10191649A>T    genomic   \n",
       "2           2488                        3-10191647-T-G    genomic   \n",
       "2           2488                        3-10191647-T-G    genomic   \n",
       "...          ...                                   ...        ...   \n",
       "1854         877            NC_000020.11:g.58903752C>T    genomic   \n",
       "1854         877            NC_000020.11:g.58903752C>T    genomic   \n",
       "1855         731            NC_000003.11:g.37056036G>A    genomic   \n",
       "1856        3045                          VHL p.F76del    protein   \n",
       "1857        3310                          HDAC9 p.L33R    protein   \n",
       "\n",
       "      variant_accepted evidence_ids evidence_status evidence_rating   \n",
       "0                 True         9347     [submitted]             [3]  \\\n",
       "0                 True         6724      [accepted]             [2]   \n",
       "1                 True         5336      [accepted]             [2]   \n",
       "2                 True        10779     [submitted]             [3]   \n",
       "2                 True         6723      [accepted]             [2]   \n",
       "...                ...          ...             ...             ...   \n",
       "1854              True         1996      [accepted]             [3]   \n",
       "1854              True         1995      [accepted]             [2]   \n",
       "1855              True         1794      [accepted]             [1]   \n",
       "1856             False         8240     [submitted]             [2]   \n",
       "1857             False         9240     [submitted]             [4]   \n",
       "\n",
       "     evidence_level  \n",
       "0               [C]  \n",
       "0               [C]  \n",
       "1               [C]  \n",
       "2               [C]  \n",
       "2               [C]  \n",
       "...             ...  \n",
       "1854            [B]  \n",
       "1854            [B]  \n",
       "1855            [C]  \n",
       "1856            [C]  \n",
       "1857            [D]  \n",
       "\n",
       "[5896 rows x 8 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "normalized_variants_evidence_statuses = []\n",
    "normalized_variants_evidence_ratings = []\n",
    "normalized_variants_evidence_levels = []\n",
    "for e in normalized_variant_evidence_id_list:\n",
    "    variant_evidence_status_list = []\n",
    "    variant_evidence_rating_list = []\n",
    "    variant_evidence_level_list = []\n",
    "    for evidence in civic_evidence_ids: \n",
    "        if int(e) == evidence.id:\n",
    "            if evidence.status not in variant_evidence_status_list:\n",
    "                variant_evidence_status_list.append(evidence.status)\n",
    "            variant_evidence_status_list = variant_evidence_status_list or \"\"\n",
    "            if evidence.rating not in variant_evidence_rating_list:\n",
    "                variant_evidence_rating_list.append(evidence.rating)\n",
    "            variant_evidence_rating_list = variant_evidence_rating_list or \"\"\n",
    "            if evidence.evidence_level not in variant_evidence_level_list:\n",
    "                variant_evidence_level_list.append(evidence.evidence_level)\n",
    "            variant_evidence_level_list = variant_evidence_level_list or \"\"\n",
    "    normalized_variants_evidence_statuses.append(variant_evidence_status_list)\n",
    "    normalized_variants_evidence_ratings.append(variant_evidence_rating_list)\n",
    "    normalized_variants_evidence_levels.append(variant_evidence_level_list)\n",
    "normalized_queries_df[\"evidence_status\"] = normalized_variants_evidence_statuses\n",
    "normalized_queries_df[\"evidence_rating\"] = normalized_variants_evidence_ratings\n",
    "normalized_queries_df[\"evidence_level\"] = normalized_variants_evidence_levels\n",
    "normalized_queries_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analysis of Unable to Normalize Queries"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import queries that were unable to normalize"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>query</th>\n",
       "      <th>query_type</th>\n",
       "      <th>variant_accepted</th>\n",
       "      <th>exception_raised</th>\n",
       "      <th>message</th>\n",
       "      <th>warnings</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>748</td>\n",
       "      <td>MLH1 *757L</td>\n",
       "      <td>protein</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3718</td>\n",
       "      <td>AR A748V</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3725</td>\n",
       "      <td>AR A765T</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>248</td>\n",
       "      <td>TERT C228T</td>\n",
       "      <td>protein</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4004</td>\n",
       "      <td>TERT C250T</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>3454</td>\n",
       "      <td>BRAF V599R</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>2825</td>\n",
       "      <td>BRAF V601E</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>3721</td>\n",
       "      <td>AR V757A</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>3722</td>\n",
       "      <td>AR V757I</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>3724</td>\n",
       "      <td>AR Y763C</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>63 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    variant_id       query query_type  variant_accepted  exception_raised   \n",
       "0          748  MLH1 *757L    protein              True             False  \\\n",
       "1         3718    AR A748V    protein             False             False   \n",
       "2         3725    AR A765T    protein             False             False   \n",
       "3          248  TERT C228T    protein              True             False   \n",
       "4         4004  TERT C250T    protein             False             False   \n",
       "..         ...         ...        ...               ...               ...   \n",
       "58        3454  BRAF V599R    protein             False             False   \n",
       "59        2825  BRAF V601E    protein             False             False   \n",
       "60        3721    AR V757A    protein             False             False   \n",
       "61        3722    AR V757I    protein             False             False   \n",
       "62        3724    AR Y763C    protein             False             False   \n",
       "\n",
       "                message                                           warnings  \n",
       "0   unable to normalize  [\"Unable to find valid result for classificati...  \n",
       "1   unable to normalize  [\"Unable to find valid result for classificati...  \n",
       "2   unable to normalize  [\"Unable to find valid result for classificati...  \n",
       "3   unable to normalize  [\"Unable to find valid result for classificati...  \n",
       "4   unable to normalize  [\"Unable to find valid result for classificati...  \n",
       "..                  ...                                                ...  \n",
       "58  unable to normalize  [\"Unable to find valid result for classificati...  \n",
       "59  unable to normalize  [\"Unable to find valid result for classificati...  \n",
       "60  unable to normalize  [\"Unable to find valid result for classificati...  \n",
       "61  unable to normalize  [\"Unable to find valid result for classificati...  \n",
       "62  unable to normalize  [\"Unable to find valid result for classificati...  \n",
       "\n",
       "[63 rows x 7 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "not_normalized_quer_df = pd.read_csv(\"./unable_to_normalize_queries.csv\", sep= \"\\t\")\n",
    "not_normalized_quer_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create a list of the variant ID's of the variants unable to normalize"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "not_normalized_variant_id_list = list(not_normalized_quer_df[\"variant_id\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Total Number of variants that were not normalized: 63'"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "total_number_not_normalized_variants = len(not_normalized_variant_id_list)\n",
    "f\"Total Number of variants that were not normalized: {total_number_not_normalized_variants}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Percentage of variants that were not normalized: 1.7953833000854944%'"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "percentage_not_normalized_variants = total_number_not_normalized_variants/total_number_variants*100\n",
    "f\"Percentage of variants that were not normalized: {percentage_not_normalized_variants}%\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import evidence ID's associated with the unable to normalize variants using variant ID\n",
    "\n",
    "Evidence items are linked to the molecular profiles associated with variant items. To pull evidence IDs, need to use molecular profiles. For more information on the structure of a variant item"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>query</th>\n",
       "      <th>query_type</th>\n",
       "      <th>variant_accepted</th>\n",
       "      <th>exception_raised</th>\n",
       "      <th>message</th>\n",
       "      <th>warnings</th>\n",
       "      <th>evidence_ids</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>748</td>\n",
       "      <td>MLH1 *757L</td>\n",
       "      <td>protein</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "      <td>[1812]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3718</td>\n",
       "      <td>AR A748V</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "      <td>[10128]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3725</td>\n",
       "      <td>AR A765T</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "      <td>[10135]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>248</td>\n",
       "      <td>TERT C228T</td>\n",
       "      <td>protein</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "      <td>[655, 1646, 6934, 6935]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4004</td>\n",
       "      <td>TERT C250T</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "      <td>[10331]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>3454</td>\n",
       "      <td>BRAF V599R</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "      <td>[9571]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>2825</td>\n",
       "      <td>BRAF V601E</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "      <td>[7633, 7627]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>3721</td>\n",
       "      <td>AR V757A</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "      <td>[10131]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>3722</td>\n",
       "      <td>AR V757I</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "      <td>[10132]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>3724</td>\n",
       "      <td>AR Y763C</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "      <td>[10134]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>63 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    variant_id       query query_type  variant_accepted  exception_raised   \n",
       "0          748  MLH1 *757L    protein              True             False  \\\n",
       "1         3718    AR A748V    protein             False             False   \n",
       "2         3725    AR A765T    protein             False             False   \n",
       "3          248  TERT C228T    protein              True             False   \n",
       "4         4004  TERT C250T    protein             False             False   \n",
       "..         ...         ...        ...               ...               ...   \n",
       "58        3454  BRAF V599R    protein             False             False   \n",
       "59        2825  BRAF V601E    protein             False             False   \n",
       "60        3721    AR V757A    protein             False             False   \n",
       "61        3722    AR V757I    protein             False             False   \n",
       "62        3724    AR Y763C    protein             False             False   \n",
       "\n",
       "                message                                           warnings   \n",
       "0   unable to normalize  [\"Unable to find valid result for classificati...  \\\n",
       "1   unable to normalize  [\"Unable to find valid result for classificati...   \n",
       "2   unable to normalize  [\"Unable to find valid result for classificati...   \n",
       "3   unable to normalize  [\"Unable to find valid result for classificati...   \n",
       "4   unable to normalize  [\"Unable to find valid result for classificati...   \n",
       "..                  ...                                                ...   \n",
       "58  unable to normalize  [\"Unable to find valid result for classificati...   \n",
       "59  unable to normalize  [\"Unable to find valid result for classificati...   \n",
       "60  unable to normalize  [\"Unable to find valid result for classificati...   \n",
       "61  unable to normalize  [\"Unable to find valid result for classificati...   \n",
       "62  unable to normalize  [\"Unable to find valid result for classificati...   \n",
       "\n",
       "               evidence_ids  \n",
       "0                    [1812]  \n",
       "1                   [10128]  \n",
       "2                   [10135]  \n",
       "3   [655, 1646, 6934, 6935]  \n",
       "4                   [10331]  \n",
       "..                      ...  \n",
       "58                   [9571]  \n",
       "59             [7633, 7627]  \n",
       "60                  [10131]  \n",
       "61                  [10132]  \n",
       "62                  [10134]  \n",
       "\n",
       "[63 rows x 8 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "not_normalized_variants_evidence_ids = []\n",
    "for v in not_normalized_variant_id_list:\n",
    "    not_normalized_variant_evidence_id_list = []\n",
    "    for variant in civic_variant_ids: \n",
    "        if int(v) == variant.id:\n",
    "            for mp in variant.molecular_profiles:\n",
    "                for e in mp.evidence_items:\n",
    "                    if e.id not in not_normalized_variant_evidence_id_list:\n",
    "                        not_normalized_variant_evidence_id_list.append(e.id)\n",
    "            not_normalized_variant_evidence_id_list = not_normalized_variant_evidence_id_list or \"\"\n",
    "    not_normalized_variants_evidence_ids.append(not_normalized_variant_evidence_id_list)\n",
    "not_normalized_quer_df[\"evidence_ids\"] = not_normalized_variants_evidence_ids\n",
    "not_normalized_quer_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Split the list of evidence ID's of each variant to be one per line"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>query</th>\n",
       "      <th>query_type</th>\n",
       "      <th>variant_accepted</th>\n",
       "      <th>exception_raised</th>\n",
       "      <th>message</th>\n",
       "      <th>warnings</th>\n",
       "      <th>evidence_ids</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>748</td>\n",
       "      <td>MLH1 *757L</td>\n",
       "      <td>protein</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "      <td>1812</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3718</td>\n",
       "      <td>AR A748V</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "      <td>10128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3725</td>\n",
       "      <td>AR A765T</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "      <td>10135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>248</td>\n",
       "      <td>TERT C228T</td>\n",
       "      <td>protein</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "      <td>655</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>248</td>\n",
       "      <td>TERT C228T</td>\n",
       "      <td>protein</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "      <td>1646</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>2825</td>\n",
       "      <td>BRAF V601E</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "      <td>7633</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>2825</td>\n",
       "      <td>BRAF V601E</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "      <td>7627</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>3721</td>\n",
       "      <td>AR V757A</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "      <td>10131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>3722</td>\n",
       "      <td>AR V757I</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "      <td>10132</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>3724</td>\n",
       "      <td>AR Y763C</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "      <td>10134</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>80 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    variant_id       query query_type  variant_accepted  exception_raised   \n",
       "0          748  MLH1 *757L    protein              True             False  \\\n",
       "1         3718    AR A748V    protein             False             False   \n",
       "2         3725    AR A765T    protein             False             False   \n",
       "3          248  TERT C228T    protein              True             False   \n",
       "3          248  TERT C228T    protein              True             False   \n",
       "..         ...         ...        ...               ...               ...   \n",
       "59        2825  BRAF V601E    protein             False             False   \n",
       "59        2825  BRAF V601E    protein             False             False   \n",
       "60        3721    AR V757A    protein             False             False   \n",
       "61        3722    AR V757I    protein             False             False   \n",
       "62        3724    AR Y763C    protein             False             False   \n",
       "\n",
       "                message                                           warnings   \n",
       "0   unable to normalize  [\"Unable to find valid result for classificati...  \\\n",
       "1   unable to normalize  [\"Unable to find valid result for classificati...   \n",
       "2   unable to normalize  [\"Unable to find valid result for classificati...   \n",
       "3   unable to normalize  [\"Unable to find valid result for classificati...   \n",
       "3   unable to normalize  [\"Unable to find valid result for classificati...   \n",
       "..                  ...                                                ...   \n",
       "59  unable to normalize  [\"Unable to find valid result for classificati...   \n",
       "59  unable to normalize  [\"Unable to find valid result for classificati...   \n",
       "60  unable to normalize  [\"Unable to find valid result for classificati...   \n",
       "61  unable to normalize  [\"Unable to find valid result for classificati...   \n",
       "62  unable to normalize  [\"Unable to find valid result for classificati...   \n",
       "\n",
       "   evidence_ids  \n",
       "0          1812  \n",
       "1         10128  \n",
       "2         10135  \n",
       "3           655  \n",
       "3          1646  \n",
       "..          ...  \n",
       "59         7633  \n",
       "59         7627  \n",
       "60        10131  \n",
       "61        10132  \n",
       "62        10134  \n",
       "\n",
       "[80 rows x 8 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "not_normalized_quer_df = not_normalized_quer_df.explode(column=\"evidence_ids\")\n",
    "not_normalized_quer_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create a list of the evidence ID's of the unable to normalize variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "not_normalized_evidence_id_list = list(not_normalized_quer_df[\"evidence_ids\"])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Total Number of times evidence items were associated with a not normalized variant: 80'"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "total_number_not_normalized_variant_evidence_items = len(not_normalized_evidence_id_list)\n",
    "f\"Total Number of times evidence items were associated with a not normalized variant: {total_number_not_normalized_variant_evidence_items}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Total Number of unique evidence items associated with a not normalized variant: 80'"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "total_number_not_normalized_variant_unique_evidence_items = len(not_normalized_evidence_id_list)\n",
    "f\"Total Number of unique evidence items associated with a not normalized variant: {total_number_not_normalized_variant_unique_evidence_items}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Percentage of all CIViC evidence items in not Normalized variants: 0.8035355564483728%'"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "not_normalized_percentage_evidence_items = total_number_not_normalized_variant_unique_evidence_items/total_number_evidences*100\n",
    "f\"Percentage of all CIViC evidence items in not Normalized variants: {not_normalized_percentage_evidence_items}%\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import evidence status, rating, and level associated with a specific evidence ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>query</th>\n",
       "      <th>query_type</th>\n",
       "      <th>variant_accepted</th>\n",
       "      <th>exception_raised</th>\n",
       "      <th>message</th>\n",
       "      <th>warnings</th>\n",
       "      <th>evidence_ids</th>\n",
       "      <th>evidence_status</th>\n",
       "      <th>evidence_rating</th>\n",
       "      <th>evidence_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>748</td>\n",
       "      <td>MLH1 *757L</td>\n",
       "      <td>protein</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "      <td>1812</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3718</td>\n",
       "      <td>AR A748V</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "      <td>10128</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3725</td>\n",
       "      <td>AR A765T</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "      <td>10135</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>248</td>\n",
       "      <td>TERT C228T</td>\n",
       "      <td>protein</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "      <td>655</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>248</td>\n",
       "      <td>TERT C228T</td>\n",
       "      <td>protein</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "      <td>1646</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>2825</td>\n",
       "      <td>BRAF V601E</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "      <td>7633</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>2825</td>\n",
       "      <td>BRAF V601E</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "      <td>7627</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>3721</td>\n",
       "      <td>AR V757A</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "      <td>10131</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>3722</td>\n",
       "      <td>AR V757I</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "      <td>10132</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>3724</td>\n",
       "      <td>AR Y763C</td>\n",
       "      <td>protein</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>unable to normalize</td>\n",
       "      <td>[\"Unable to find valid result for classificati...</td>\n",
       "      <td>10134</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>80 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    variant_id       query query_type  variant_accepted  exception_raised   \n",
       "0          748  MLH1 *757L    protein              True             False  \\\n",
       "1         3718    AR A748V    protein             False             False   \n",
       "2         3725    AR A765T    protein             False             False   \n",
       "3          248  TERT C228T    protein              True             False   \n",
       "3          248  TERT C228T    protein              True             False   \n",
       "..         ...         ...        ...               ...               ...   \n",
       "59        2825  BRAF V601E    protein             False             False   \n",
       "59        2825  BRAF V601E    protein             False             False   \n",
       "60        3721    AR V757A    protein             False             False   \n",
       "61        3722    AR V757I    protein             False             False   \n",
       "62        3724    AR Y763C    protein             False             False   \n",
       "\n",
       "                message                                           warnings   \n",
       "0   unable to normalize  [\"Unable to find valid result for classificati...  \\\n",
       "1   unable to normalize  [\"Unable to find valid result for classificati...   \n",
       "2   unable to normalize  [\"Unable to find valid result for classificati...   \n",
       "3   unable to normalize  [\"Unable to find valid result for classificati...   \n",
       "3   unable to normalize  [\"Unable to find valid result for classificati...   \n",
       "..                  ...                                                ...   \n",
       "59  unable to normalize  [\"Unable to find valid result for classificati...   \n",
       "59  unable to normalize  [\"Unable to find valid result for classificati...   \n",
       "60  unable to normalize  [\"Unable to find valid result for classificati...   \n",
       "61  unable to normalize  [\"Unable to find valid result for classificati...   \n",
       "62  unable to normalize  [\"Unable to find valid result for classificati...   \n",
       "\n",
       "   evidence_ids evidence_status evidence_rating evidence_level  \n",
       "0          1812      [accepted]             [1]            [C]  \n",
       "1         10128     [submitted]             [3]            [D]  \n",
       "2         10135     [submitted]             [3]            [D]  \n",
       "3           655      [accepted]             [5]            [B]  \n",
       "3          1646      [accepted]             [3]            [B]  \n",
       "..          ...             ...             ...            ...  \n",
       "59         7633     [submitted]             [3]            [B]  \n",
       "59         7627     [submitted]             [3]            [C]  \n",
       "60        10131     [submitted]             [3]            [D]  \n",
       "61        10132     [submitted]             [3]            [D]  \n",
       "62        10134     [submitted]             [3]            [D]  \n",
       "\n",
       "[80 rows x 11 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "not_normalized_variants_evidence_statuses = []\n",
    "not_normalized_variants_evidence_ratings = []\n",
    "not_normalized_variants_evidence_levels = []\n",
    "for e in not_normalized_evidence_id_list:\n",
    "    not_normalized_variant_evidence_status_list = []\n",
    "    not_normalized_variant_evidence_rating_list = []\n",
    "    not_normalized_variant_evidence_level_list = []\n",
    "    for evidence in civic_evidence_ids: \n",
    "        if int(e) == evidence.id:\n",
    "            if evidence.status not in not_normalized_variant_evidence_status_list:\n",
    "                not_normalized_variant_evidence_status_list.append(evidence.status)\n",
    "            not_normalized_variant_evidence_status_list = not_normalized_variant_evidence_status_list or \"\"\n",
    "            if evidence.rating not in not_normalized_variant_evidence_rating_list:\n",
    "                not_normalized_variant_evidence_rating_list.append(evidence.rating)\n",
    "            not_normalized_variant_evidence_rating_list = not_normalized_variant_evidence_rating_list or \"\"\n",
    "            if evidence.evidence_level not in not_normalized_variant_evidence_level_list:\n",
    "                not_normalized_variant_evidence_level_list.append(evidence.evidence_level)\n",
    "            not_normalized_variant_evidence_level_list = not_normalized_variant_evidence_level_list or \"\"\n",
    "    not_normalized_variants_evidence_statuses.append(not_normalized_variant_evidence_status_list)\n",
    "    not_normalized_variants_evidence_ratings.append(not_normalized_variant_evidence_rating_list)\n",
    "    not_normalized_variants_evidence_levels.append(not_normalized_variant_evidence_level_list)\n",
    "not_normalized_quer_df[\"evidence_status\"] = not_normalized_variants_evidence_statuses\n",
    "not_normalized_quer_df[\"evidence_rating\"] = not_normalized_variants_evidence_ratings\n",
    "not_normalized_quer_df[\"evidence_level\"] = not_normalized_variants_evidence_levels\n",
    "not_normalized_quer_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "80"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(not_normalized_quer_df.evidence_ids))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analysis of Not Supported Variants"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import variants that were not supported"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>variant_name</th>\n",
       "      <th>category</th>\n",
       "      <th>variant_accepted</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4417</td>\n",
       "      <td>ALK</td>\n",
       "      <td>FBXO11::ALK</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4188</td>\n",
       "      <td>VHL</td>\n",
       "      <td>10kb Deletion</td>\n",
       "      <td>Region Defined Variant</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>785</td>\n",
       "      <td>CHEK2</td>\n",
       "      <td>1100DELC</td>\n",
       "      <td>Protein Consequence</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2438</td>\n",
       "      <td>VHL</td>\n",
       "      <td>235 (CAG-TAG)</td>\n",
       "      <td>Other</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4186</td>\n",
       "      <td>VHL</td>\n",
       "      <td>2kb Deletion</td>\n",
       "      <td>Region Defined Variant</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1199</th>\n",
       "      <td>4008</td>\n",
       "      <td>VHL</td>\n",
       "      <td>t(3;6)(p13;q25.1)</td>\n",
       "      <td>Rearrangements</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1200</th>\n",
       "      <td>3351</td>\n",
       "      <td>VHL</td>\n",
       "      <td>t(3;8)(p13;q24.1)</td>\n",
       "      <td>Rearrangements</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1201</th>\n",
       "      <td>3478</td>\n",
       "      <td>ESR2</td>\n",
       "      <td>underexpression beta-1</td>\n",
       "      <td>Other</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1202</th>\n",
       "      <td>3508</td>\n",
       "      <td>CD274</td>\n",
       "      <td>v242</td>\n",
       "      <td>Protein Consequence</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1203</th>\n",
       "      <td>2422</td>\n",
       "      <td>NTRK3</td>\n",
       "      <td>~DEPRECATED-ETV6-NTRK3</td>\n",
       "      <td>Other</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1204 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      variant_id gene_name            variant_name                category   \n",
       "0           4417       ALK             FBXO11::ALK                  Fusion  \\\n",
       "1           4188       VHL           10kb Deletion  Region Defined Variant   \n",
       "2            785     CHEK2                1100DELC     Protein Consequence   \n",
       "3           2438       VHL           235 (CAG-TAG)                   Other   \n",
       "4           4186       VHL            2kb Deletion  Region Defined Variant   \n",
       "...          ...       ...                     ...                     ...   \n",
       "1199        4008       VHL       t(3;6)(p13;q25.1)          Rearrangements   \n",
       "1200        3351       VHL       t(3;8)(p13;q24.1)          Rearrangements   \n",
       "1201        3478      ESR2  underexpression beta-1                   Other   \n",
       "1202        3508     CD274                    v242     Protein Consequence   \n",
       "1203        2422     NTRK3  ~DEPRECATED-ETV6-NTRK3                   Other   \n",
       "\n",
       "      variant_accepted  \n",
       "0                False  \n",
       "1                False  \n",
       "2                 True  \n",
       "3                False  \n",
       "4                False  \n",
       "...                ...  \n",
       "1199             False  \n",
       "1200             False  \n",
       "1201             False  \n",
       "1202             False  \n",
       "1203             False  \n",
       "\n",
       "[1204 rows x 5 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "not_supported_variants_df = pd.read_csv(\"./not_supported_variants.csv\", sep= \"\\t\")\n",
    "not_supported_variants_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create a list of the variant ID's of the not supported variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "not_supported_variant_id_list = list(not_supported_variants_df[\"variant_id\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Total Number of variants that were not supported: 1204'"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "total_number_not_supported_variants = len(not_supported_variants_df)\n",
    "f\"Total Number of variants that were not supported: {total_number_not_supported_variants}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Percentage of variants that were not supported: 34.31176973496723%'"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "percentage_not_supported_variants = total_number_not_supported_variants/total_number_variants*100\n",
    "f\"Percentage of variants that were not supported: {percentage_not_supported_variants}%\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import evidence ID's associated with the not supported variants using variant ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>variant_name</th>\n",
       "      <th>category</th>\n",
       "      <th>variant_accepted</th>\n",
       "      <th>evidence_ids</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4417</td>\n",
       "      <td>ALK</td>\n",
       "      <td>FBXO11::ALK</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>False</td>\n",
       "      <td>[7428]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4188</td>\n",
       "      <td>VHL</td>\n",
       "      <td>10kb Deletion</td>\n",
       "      <td>Region Defined Variant</td>\n",
       "      <td>False</td>\n",
       "      <td>[10678]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>785</td>\n",
       "      <td>CHEK2</td>\n",
       "      <td>1100DELC</td>\n",
       "      <td>Protein Consequence</td>\n",
       "      <td>True</td>\n",
       "      <td>[7235, 1849, 1850]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2438</td>\n",
       "      <td>VHL</td>\n",
       "      <td>235 (CAG-TAG)</td>\n",
       "      <td>Other</td>\n",
       "      <td>False</td>\n",
       "      <td>[6524]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4186</td>\n",
       "      <td>VHL</td>\n",
       "      <td>2kb Deletion</td>\n",
       "      <td>Region Defined Variant</td>\n",
       "      <td>False</td>\n",
       "      <td>[10791, 10676]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1199</th>\n",
       "      <td>4008</td>\n",
       "      <td>VHL</td>\n",
       "      <td>t(3;6)(p13;q25.1)</td>\n",
       "      <td>Rearrangements</td>\n",
       "      <td>False</td>\n",
       "      <td>[10357, 10350]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1200</th>\n",
       "      <td>3351</td>\n",
       "      <td>VHL</td>\n",
       "      <td>t(3;8)(p13;q24.1)</td>\n",
       "      <td>Rearrangements</td>\n",
       "      <td>False</td>\n",
       "      <td>[9340]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1201</th>\n",
       "      <td>3478</td>\n",
       "      <td>ESR2</td>\n",
       "      <td>underexpression beta-1</td>\n",
       "      <td>Other</td>\n",
       "      <td>False</td>\n",
       "      <td>[9612, 9613, 9618, 9619]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1202</th>\n",
       "      <td>3508</td>\n",
       "      <td>CD274</td>\n",
       "      <td>v242</td>\n",
       "      <td>Protein Consequence</td>\n",
       "      <td>False</td>\n",
       "      <td>[9695]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1203</th>\n",
       "      <td>2422</td>\n",
       "      <td>NTRK3</td>\n",
       "      <td>~DEPRECATED-ETV6-NTRK3</td>\n",
       "      <td>Other</td>\n",
       "      <td>False</td>\n",
       "      <td>[10692, 10717]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1204 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      variant_id gene_name            variant_name                category   \n",
       "0           4417       ALK             FBXO11::ALK                  Fusion  \\\n",
       "1           4188       VHL           10kb Deletion  Region Defined Variant   \n",
       "2            785     CHEK2                1100DELC     Protein Consequence   \n",
       "3           2438       VHL           235 (CAG-TAG)                   Other   \n",
       "4           4186       VHL            2kb Deletion  Region Defined Variant   \n",
       "...          ...       ...                     ...                     ...   \n",
       "1199        4008       VHL       t(3;6)(p13;q25.1)          Rearrangements   \n",
       "1200        3351       VHL       t(3;8)(p13;q24.1)          Rearrangements   \n",
       "1201        3478      ESR2  underexpression beta-1                   Other   \n",
       "1202        3508     CD274                    v242     Protein Consequence   \n",
       "1203        2422     NTRK3  ~DEPRECATED-ETV6-NTRK3                   Other   \n",
       "\n",
       "      variant_accepted              evidence_ids  \n",
       "0                False                    [7428]  \n",
       "1                False                   [10678]  \n",
       "2                 True        [7235, 1849, 1850]  \n",
       "3                False                    [6524]  \n",
       "4                False            [10791, 10676]  \n",
       "...                ...                       ...  \n",
       "1199             False            [10357, 10350]  \n",
       "1200             False                    [9340]  \n",
       "1201             False  [9612, 9613, 9618, 9619]  \n",
       "1202             False                    [9695]  \n",
       "1203             False            [10692, 10717]  \n",
       "\n",
       "[1204 rows x 6 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "not_supported_variants_evidence_ids = []\n",
    "for v in not_supported_variant_id_list:\n",
    "    not_supported_variant_evidence_id_list = []\n",
    "    for variant in civic_variant_ids:\n",
    "        if int(v) == variant.id:\n",
    "            for mp in variant.molecular_profiles:\n",
    "                for e in mp.evidence_items:\n",
    "                    if e.id not in not_supported_variant_evidence_id_list:\n",
    "                        not_supported_variant_evidence_id_list.append(e.id)\n",
    "            not_supported_variant_evidence_id_list = not_supported_variant_evidence_id_list or \"\"\n",
    "    not_supported_variants_evidence_ids.append(not_supported_variant_evidence_id_list)\n",
    "not_supported_variants_df[\"evidence_ids\"] = not_supported_variants_evidence_ids\n",
    "not_supported_variants_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Split the list of evidence ID's of each variant to be one per line"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>variant_name</th>\n",
       "      <th>category</th>\n",
       "      <th>variant_accepted</th>\n",
       "      <th>evidence_ids</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4417</td>\n",
       "      <td>ALK</td>\n",
       "      <td>FBXO11::ALK</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>False</td>\n",
       "      <td>7428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4188</td>\n",
       "      <td>VHL</td>\n",
       "      <td>10kb Deletion</td>\n",
       "      <td>Region Defined Variant</td>\n",
       "      <td>False</td>\n",
       "      <td>10678</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>785</td>\n",
       "      <td>CHEK2</td>\n",
       "      <td>1100DELC</td>\n",
       "      <td>Protein Consequence</td>\n",
       "      <td>True</td>\n",
       "      <td>7235</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>785</td>\n",
       "      <td>CHEK2</td>\n",
       "      <td>1100DELC</td>\n",
       "      <td>Protein Consequence</td>\n",
       "      <td>True</td>\n",
       "      <td>1849</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>785</td>\n",
       "      <td>CHEK2</td>\n",
       "      <td>1100DELC</td>\n",
       "      <td>Protein Consequence</td>\n",
       "      <td>True</td>\n",
       "      <td>1850</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1201</th>\n",
       "      <td>3478</td>\n",
       "      <td>ESR2</td>\n",
       "      <td>underexpression beta-1</td>\n",
       "      <td>Other</td>\n",
       "      <td>False</td>\n",
       "      <td>9618</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1201</th>\n",
       "      <td>3478</td>\n",
       "      <td>ESR2</td>\n",
       "      <td>underexpression beta-1</td>\n",
       "      <td>Other</td>\n",
       "      <td>False</td>\n",
       "      <td>9619</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1202</th>\n",
       "      <td>3508</td>\n",
       "      <td>CD274</td>\n",
       "      <td>v242</td>\n",
       "      <td>Protein Consequence</td>\n",
       "      <td>False</td>\n",
       "      <td>9695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1203</th>\n",
       "      <td>2422</td>\n",
       "      <td>NTRK3</td>\n",
       "      <td>~DEPRECATED-ETV6-NTRK3</td>\n",
       "      <td>Other</td>\n",
       "      <td>False</td>\n",
       "      <td>10692</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1203</th>\n",
       "      <td>2422</td>\n",
       "      <td>NTRK3</td>\n",
       "      <td>~DEPRECATED-ETV6-NTRK3</td>\n",
       "      <td>Other</td>\n",
       "      <td>False</td>\n",
       "      <td>10717</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3766 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      variant_id gene_name            variant_name                category   \n",
       "0           4417       ALK             FBXO11::ALK                  Fusion  \\\n",
       "1           4188       VHL           10kb Deletion  Region Defined Variant   \n",
       "2            785     CHEK2                1100DELC     Protein Consequence   \n",
       "2            785     CHEK2                1100DELC     Protein Consequence   \n",
       "2            785     CHEK2                1100DELC     Protein Consequence   \n",
       "...          ...       ...                     ...                     ...   \n",
       "1201        3478      ESR2  underexpression beta-1                   Other   \n",
       "1201        3478      ESR2  underexpression beta-1                   Other   \n",
       "1202        3508     CD274                    v242     Protein Consequence   \n",
       "1203        2422     NTRK3  ~DEPRECATED-ETV6-NTRK3                   Other   \n",
       "1203        2422     NTRK3  ~DEPRECATED-ETV6-NTRK3                   Other   \n",
       "\n",
       "      variant_accepted evidence_ids  \n",
       "0                False         7428  \n",
       "1                False        10678  \n",
       "2                 True         7235  \n",
       "2                 True         1849  \n",
       "2                 True         1850  \n",
       "...                ...          ...  \n",
       "1201             False         9618  \n",
       "1201             False         9619  \n",
       "1202             False         9695  \n",
       "1203             False        10692  \n",
       "1203             False        10717  \n",
       "\n",
       "[3766 rows x 6 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "not_supported_variants_df = not_supported_variants_df.explode(column=\"evidence_ids\")\n",
    "not_supported_variants_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create a list of the evidence ID's of the not supported variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "not_supported_evidence_id_list = list(not_supported_variants_df[\"evidence_ids\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Total Number of times evidence items were associated with a not supported variant: 3766'"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "total_number_not_supported_variant_evidence_items = len(not_supported_evidence_id_list)\n",
    "f\"Total Number of times evidence items were associated with a not supported variant: {total_number_not_supported_variant_evidence_items}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Total Number of unique evidence items were associated with a not supported variant: 3738'"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "total_number_not_supported_variant_unique_evidence_items = len(set(not_supported_variants_df.evidence_ids))\n",
    "f\"Total Number of unique evidence items were associated with a not supported variant: {total_number_not_supported_variant_unique_evidence_items}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Percentage of all CIViC evidence items in Not Supported variants: 37.545198875050225%'"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "not_supported_variant_percentage_evidence_items = total_number_not_supported_variant_unique_evidence_items/total_number_evidences*100\n",
    "f\"Percentage of all CIViC evidence items in Not Supported variants: {not_supported_variant_percentage_evidence_items}%\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import evidence status, rating, and level associated with a specific evidence ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>variant_name</th>\n",
       "      <th>category</th>\n",
       "      <th>variant_accepted</th>\n",
       "      <th>evidence_ids</th>\n",
       "      <th>evidence_status</th>\n",
       "      <th>evidence_rating</th>\n",
       "      <th>evidence_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4417</td>\n",
       "      <td>ALK</td>\n",
       "      <td>FBXO11::ALK</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>False</td>\n",
       "      <td>7428</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4188</td>\n",
       "      <td>VHL</td>\n",
       "      <td>10kb Deletion</td>\n",
       "      <td>Region Defined Variant</td>\n",
       "      <td>False</td>\n",
       "      <td>10678</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>785</td>\n",
       "      <td>CHEK2</td>\n",
       "      <td>1100DELC</td>\n",
       "      <td>Protein Consequence</td>\n",
       "      <td>True</td>\n",
       "      <td>7235</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>785</td>\n",
       "      <td>CHEK2</td>\n",
       "      <td>1100DELC</td>\n",
       "      <td>Protein Consequence</td>\n",
       "      <td>True</td>\n",
       "      <td>1849</td>\n",
       "      <td>[rejected]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>785</td>\n",
       "      <td>CHEK2</td>\n",
       "      <td>1100DELC</td>\n",
       "      <td>Protein Consequence</td>\n",
       "      <td>True</td>\n",
       "      <td>1850</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1201</th>\n",
       "      <td>3478</td>\n",
       "      <td>ESR2</td>\n",
       "      <td>underexpression beta-1</td>\n",
       "      <td>Other</td>\n",
       "      <td>False</td>\n",
       "      <td>9618</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1201</th>\n",
       "      <td>3478</td>\n",
       "      <td>ESR2</td>\n",
       "      <td>underexpression beta-1</td>\n",
       "      <td>Other</td>\n",
       "      <td>False</td>\n",
       "      <td>9619</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1202</th>\n",
       "      <td>3508</td>\n",
       "      <td>CD274</td>\n",
       "      <td>v242</td>\n",
       "      <td>Protein Consequence</td>\n",
       "      <td>False</td>\n",
       "      <td>9695</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[E]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1203</th>\n",
       "      <td>2422</td>\n",
       "      <td>NTRK3</td>\n",
       "      <td>~DEPRECATED-ETV6-NTRK3</td>\n",
       "      <td>Other</td>\n",
       "      <td>False</td>\n",
       "      <td>10692</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1203</th>\n",
       "      <td>2422</td>\n",
       "      <td>NTRK3</td>\n",
       "      <td>~DEPRECATED-ETV6-NTRK3</td>\n",
       "      <td>Other</td>\n",
       "      <td>False</td>\n",
       "      <td>10717</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3766 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      variant_id gene_name            variant_name                category   \n",
       "0           4417       ALK             FBXO11::ALK                  Fusion  \\\n",
       "1           4188       VHL           10kb Deletion  Region Defined Variant   \n",
       "2            785     CHEK2                1100DELC     Protein Consequence   \n",
       "2            785     CHEK2                1100DELC     Protein Consequence   \n",
       "2            785     CHEK2                1100DELC     Protein Consequence   \n",
       "...          ...       ...                     ...                     ...   \n",
       "1201        3478      ESR2  underexpression beta-1                   Other   \n",
       "1201        3478      ESR2  underexpression beta-1                   Other   \n",
       "1202        3508     CD274                    v242     Protein Consequence   \n",
       "1203        2422     NTRK3  ~DEPRECATED-ETV6-NTRK3                   Other   \n",
       "1203        2422     NTRK3  ~DEPRECATED-ETV6-NTRK3                   Other   \n",
       "\n",
       "      variant_accepted evidence_ids evidence_status evidence_rating   \n",
       "0                False         7428     [submitted]             [3]  \\\n",
       "1                False        10678     [submitted]             [3]   \n",
       "2                 True         7235     [submitted]             [4]   \n",
       "2                 True         1849      [rejected]             [3]   \n",
       "2                 True         1850      [accepted]             [3]   \n",
       "...                ...          ...             ...             ...   \n",
       "1201             False         9618     [submitted]             [4]   \n",
       "1201             False         9619     [submitted]             [4]   \n",
       "1202             False         9695     [submitted]             [4]   \n",
       "1203             False        10692     [submitted]             [3]   \n",
       "1203             False        10717     [submitted]             [3]   \n",
       "\n",
       "     evidence_level  \n",
       "0               [C]  \n",
       "1               [C]  \n",
       "2               [B]  \n",
       "2               [B]  \n",
       "2               [B]  \n",
       "...             ...  \n",
       "1201            [B]  \n",
       "1201            [B]  \n",
       "1202            [E]  \n",
       "1203            [C]  \n",
       "1203            [C]  \n",
       "\n",
       "[3766 rows x 9 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "not_supported_variants_evidence_statuses = []\n",
    "not_supported_variants_evidence_ratings = []\n",
    "not_supported_variants_evidence_levels = []\n",
    "for e in not_supported_evidence_id_list:\n",
    "    not_supported_variant_evidence_status_list = []\n",
    "    not_supported_variant_evidence_rating_list = []\n",
    "    not_supported_variant_evidence_level_list = []\n",
    "    for evidence in civic_evidence_ids:\n",
    "        if int(e) == evidence.id:\n",
    "            if evidence.status not in not_supported_variant_evidence_status_list:\n",
    "                not_supported_variant_evidence_status_list.append(evidence.status)\n",
    "            not_supported_variant_evidence_status_list = not_supported_variant_evidence_status_list or \"\"\n",
    "            if evidence.rating not in not_supported_variant_evidence_rating_list:\n",
    "                not_supported_variant_evidence_rating_list.append(evidence.rating)\n",
    "            not_supported_variant_evidence_rating_list = not_supported_variant_evidence_rating_list or \"\"\n",
    "            if evidence.evidence_level not in not_supported_variant_evidence_level_list:\n",
    "                not_supported_variant_evidence_level_list.append(evidence.evidence_level)\n",
    "            not_supported_variant_evidence_level_list = not_supported_variant_evidence_level_list or \"\"\n",
    "    not_supported_variants_evidence_statuses.append(not_supported_variant_evidence_status_list)\n",
    "    not_supported_variants_evidence_ratings.append(not_supported_variant_evidence_rating_list)\n",
    "    not_supported_variants_evidence_levels.append(not_supported_variant_evidence_level_list)\n",
    "not_supported_variants_df[\"evidence_status\"] = not_supported_variants_evidence_statuses\n",
    "not_supported_variants_df[\"evidence_rating\"] = not_supported_variants_evidence_ratings\n",
    "not_supported_variants_df[\"evidence_level\"] = not_supported_variants_evidence_levels\n",
    "not_supported_variants_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## List all the possible variant categories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Fusion', 'Region Defined Variant', 'Protein Consequence', 'Other',\n",
       "       'Rearrangements', 'Copy Number', 'Expression', 'Gene Function',\n",
       "       'Genotypes Compound', 'Epigenetic Modification', 'Genotypes Easy'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "not_supported_variant_categories = not_supported_variants_df.category.unique()\n",
    "not_supported_variant_categories"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Find evidence ID's that qualify for more than one kind of variant category"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>variant_name</th>\n",
       "      <th>category</th>\n",
       "      <th>variant_accepted</th>\n",
       "      <th>evidence_ids</th>\n",
       "      <th>evidence_status</th>\n",
       "      <th>evidence_rating</th>\n",
       "      <th>evidence_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4417</td>\n",
       "      <td>ALK</td>\n",
       "      <td>FBXO11::ALK</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>False</td>\n",
       "      <td>7428</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>181</th>\n",
       "      <td>200</td>\n",
       "      <td>IKZF1</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>Gene Function</td>\n",
       "      <td>True</td>\n",
       "      <td>7786</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>189</th>\n",
       "      <td>2654</td>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>Gene Function</td>\n",
       "      <td>True</td>\n",
       "      <td>7786</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>192</th>\n",
       "      <td>4314</td>\n",
       "      <td>CDKN2B</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>Gene Function</td>\n",
       "      <td>False</td>\n",
       "      <td>7786</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193</th>\n",
       "      <td>4316</td>\n",
       "      <td>PAX5</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>Gene Function</td>\n",
       "      <td>False</td>\n",
       "      <td>7786</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194</th>\n",
       "      <td>4317</td>\n",
       "      <td>ERG</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>Gene Function</td>\n",
       "      <td>False</td>\n",
       "      <td>7786</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>218</th>\n",
       "      <td>4241</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>EGFR::SEPT14</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>True</td>\n",
       "      <td>11152</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>218</th>\n",
       "      <td>4241</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>EGFR::SEPT14</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>True</td>\n",
       "      <td>11154</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>224</th>\n",
       "      <td>1516</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>EGFRVIII</td>\n",
       "      <td>Gene Function</td>\n",
       "      <td>False</td>\n",
       "      <td>11152</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>224</th>\n",
       "      <td>1516</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>EGFRVIII</td>\n",
       "      <td>Gene Function</td>\n",
       "      <td>False</td>\n",
       "      <td>11154</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>225</th>\n",
       "      <td>5</td>\n",
       "      <td>ALK</td>\n",
       "      <td>EML4::ALK</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>True</td>\n",
       "      <td>444</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>225</th>\n",
       "      <td>5</td>\n",
       "      <td>ALK</td>\n",
       "      <td>EML4::ALK</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>True</td>\n",
       "      <td>1347</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>225</th>\n",
       "      <td>5</td>\n",
       "      <td>ALK</td>\n",
       "      <td>EML4::ALK</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>True</td>\n",
       "      <td>1348</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>225</th>\n",
       "      <td>5</td>\n",
       "      <td>ALK</td>\n",
       "      <td>EML4::ALK</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>True</td>\n",
       "      <td>1349</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>225</th>\n",
       "      <td>5</td>\n",
       "      <td>ALK</td>\n",
       "      <td>EML4::ALK</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>True</td>\n",
       "      <td>4608</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>232</th>\n",
       "      <td>2750</td>\n",
       "      <td>ALK</td>\n",
       "      <td>EML6::ALK</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>False</td>\n",
       "      <td>7428</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>390</th>\n",
       "      <td>4336</td>\n",
       "      <td>FLT3LG</td>\n",
       "      <td>Expression</td>\n",
       "      <td>Expression</td>\n",
       "      <td>False</td>\n",
       "      <td>11088</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>390</th>\n",
       "      <td>4336</td>\n",
       "      <td>FLT3LG</td>\n",
       "      <td>Expression</td>\n",
       "      <td>Expression</td>\n",
       "      <td>False</td>\n",
       "      <td>11089</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>429</th>\n",
       "      <td>2639</td>\n",
       "      <td>BRAF</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>False</td>\n",
       "      <td>7149</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>438</th>\n",
       "      <td>76</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>G12</td>\n",
       "      <td>Protein Consequence</td>\n",
       "      <td>True</td>\n",
       "      <td>11055</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>438</th>\n",
       "      <td>76</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>G12</td>\n",
       "      <td>Protein Consequence</td>\n",
       "      <td>True</td>\n",
       "      <td>11056</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>442</th>\n",
       "      <td>80</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>G13</td>\n",
       "      <td>Protein Consequence</td>\n",
       "      <td>True</td>\n",
       "      <td>11055</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>442</th>\n",
       "      <td>80</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>G13</td>\n",
       "      <td>Protein Consequence</td>\n",
       "      <td>True</td>\n",
       "      <td>11056</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>480</th>\n",
       "      <td>55</td>\n",
       "      <td>FLT3</td>\n",
       "      <td>ITD</td>\n",
       "      <td>Rearrangements</td>\n",
       "      <td>True</td>\n",
       "      <td>11035</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>580</th>\n",
       "      <td>4310</td>\n",
       "      <td>KDM6A</td>\n",
       "      <td>Loss</td>\n",
       "      <td>Gene Function</td>\n",
       "      <td>False</td>\n",
       "      <td>11035</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>651</th>\n",
       "      <td>208</td>\n",
       "      <td>NRAS</td>\n",
       "      <td>Mutation</td>\n",
       "      <td>Region Defined Variant</td>\n",
       "      <td>True</td>\n",
       "      <td>11061</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>665</th>\n",
       "      <td>336</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>Mutation</td>\n",
       "      <td>Region Defined Variant</td>\n",
       "      <td>True</td>\n",
       "      <td>11058</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>665</th>\n",
       "      <td>336</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>Mutation</td>\n",
       "      <td>Region Defined Variant</td>\n",
       "      <td>True</td>\n",
       "      <td>11062</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>665</th>\n",
       "      <td>336</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>Mutation</td>\n",
       "      <td>Region Defined Variant</td>\n",
       "      <td>True</td>\n",
       "      <td>11059</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>665</th>\n",
       "      <td>336</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>Mutation</td>\n",
       "      <td>Region Defined Variant</td>\n",
       "      <td>True</td>\n",
       "      <td>11061</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>667</th>\n",
       "      <td>399</td>\n",
       "      <td>BRAF</td>\n",
       "      <td>Mutation</td>\n",
       "      <td>Region Defined Variant</td>\n",
       "      <td>True</td>\n",
       "      <td>7149</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>667</th>\n",
       "      <td>399</td>\n",
       "      <td>BRAF</td>\n",
       "      <td>Mutation</td>\n",
       "      <td>Region Defined Variant</td>\n",
       "      <td>True</td>\n",
       "      <td>11061</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>784</th>\n",
       "      <td>193</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>Overexpression</td>\n",
       "      <td>Expression</td>\n",
       "      <td>True</td>\n",
       "      <td>11058</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>784</th>\n",
       "      <td>193</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>Overexpression</td>\n",
       "      <td>Expression</td>\n",
       "      <td>True</td>\n",
       "      <td>11062</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>784</th>\n",
       "      <td>193</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>Overexpression</td>\n",
       "      <td>Expression</td>\n",
       "      <td>True</td>\n",
       "      <td>11059</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>806</th>\n",
       "      <td>603</td>\n",
       "      <td>FLT3</td>\n",
       "      <td>Overexpression</td>\n",
       "      <td>Expression</td>\n",
       "      <td>True</td>\n",
       "      <td>11088</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>806</th>\n",
       "      <td>603</td>\n",
       "      <td>FLT3</td>\n",
       "      <td>Overexpression</td>\n",
       "      <td>Expression</td>\n",
       "      <td>True</td>\n",
       "      <td>11089</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>895</th>\n",
       "      <td>2350</td>\n",
       "      <td>CRLF2</td>\n",
       "      <td>P2RY8::CRLF2</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>False</td>\n",
       "      <td>7786</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>954</th>\n",
       "      <td>203</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>Q61</td>\n",
       "      <td>Protein Consequence</td>\n",
       "      <td>True</td>\n",
       "      <td>11055</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>954</th>\n",
       "      <td>203</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>Q61</td>\n",
       "      <td>Protein Consequence</td>\n",
       "      <td>True</td>\n",
       "      <td>11056</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1046</th>\n",
       "      <td>4389</td>\n",
       "      <td>ALK</td>\n",
       "      <td>T1151dup</td>\n",
       "      <td>Copy Number</td>\n",
       "      <td>True</td>\n",
       "      <td>444</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1046</th>\n",
       "      <td>4389</td>\n",
       "      <td>ALK</td>\n",
       "      <td>T1151dup</td>\n",
       "      <td>Copy Number</td>\n",
       "      <td>True</td>\n",
       "      <td>1347</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1046</th>\n",
       "      <td>4389</td>\n",
       "      <td>ALK</td>\n",
       "      <td>T1151dup</td>\n",
       "      <td>Copy Number</td>\n",
       "      <td>True</td>\n",
       "      <td>1348</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1046</th>\n",
       "      <td>4389</td>\n",
       "      <td>ALK</td>\n",
       "      <td>T1151dup</td>\n",
       "      <td>Copy Number</td>\n",
       "      <td>True</td>\n",
       "      <td>1349</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1046</th>\n",
       "      <td>4389</td>\n",
       "      <td>ALK</td>\n",
       "      <td>T1151dup</td>\n",
       "      <td>Copy Number</td>\n",
       "      <td>True</td>\n",
       "      <td>4608</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1079</th>\n",
       "      <td>4428</td>\n",
       "      <td>MYC</td>\n",
       "      <td>Translocation</td>\n",
       "      <td>Rearrangements</td>\n",
       "      <td>False</td>\n",
       "      <td>9813</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1080</th>\n",
       "      <td>4429</td>\n",
       "      <td>BCL2</td>\n",
       "      <td>Translocation</td>\n",
       "      <td>Rearrangements</td>\n",
       "      <td>False</td>\n",
       "      <td>9813</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1081</th>\n",
       "      <td>4430</td>\n",
       "      <td>BCL6</td>\n",
       "      <td>Translocation</td>\n",
       "      <td>Rearrangements</td>\n",
       "      <td>False</td>\n",
       "      <td>9813</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      variant_id gene_name    variant_name                category   \n",
       "0           4417       ALK     FBXO11::ALK                  Fusion  \\\n",
       "181          200     IKZF1        Deletion           Gene Function   \n",
       "189         2654    CDKN2A        Deletion           Gene Function   \n",
       "192         4314    CDKN2B        Deletion           Gene Function   \n",
       "193         4316      PAX5        Deletion           Gene Function   \n",
       "194         4317       ERG        Deletion           Gene Function   \n",
       "218         4241      EGFR    EGFR::SEPT14                  Fusion   \n",
       "218         4241      EGFR    EGFR::SEPT14                  Fusion   \n",
       "224         1516      EGFR        EGFRVIII           Gene Function   \n",
       "224         1516      EGFR        EGFRVIII           Gene Function   \n",
       "225            5       ALK       EML4::ALK                  Fusion   \n",
       "225            5       ALK       EML4::ALK                  Fusion   \n",
       "225            5       ALK       EML4::ALK                  Fusion   \n",
       "225            5       ALK       EML4::ALK                  Fusion   \n",
       "225            5       ALK       EML4::ALK                  Fusion   \n",
       "232         2750       ALK       EML6::ALK                  Fusion   \n",
       "390         4336    FLT3LG      Expression              Expression   \n",
       "390         4336    FLT3LG      Expression              Expression   \n",
       "429         2639      BRAF          Fusion                  Fusion   \n",
       "438           76      KRAS             G12     Protein Consequence   \n",
       "438           76      KRAS             G12     Protein Consequence   \n",
       "442           80      KRAS             G13     Protein Consequence   \n",
       "442           80      KRAS             G13     Protein Consequence   \n",
       "480           55      FLT3             ITD          Rearrangements   \n",
       "580         4310     KDM6A            Loss           Gene Function   \n",
       "651          208      NRAS        Mutation  Region Defined Variant   \n",
       "665          336      KRAS        Mutation  Region Defined Variant   \n",
       "665          336      KRAS        Mutation  Region Defined Variant   \n",
       "665          336      KRAS        Mutation  Region Defined Variant   \n",
       "665          336      KRAS        Mutation  Region Defined Variant   \n",
       "667          399      BRAF        Mutation  Region Defined Variant   \n",
       "667          399      BRAF        Mutation  Region Defined Variant   \n",
       "784          193      EGFR  Overexpression              Expression   \n",
       "784          193      EGFR  Overexpression              Expression   \n",
       "784          193      EGFR  Overexpression              Expression   \n",
       "806          603      FLT3  Overexpression              Expression   \n",
       "806          603      FLT3  Overexpression              Expression   \n",
       "895         2350     CRLF2    P2RY8::CRLF2                  Fusion   \n",
       "954          203      KRAS             Q61     Protein Consequence   \n",
       "954          203      KRAS             Q61     Protein Consequence   \n",
       "1046        4389       ALK        T1151dup             Copy Number   \n",
       "1046        4389       ALK        T1151dup             Copy Number   \n",
       "1046        4389       ALK        T1151dup             Copy Number   \n",
       "1046        4389       ALK        T1151dup             Copy Number   \n",
       "1046        4389       ALK        T1151dup             Copy Number   \n",
       "1079        4428       MYC   Translocation          Rearrangements   \n",
       "1080        4429      BCL2   Translocation          Rearrangements   \n",
       "1081        4430      BCL6   Translocation          Rearrangements   \n",
       "\n",
       "      variant_accepted evidence_ids evidence_status evidence_rating   \n",
       "0                False         7428     [submitted]             [3]  \\\n",
       "181               True         7786     [submitted]             [5]   \n",
       "189               True         7786     [submitted]             [5]   \n",
       "192              False         7786     [submitted]             [5]   \n",
       "193              False         7786     [submitted]             [5]   \n",
       "194              False         7786     [submitted]             [5]   \n",
       "218               True        11152     [submitted]             [2]   \n",
       "218               True        11154     [submitted]             [3]   \n",
       "224              False        11152     [submitted]             [2]   \n",
       "224              False        11154     [submitted]             [3]   \n",
       "225               True          444      [accepted]             [4]   \n",
       "225               True         1347      [accepted]             [3]   \n",
       "225               True         1348      [accepted]             [3]   \n",
       "225               True         1349      [accepted]             [2]   \n",
       "225               True         4608     [submitted]             [2]   \n",
       "232              False         7428     [submitted]             [3]   \n",
       "390              False        11088     [submitted]             [3]   \n",
       "390              False        11089     [submitted]             [3]   \n",
       "429              False         7149     [submitted]             [3]   \n",
       "438               True        11055     [submitted]             [4]   \n",
       "438               True        11056     [submitted]             [3]   \n",
       "442               True        11055     [submitted]             [4]   \n",
       "442               True        11056     [submitted]             [3]   \n",
       "480               True        11035     [submitted]             [3]   \n",
       "580              False        11035     [submitted]             [3]   \n",
       "651               True        11061     [submitted]             [4]   \n",
       "665               True        11058     [submitted]             [3]   \n",
       "665               True        11062     [submitted]             [3]   \n",
       "665               True        11059     [submitted]             [3]   \n",
       "665               True        11061     [submitted]             [4]   \n",
       "667               True         7149     [submitted]             [3]   \n",
       "667               True        11061     [submitted]             [4]   \n",
       "784               True        11058     [submitted]             [3]   \n",
       "784               True        11062     [submitted]             [3]   \n",
       "784               True        11059     [submitted]             [3]   \n",
       "806               True        11088     [submitted]             [3]   \n",
       "806               True        11089     [submitted]             [3]   \n",
       "895              False         7786     [submitted]             [5]   \n",
       "954               True        11055     [submitted]             [4]   \n",
       "954               True        11056     [submitted]             [3]   \n",
       "1046              True          444      [accepted]             [4]   \n",
       "1046              True         1347      [accepted]             [3]   \n",
       "1046              True         1348      [accepted]             [3]   \n",
       "1046              True         1349      [accepted]             [2]   \n",
       "1046              True         4608     [submitted]             [2]   \n",
       "1079             False         9813     [submitted]             [2]   \n",
       "1080             False         9813     [submitted]             [2]   \n",
       "1081             False         9813     [submitted]             [2]   \n",
       "\n",
       "     evidence_level  \n",
       "0               [C]  \n",
       "181             [B]  \n",
       "189             [B]  \n",
       "192             [B]  \n",
       "193             [B]  \n",
       "194             [B]  \n",
       "218             [B]  \n",
       "218             [B]  \n",
       "224             [B]  \n",
       "224             [B]  \n",
       "225             [C]  \n",
       "225             [D]  \n",
       "225             [D]  \n",
       "225             [D]  \n",
       "225             [D]  \n",
       "232             [C]  \n",
       "390             [D]  \n",
       "390             [D]  \n",
       "429             [C]  \n",
       "438             [B]  \n",
       "438             [B]  \n",
       "442             [B]  \n",
       "442             [B]  \n",
       "480             [D]  \n",
       "580             [D]  \n",
       "651             [B]  \n",
       "665             [B]  \n",
       "665             [B]  \n",
       "665             [B]  \n",
       "665             [B]  \n",
       "667             [C]  \n",
       "667             [B]  \n",
       "784             [B]  \n",
       "784             [B]  \n",
       "784             [B]  \n",
       "806             [D]  \n",
       "806             [D]  \n",
       "895             [B]  \n",
       "954             [B]  \n",
       "954             [B]  \n",
       "1046            [C]  \n",
       "1046            [D]  \n",
       "1046            [D]  \n",
       "1046            [D]  \n",
       "1046            [D]  \n",
       "1079            [C]  \n",
       "1080            [C]  \n",
       "1081            [C]  "
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "duplicate = not_supported_variants_df[not_supported_variants_df.duplicated('evidence_ids', keep=False)]\n",
    "duplicate"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### How many unique evidence ID's there are"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3738"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(not_supported_variants_df.evidence_ids))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### How many evidence ID's are paired with a not supported variant"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3766"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(not_supported_variants_df.evidence_ids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create individual tables seperating variants of different categories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "expression_non_supp_var_df = not_supported_variants_df[not_supported_variants_df.category == 'Expression']\n",
    "epigenetic_modification_non_supp_var_df = not_supported_variants_df[not_supported_variants_df.category == 'Epigenetic Modification']\n",
    "fusion_non_supp_var_df = not_supported_variants_df[not_supported_variants_df.category == 'Fusion']\n",
    "protein_consequence_non_supp_var_df = not_supported_variants_df[not_supported_variants_df.category == 'Protein Consequence']\n",
    "gene_function_non_supp_var_df = not_supported_variants_df[not_supported_variants_df.category == 'Gene Function']\n",
    "rearrangements_non_supp_var_df = not_supported_variants_df[not_supported_variants_df.category == 'Rearrangements']\n",
    "copy_number_non_supp_var_df = not_supported_variants_df[not_supported_variants_df.category == 'Copy Number']\n",
    "other_non_supp_var_df = not_supported_variants_df[not_supported_variants_df.category == 'Other']\n",
    "genotypes_easy_non_supp_var_df = not_supported_variants_df[not_supported_variants_df.category == 'Genotypes Easy']\n",
    "genotypes_compound_non_supp_var_df = not_supported_variants_df[not_supported_variants_df.category == 'Genotypes Compound']\n",
    "region_defined_variant_non_supp_var_df = not_supported_variants_df[not_supported_variants_df.category == 'Region Defined Variant']\n",
    "transcript_variant_non_supp_var_df = not_supported_variants_df[not_supported_variants_df.category == 'Transcript Variant']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>variant_name</th>\n",
       "      <th>category</th>\n",
       "      <th>variant_accepted</th>\n",
       "      <th>evidence_ids</th>\n",
       "      <th>evidence_status</th>\n",
       "      <th>evidence_rating</th>\n",
       "      <th>evidence_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>309</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>AUTOCRINE ACTIVATION</td>\n",
       "      <td>Expression</td>\n",
       "      <td>False</td>\n",
       "      <td>769</td>\n",
       "      <td>[rejected]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>309</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>AUTOCRINE ACTIVATION</td>\n",
       "      <td>Expression</td>\n",
       "      <td>False</td>\n",
       "      <td>768</td>\n",
       "      <td>[rejected]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>3385</td>\n",
       "      <td>CD44</td>\n",
       "      <td>CD44s</td>\n",
       "      <td>Expression</td>\n",
       "      <td>False</td>\n",
       "      <td>9426</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>341</td>\n",
       "      <td>CD44</td>\n",
       "      <td>CD44s Expression</td>\n",
       "      <td>Expression</td>\n",
       "      <td>True</td>\n",
       "      <td>825</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>3386</td>\n",
       "      <td>CD44</td>\n",
       "      <td>CD44v5 and CD44v6</td>\n",
       "      <td>Expression</td>\n",
       "      <td>False</td>\n",
       "      <td>9432</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1183</th>\n",
       "      <td>272</td>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>p16 Expression</td>\n",
       "      <td>Expression</td>\n",
       "      <td>True</td>\n",
       "      <td>753</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1183</th>\n",
       "      <td>272</td>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>p16 Expression</td>\n",
       "      <td>Expression</td>\n",
       "      <td>True</td>\n",
       "      <td>804</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1183</th>\n",
       "      <td>272</td>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>p16 Expression</td>\n",
       "      <td>Expression</td>\n",
       "      <td>True</td>\n",
       "      <td>1155</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1183</th>\n",
       "      <td>272</td>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>p16 Expression</td>\n",
       "      <td>Expression</td>\n",
       "      <td>True</td>\n",
       "      <td>1314</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1183</th>\n",
       "      <td>272</td>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>p16 Expression</td>\n",
       "      <td>Expression</td>\n",
       "      <td>True</td>\n",
       "      <td>4873</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>619 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      variant_id gene_name          variant_name    category   \n",
       "33           309      EGFR  AUTOCRINE ACTIVATION  Expression  \\\n",
       "33           309      EGFR  AUTOCRINE ACTIVATION  Expression   \n",
       "126         3385      CD44                 CD44s  Expression   \n",
       "127          341      CD44      CD44s Expression  Expression   \n",
       "128         3386      CD44     CD44v5 and CD44v6  Expression   \n",
       "...          ...       ...                   ...         ...   \n",
       "1183         272    CDKN2A        p16 Expression  Expression   \n",
       "1183         272    CDKN2A        p16 Expression  Expression   \n",
       "1183         272    CDKN2A        p16 Expression  Expression   \n",
       "1183         272    CDKN2A        p16 Expression  Expression   \n",
       "1183         272    CDKN2A        p16 Expression  Expression   \n",
       "\n",
       "      variant_accepted evidence_ids evidence_status evidence_rating   \n",
       "33               False          769      [rejected]             [2]  \\\n",
       "33               False          768      [rejected]             [3]   \n",
       "126              False         9426     [submitted]             [4]   \n",
       "127               True          825      [accepted]             [3]   \n",
       "128              False         9432     [submitted]             [4]   \n",
       "...                ...          ...             ...             ...   \n",
       "1183              True          753      [accepted]             [3]   \n",
       "1183              True          804      [accepted]             [3]   \n",
       "1183              True         1155      [accepted]             [2]   \n",
       "1183              True         1314      [accepted]             [2]   \n",
       "1183              True         4873     [submitted]             [4]   \n",
       "\n",
       "     evidence_level  \n",
       "33              [B]  \n",
       "33              [D]  \n",
       "126             [B]  \n",
       "127             [D]  \n",
       "128             [B]  \n",
       "...             ...  \n",
       "1183            [B]  \n",
       "1183            [B]  \n",
       "1183            [B]  \n",
       "1183            [B]  \n",
       "1183            [D]  \n",
       "\n",
       "[619 rows x 9 columns]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expression_non_supp_var_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"Number of unique variant ID's in Expression Variant category: 291\""
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_expression_variant_ids = len(set(expression_non_supp_var_df.variant_id))\n",
    "f\"Number of unique variant ID's in Expression Variant category: {unique_expression_variant_ids}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Percentage of all CIViC Variant Items in Expression category: 8.29296095753776%'"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expression_percentage_civic_variant_items = unique_expression_variant_ids/total_number_variants*100\n",
    "f\"Percentage of all CIViC Variant Items in Expression category: {expression_percentage_civic_variant_items}%\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"Number of unique evidence ID's in Expression category: 617\""
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_expression_evidence_ids = len(set(expression_non_supp_var_df.evidence_ids))\n",
    "f\"Number of unique evidence ID's in Expression category: {unique_expression_evidence_ids}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"Number of evidence ID's in Expression category: 619\""
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expression_evidence_ids = len(expression_non_supp_var_df)\n",
    "f\"Number of evidence ID's in Expression category: {expression_evidence_ids}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Percentage of all CIViC Evidence Items in Expression category:6.197267979108076%'"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expression_percentage_civic_evidence_items = unique_expression_evidence_ids/total_number_evidences*100\n",
    "f\"Percentage of all CIViC Evidence Items in Expression category:{expression_percentage_civic_evidence_items}%\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Percentage of all Not Supported Evidence Items in Expression category:16.506153023006956%'"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expression_percentage_not_supported_evidence_items = unique_expression_evidence_ids/total_number_not_supported_variant_unique_evidence_items*100\n",
    "f\"Percentage of all Not Supported Evidence Items in Expression category:{expression_percentage_not_supported_evidence_items}%\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Epigenetic Modification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>variant_name</th>\n",
       "      <th>category</th>\n",
       "      <th>variant_accepted</th>\n",
       "      <th>evidence_ids</th>\n",
       "      <th>evidence_status</th>\n",
       "      <th>evidence_rating</th>\n",
       "      <th>evidence_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>613</th>\n",
       "      <td>538</td>\n",
       "      <td>MLH1</td>\n",
       "      <td>METHYLATION</td>\n",
       "      <td>Epigenetic Modification</td>\n",
       "      <td>True</td>\n",
       "      <td>1315</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>614</th>\n",
       "      <td>784</td>\n",
       "      <td>CDKN2B</td>\n",
       "      <td>METHYLATION</td>\n",
       "      <td>Epigenetic Modification</td>\n",
       "      <td>True</td>\n",
       "      <td>1848</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>908</th>\n",
       "      <td>387</td>\n",
       "      <td>EIF4EBP1</td>\n",
       "      <td>PHOSPHORYLATION</td>\n",
       "      <td>Epigenetic Modification</td>\n",
       "      <td>True</td>\n",
       "      <td>918</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>908</th>\n",
       "      <td>387</td>\n",
       "      <td>EIF4EBP1</td>\n",
       "      <td>PHOSPHORYLATION</td>\n",
       "      <td>Epigenetic Modification</td>\n",
       "      <td>True</td>\n",
       "      <td>932</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>909</th>\n",
       "      <td>395</td>\n",
       "      <td>RPS6</td>\n",
       "      <td>PHOSPHORYLATION</td>\n",
       "      <td>Epigenetic Modification</td>\n",
       "      <td>True</td>\n",
       "      <td>930</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>910</th>\n",
       "      <td>632</td>\n",
       "      <td>RB1</td>\n",
       "      <td>PHOSPHORYLATION</td>\n",
       "      <td>Epigenetic Modification</td>\n",
       "      <td>True</td>\n",
       "      <td>1609</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>919</th>\n",
       "      <td>22</td>\n",
       "      <td>CCND2</td>\n",
       "      <td>PROMOTER DEMETHYLATION</td>\n",
       "      <td>Epigenetic Modification</td>\n",
       "      <td>True</td>\n",
       "      <td>219</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>920</th>\n",
       "      <td>2254</td>\n",
       "      <td>DBI</td>\n",
       "      <td>PROMOTER DEMETHYLATION</td>\n",
       "      <td>Epigenetic Modification</td>\n",
       "      <td>False</td>\n",
       "      <td>6033</td>\n",
       "      <td>[rejected]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>921</th>\n",
       "      <td>711</td>\n",
       "      <td>KLLN</td>\n",
       "      <td>PROMOTER METHYLATION</td>\n",
       "      <td>Epigenetic Modification</td>\n",
       "      <td>True</td>\n",
       "      <td>1761</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>921</th>\n",
       "      <td>711</td>\n",
       "      <td>KLLN</td>\n",
       "      <td>PROMOTER METHYLATION</td>\n",
       "      <td>Epigenetic Modification</td>\n",
       "      <td>True</td>\n",
       "      <td>1762</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>942</th>\n",
       "      <td>27</td>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>Promoter Hypermethylation</td>\n",
       "      <td>Epigenetic Modification</td>\n",
       "      <td>True</td>\n",
       "      <td>367</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>942</th>\n",
       "      <td>27</td>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>Promoter Hypermethylation</td>\n",
       "      <td>Epigenetic Modification</td>\n",
       "      <td>True</td>\n",
       "      <td>368</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>943</th>\n",
       "      <td>333</td>\n",
       "      <td>PTPRT</td>\n",
       "      <td>Promoter Hypermethylation</td>\n",
       "      <td>Epigenetic Modification</td>\n",
       "      <td>True</td>\n",
       "      <td>817</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>944</th>\n",
       "      <td>374</td>\n",
       "      <td>NAPRT</td>\n",
       "      <td>Promoter Hypermethylation</td>\n",
       "      <td>Epigenetic Modification</td>\n",
       "      <td>True</td>\n",
       "      <td>897</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>945</th>\n",
       "      <td>85</td>\n",
       "      <td>MGMT</td>\n",
       "      <td>Promoter Methylation</td>\n",
       "      <td>Epigenetic Modification</td>\n",
       "      <td>True</td>\n",
       "      <td>307</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>[A]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>945</th>\n",
       "      <td>85</td>\n",
       "      <td>MGMT</td>\n",
       "      <td>Promoter Methylation</td>\n",
       "      <td>Epigenetic Modification</td>\n",
       "      <td>True</td>\n",
       "      <td>7303</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[A]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>945</th>\n",
       "      <td>85</td>\n",
       "      <td>MGMT</td>\n",
       "      <td>Promoter Methylation</td>\n",
       "      <td>Epigenetic Modification</td>\n",
       "      <td>True</td>\n",
       "      <td>491</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>945</th>\n",
       "      <td>85</td>\n",
       "      <td>MGMT</td>\n",
       "      <td>Promoter Methylation</td>\n",
       "      <td>Epigenetic Modification</td>\n",
       "      <td>True</td>\n",
       "      <td>172</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>945</th>\n",
       "      <td>85</td>\n",
       "      <td>MGMT</td>\n",
       "      <td>Promoter Methylation</td>\n",
       "      <td>Epigenetic Modification</td>\n",
       "      <td>True</td>\n",
       "      <td>308</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>945</th>\n",
       "      <td>85</td>\n",
       "      <td>MGMT</td>\n",
       "      <td>Promoter Methylation</td>\n",
       "      <td>Epigenetic Modification</td>\n",
       "      <td>True</td>\n",
       "      <td>490</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>945</th>\n",
       "      <td>85</td>\n",
       "      <td>MGMT</td>\n",
       "      <td>Promoter Methylation</td>\n",
       "      <td>Epigenetic Modification</td>\n",
       "      <td>True</td>\n",
       "      <td>309</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[E]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1013</th>\n",
       "      <td>3257</td>\n",
       "      <td>AKT1</td>\n",
       "      <td>S473 Phosphorylation</td>\n",
       "      <td>Epigenetic Modification</td>\n",
       "      <td>True</td>\n",
       "      <td>4051</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1050</th>\n",
       "      <td>385</td>\n",
       "      <td>PRKAA2</td>\n",
       "      <td>T172 PHOSPHORYLATION</td>\n",
       "      <td>Epigenetic Modification</td>\n",
       "      <td>True</td>\n",
       "      <td>913</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[E]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1149</th>\n",
       "      <td>390</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>Y1092 PHOSPHORYLATION</td>\n",
       "      <td>Epigenetic Modification</td>\n",
       "      <td>True</td>\n",
       "      <td>923</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      variant_id gene_name               variant_name   \n",
       "613          538      MLH1                METHYLATION  \\\n",
       "614          784    CDKN2B                METHYLATION   \n",
       "908          387  EIF4EBP1            PHOSPHORYLATION   \n",
       "908          387  EIF4EBP1            PHOSPHORYLATION   \n",
       "909          395      RPS6            PHOSPHORYLATION   \n",
       "910          632       RB1            PHOSPHORYLATION   \n",
       "919           22     CCND2     PROMOTER DEMETHYLATION   \n",
       "920         2254       DBI     PROMOTER DEMETHYLATION   \n",
       "921          711      KLLN       PROMOTER METHYLATION   \n",
       "921          711      KLLN       PROMOTER METHYLATION   \n",
       "942           27    CDKN2A  Promoter Hypermethylation   \n",
       "942           27    CDKN2A  Promoter Hypermethylation   \n",
       "943          333     PTPRT  Promoter Hypermethylation   \n",
       "944          374     NAPRT  Promoter Hypermethylation   \n",
       "945           85      MGMT       Promoter Methylation   \n",
       "945           85      MGMT       Promoter Methylation   \n",
       "945           85      MGMT       Promoter Methylation   \n",
       "945           85      MGMT       Promoter Methylation   \n",
       "945           85      MGMT       Promoter Methylation   \n",
       "945           85      MGMT       Promoter Methylation   \n",
       "945           85      MGMT       Promoter Methylation   \n",
       "1013        3257      AKT1       S473 Phosphorylation   \n",
       "1050         385    PRKAA2       T172 PHOSPHORYLATION   \n",
       "1149         390      EGFR      Y1092 PHOSPHORYLATION   \n",
       "\n",
       "                     category  variant_accepted evidence_ids evidence_status   \n",
       "613   Epigenetic Modification              True         1315      [accepted]  \\\n",
       "614   Epigenetic Modification              True         1848      [accepted]   \n",
       "908   Epigenetic Modification              True          918      [accepted]   \n",
       "908   Epigenetic Modification              True          932      [accepted]   \n",
       "909   Epigenetic Modification              True          930      [accepted]   \n",
       "910   Epigenetic Modification              True         1609      [accepted]   \n",
       "919   Epigenetic Modification              True          219      [accepted]   \n",
       "920   Epigenetic Modification             False         6033      [rejected]   \n",
       "921   Epigenetic Modification              True         1761      [accepted]   \n",
       "921   Epigenetic Modification              True         1762      [accepted]   \n",
       "942   Epigenetic Modification              True          367      [accepted]   \n",
       "942   Epigenetic Modification              True          368      [accepted]   \n",
       "943   Epigenetic Modification              True          817      [accepted]   \n",
       "944   Epigenetic Modification              True          897      [accepted]   \n",
       "945   Epigenetic Modification              True          307      [accepted]   \n",
       "945   Epigenetic Modification              True         7303     [submitted]   \n",
       "945   Epigenetic Modification              True          491      [accepted]   \n",
       "945   Epigenetic Modification              True          172      [accepted]   \n",
       "945   Epigenetic Modification              True          308      [accepted]   \n",
       "945   Epigenetic Modification              True          490      [accepted]   \n",
       "945   Epigenetic Modification              True          309      [accepted]   \n",
       "1013  Epigenetic Modification              True         4051      [accepted]   \n",
       "1050  Epigenetic Modification              True          913      [accepted]   \n",
       "1149  Epigenetic Modification              True          923      [accepted]   \n",
       "\n",
       "     evidence_rating evidence_level  \n",
       "613              [2]            [B]  \n",
       "614              [4]            [B]  \n",
       "908              [2]            [D]  \n",
       "908              [2]            [D]  \n",
       "909              [2]            [D]  \n",
       "910              [3]            [B]  \n",
       "919              [3]            [D]  \n",
       "920              [5]            [B]  \n",
       "921              [3]            [B]  \n",
       "921              [3]            [B]  \n",
       "942              [3]            [B]  \n",
       "942              [3]            [B]  \n",
       "943              [2]            [D]  \n",
       "944              [3]            [D]  \n",
       "945              [5]            [A]  \n",
       "945              [3]            [A]  \n",
       "945              [5]            [B]  \n",
       "945              [4]            [B]  \n",
       "945              [4]            [B]  \n",
       "945              [4]            [B]  \n",
       "945              [4]            [E]  \n",
       "1013             [3]            [B]  \n",
       "1050             [2]            [E]  \n",
       "1149             [3]            [B]  "
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "epigenetic_modification_non_supp_var_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"Number of unique variant ID's in Epigenetic Modification Variant category: 15\""
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_epigenetic_modification_variant_ids = len(set(epigenetic_modification_non_supp_var_df.variant_id))\n",
    "f\"Number of unique variant ID's in Epigenetic Modification Variant category: {unique_epigenetic_modification_variant_ids}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Percentage of all CIViC Variant Items in Epigenetic Modification category: 0.42747221430607013%'"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "epigenetic_modification_percentage_civic_variant_items = unique_epigenetic_modification_variant_ids/total_number_variants*100\n",
    "f\"Percentage of all CIViC Variant Items in Epigenetic Modification category: {epigenetic_modification_percentage_civic_variant_items}%\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"Number of unique evidence ID's in Epigenetic Modification category: 24\""
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_epigenetic_evidence_ids = len(set(epigenetic_modification_non_supp_var_df.evidence_ids))\n",
    "f\"Number of unique evidence ID's in Epigenetic Modification category: {unique_epigenetic_evidence_ids}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"Number of evidence ID's in Epigenetic Modification category: 24\""
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "epigenetic_evidence_ids = len(epigenetic_modification_non_supp_var_df)\n",
    "f\"Number of evidence ID's in Epigenetic Modification category: {epigenetic_evidence_ids}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Percentage of all CIViC Evidence Items in Epigenetic Modification category:0.24106066693451184%'"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "epigenetic_percentage_civic_evidence_items = unique_epigenetic_evidence_ids/total_number_evidences*100\n",
    "f\"Percentage of all CIViC Evidence Items in Epigenetic Modification category:{epigenetic_percentage_civic_evidence_items}%\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Percentage of all Not Supported Evidence Items in Epigenetic Modification category:0.6420545746388443%'"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "epigenetic_percentage_not_supported_evidence_items = unique_epigenetic_evidence_ids/total_number_not_supported_variant_unique_evidence_items*100\n",
    "f\"Percentage of all Not Supported Evidence Items in Epigenetic Modification category:{epigenetic_percentage_not_supported_evidence_items}%\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Fusion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>variant_name</th>\n",
       "      <th>category</th>\n",
       "      <th>variant_accepted</th>\n",
       "      <th>evidence_ids</th>\n",
       "      <th>evidence_status</th>\n",
       "      <th>evidence_rating</th>\n",
       "      <th>evidence_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4417</td>\n",
       "      <td>ALK</td>\n",
       "      <td>FBXO11::ALK</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>False</td>\n",
       "      <td>7428</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>780</td>\n",
       "      <td>GLI1</td>\n",
       "      <td>ACTB::GLI1</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>True</td>\n",
       "      <td>1844</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>3702</td>\n",
       "      <td>NTRK2</td>\n",
       "      <td>AFAP1::NTRK2</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>True</td>\n",
       "      <td>10098</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>2577</td>\n",
       "      <td>PDGFRB</td>\n",
       "      <td>AGGF1::PDGFRB C843G</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>False</td>\n",
       "      <td>6999</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>285</td>\n",
       "      <td>BRAF</td>\n",
       "      <td>AGK::BRAF</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>True</td>\n",
       "      <td>723</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1161</th>\n",
       "      <td>466</td>\n",
       "      <td>FGFR1</td>\n",
       "      <td>ZNF198::FGFR1</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>True</td>\n",
       "      <td>1104</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1162</th>\n",
       "      <td>2864</td>\n",
       "      <td>JAK2</td>\n",
       "      <td>ZNF274::JAK2</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>True</td>\n",
       "      <td>7706</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1163</th>\n",
       "      <td>2977</td>\n",
       "      <td>JAK2</td>\n",
       "      <td>ZNF430::JAK2</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>True</td>\n",
       "      <td>7999</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1172</th>\n",
       "      <td>4406</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>fusion</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>False</td>\n",
       "      <td>11143</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1172</th>\n",
       "      <td>4406</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>fusion</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>False</td>\n",
       "      <td>11169</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1187 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      variant_id gene_name         variant_name category  variant_accepted   \n",
       "0           4417       ALK          FBXO11::ALK   Fusion             False  \\\n",
       "18           780      GLI1           ACTB::GLI1   Fusion              True   \n",
       "19          3702     NTRK2         AFAP1::NTRK2   Fusion              True   \n",
       "20          2577    PDGFRB  AGGF1::PDGFRB C843G   Fusion             False   \n",
       "21           285      BRAF            AGK::BRAF   Fusion              True   \n",
       "...          ...       ...                  ...      ...               ...   \n",
       "1161         466     FGFR1        ZNF198::FGFR1   Fusion              True   \n",
       "1162        2864      JAK2         ZNF274::JAK2   Fusion              True   \n",
       "1163        2977      JAK2         ZNF430::JAK2   Fusion              True   \n",
       "1172        4406      EGFR               fusion   Fusion             False   \n",
       "1172        4406      EGFR               fusion   Fusion             False   \n",
       "\n",
       "     evidence_ids evidence_status evidence_rating evidence_level  \n",
       "0            7428     [submitted]             [3]            [C]  \n",
       "18           1844      [accepted]             [3]            [B]  \n",
       "19          10098      [accepted]             [3]            [D]  \n",
       "20           6999     [submitted]             [3]            [D]  \n",
       "21            723      [accepted]             [2]            [C]  \n",
       "...           ...             ...             ...            ...  \n",
       "1161         1104      [accepted]             [4]            [C]  \n",
       "1162         7706      [accepted]             [3]            [C]  \n",
       "1163         7999      [accepted]             [3]            [C]  \n",
       "1172        11143     [submitted]             [1]            [C]  \n",
       "1172        11169     [submitted]             [1]            [C]  \n",
       "\n",
       "[1187 rows x 9 columns]"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fusion_non_supp_var_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"Number of unique variant ID's in Fusion Variant category: 301\""
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_fusion_variant_ids = len(set(fusion_non_supp_var_df.variant_id))\n",
    "f\"Number of unique variant ID's in Fusion Variant category: {unique_fusion_variant_ids}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Percentage of all CIViC Variant Items in Fusion category: 8.577942433741807%'"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fusion_percentage_civic_variant_items = unique_fusion_variant_ids/total_number_variants*100\n",
    "f\"Percentage of all CIViC Variant Items in Fusion category: {fusion_percentage_civic_variant_items}%\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"Number of unique evidence ID's in Fusion category: 1186\""
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_fusion_evidence_ids = len(set(fusion_non_supp_var_df.evidence_ids))\n",
    "f\"Number of unique evidence ID's in Fusion category: {unique_fusion_evidence_ids}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"Number of evidence ID's in Fusion category: 1187\""
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fusion_evidence_ids = len(fusion_non_supp_var_df.evidence_ids)\n",
    "f\"Number of evidence ID's in Fusion category: {fusion_evidence_ids}\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Percentage of all CIViC Evidence Items in Fusion category:11.912414624347127%'"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fusion_percentage_civic_evidence_items = unique_fusion_evidence_ids/total_number_evidences*100\n",
    "f\"Percentage of all CIViC Evidence Items in Fusion category:{fusion_percentage_civic_evidence_items}%\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Percentage of all Not Supported Evidence Items in Fusion category:31.728196896736222%'"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fusion_percentage_not_supported_evidence_items = unique_fusion_evidence_ids/total_number_not_supported_variant_unique_evidence_items*100\n",
    "f\"Percentage of all Not Supported Evidence Items in Fusion category:{fusion_percentage_not_supported_evidence_items}%\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Protein Consequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>variant_name</th>\n",
       "      <th>category</th>\n",
       "      <th>variant_accepted</th>\n",
       "      <th>evidence_ids</th>\n",
       "      <th>evidence_status</th>\n",
       "      <th>evidence_rating</th>\n",
       "      <th>evidence_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>785</td>\n",
       "      <td>CHEK2</td>\n",
       "      <td>1100DELC</td>\n",
       "      <td>Protein Consequence</td>\n",
       "      <td>True</td>\n",
       "      <td>7235</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>785</td>\n",
       "      <td>CHEK2</td>\n",
       "      <td>1100DELC</td>\n",
       "      <td>Protein Consequence</td>\n",
       "      <td>True</td>\n",
       "      <td>1849</td>\n",
       "      <td>[rejected]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>785</td>\n",
       "      <td>CHEK2</td>\n",
       "      <td>1100DELC</td>\n",
       "      <td>Protein Consequence</td>\n",
       "      <td>True</td>\n",
       "      <td>1850</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>2727</td>\n",
       "      <td>ERBB2</td>\n",
       "      <td>A129</td>\n",
       "      <td>Protein Consequence</td>\n",
       "      <td>False</td>\n",
       "      <td>7346</td>\n",
       "      <td>[rejected]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>2729</td>\n",
       "      <td>TP53</td>\n",
       "      <td>A129</td>\n",
       "      <td>Protein Consequence</td>\n",
       "      <td>False</td>\n",
       "      <td>7348</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1154</th>\n",
       "      <td>3055</td>\n",
       "      <td>FLT3</td>\n",
       "      <td>Y693</td>\n",
       "      <td>Protein Consequence</td>\n",
       "      <td>False</td>\n",
       "      <td>8285</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1155</th>\n",
       "      <td>1672</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>Y69FS*11</td>\n",
       "      <td>Protein Consequence</td>\n",
       "      <td>False</td>\n",
       "      <td>4808</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[None]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1157</th>\n",
       "      <td>3410</td>\n",
       "      <td>TEK</td>\n",
       "      <td>Y897C and R915C</td>\n",
       "      <td>Protein Consequence</td>\n",
       "      <td>False</td>\n",
       "      <td>9476</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1182</th>\n",
       "      <td>3057</td>\n",
       "      <td>GATA1</td>\n",
       "      <td>p.Ser51Alafs*86</td>\n",
       "      <td>Protein Consequence</td>\n",
       "      <td>False</td>\n",
       "      <td>8302</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1202</th>\n",
       "      <td>3508</td>\n",
       "      <td>CD274</td>\n",
       "      <td>v242</td>\n",
       "      <td>Protein Consequence</td>\n",
       "      <td>False</td>\n",
       "      <td>9695</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[E]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>299 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      variant_id gene_name     variant_name             category   \n",
       "2            785     CHEK2         1100DELC  Protein Consequence  \\\n",
       "2            785     CHEK2         1100DELC  Protein Consequence   \n",
       "2            785     CHEK2         1100DELC  Protein Consequence   \n",
       "12          2727     ERBB2             A129  Protein Consequence   \n",
       "13          2729      TP53             A129  Protein Consequence   \n",
       "...          ...       ...              ...                  ...   \n",
       "1154        3055      FLT3             Y693  Protein Consequence   \n",
       "1155        1672      EGFR         Y69FS*11  Protein Consequence   \n",
       "1157        3410       TEK  Y897C and R915C  Protein Consequence   \n",
       "1182        3057     GATA1  p.Ser51Alafs*86  Protein Consequence   \n",
       "1202        3508     CD274             v242  Protein Consequence   \n",
       "\n",
       "      variant_accepted evidence_ids evidence_status evidence_rating   \n",
       "2                 True         7235     [submitted]             [4]  \\\n",
       "2                 True         1849      [rejected]             [3]   \n",
       "2                 True         1850      [accepted]             [3]   \n",
       "12               False         7346      [rejected]             [4]   \n",
       "13               False         7348     [submitted]             [4]   \n",
       "...                ...          ...             ...             ...   \n",
       "1154             False         8285     [submitted]             [3]   \n",
       "1155             False         4808     [submitted]          [None]   \n",
       "1157             False         9476     [submitted]             [2]   \n",
       "1182             False         8302     [submitted]             [2]   \n",
       "1202             False         9695     [submitted]             [4]   \n",
       "\n",
       "     evidence_level  \n",
       "2               [B]  \n",
       "2               [B]  \n",
       "2               [B]  \n",
       "12              [D]  \n",
       "13              [D]  \n",
       "...             ...  \n",
       "1154            [D]  \n",
       "1155            [D]  \n",
       "1157            [C]  \n",
       "1182            [C]  \n",
       "1202            [E]  \n",
       "\n",
       "[299 rows x 9 columns]"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein_consequence_non_supp_var_df "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"Number of unique variant ID's in Protein Consequence Variant category: 128\""
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_protein_consequence_variant_ids = len(set(protein_consequence_non_supp_var_df.variant_id))\n",
    "f\"Number of unique variant ID's in Protein Consequence Variant category: {unique_protein_consequence_variant_ids}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Percentage of all CIViC Variant Items in Protein Consequence category: 3.647762895411798%'"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein_consequence_percentage_civic_variant_items = unique_protein_consequence_variant_ids/total_number_variants*100\n",
    "f\"Percentage of all CIViC Variant Items in Protein Consequence category: {protein_consequence_percentage_civic_variant_items}%\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"Number of unique evidence ID's in Protein Consequence category: 295\""
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_protein_evidence_ids = len(set(protein_consequence_non_supp_var_df.evidence_ids))\n",
    "f\"Number of unique evidence ID's in Protein Consequence category: {unique_protein_evidence_ids}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"Number of evidence ID's in Protein Consequence category: 299\""
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein_evidece_ids = len(protein_consequence_non_supp_var_df.evidence_ids)\n",
    "f\"Number of evidence ID's in Protein Consequence category: {protein_evidece_ids}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Percentage of all CIViC Evidence Items in Protein Evidence category:2.963037364403375%'"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein_evidece_percentage_civic_evidence_items = unique_protein_evidence_ids/total_number_evidences*100\n",
    "f\"Percentage of all CIViC Evidence Items in Protein Evidence category:{protein_evidece_percentage_civic_evidence_items}%\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Percentage of all Not Supported Evidence Items in Protein Evidence category:7.891920813269128%'"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein_evidece_percentage_not_supported_evidence_items = unique_protein_evidence_ids/total_number_not_supported_variant_unique_evidence_items*100\n",
    "f\"Percentage of all Not Supported Evidence Items in Protein Evidence category:{protein_evidece_percentage_not_supported_evidence_items}%\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Gene Function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>variant_name</th>\n",
       "      <th>category</th>\n",
       "      <th>variant_accepted</th>\n",
       "      <th>evidence_ids</th>\n",
       "      <th>evidence_status</th>\n",
       "      <th>evidence_rating</th>\n",
       "      <th>evidence_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>1296</td>\n",
       "      <td>CTNNB1</td>\n",
       "      <td>Activating Mutation</td>\n",
       "      <td>Gene Function</td>\n",
       "      <td>True</td>\n",
       "      <td>2988</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>2649</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>Activating Mutation</td>\n",
       "      <td>Gene Function</td>\n",
       "      <td>False</td>\n",
       "      <td>7180</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>2649</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>Activating Mutation</td>\n",
       "      <td>Gene Function</td>\n",
       "      <td>False</td>\n",
       "      <td>7182</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>2649</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>Activating Mutation</td>\n",
       "      <td>Gene Function</td>\n",
       "      <td>False</td>\n",
       "      <td>7184</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>2657</td>\n",
       "      <td>ERBB2</td>\n",
       "      <td>Activating Mutation</td>\n",
       "      <td>Gene Function</td>\n",
       "      <td>False</td>\n",
       "      <td>7201</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1133</th>\n",
       "      <td>312</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>VIII</td>\n",
       "      <td>Gene Function</td>\n",
       "      <td>True</td>\n",
       "      <td>773</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1133</th>\n",
       "      <td>312</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>VIII</td>\n",
       "      <td>Gene Function</td>\n",
       "      <td>True</td>\n",
       "      <td>8191</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1133</th>\n",
       "      <td>312</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>VIII</td>\n",
       "      <td>Gene Function</td>\n",
       "      <td>True</td>\n",
       "      <td>4500</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1133</th>\n",
       "      <td>312</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>VIII</td>\n",
       "      <td>Gene Function</td>\n",
       "      <td>True</td>\n",
       "      <td>8192</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1133</th>\n",
       "      <td>312</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>VIII</td>\n",
       "      <td>Gene Function</td>\n",
       "      <td>True</td>\n",
       "      <td>1017</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>347 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      variant_id gene_name         variant_name       category   \n",
       "34          1296    CTNNB1  Activating Mutation  Gene Function  \\\n",
       "35          2649      KRAS  Activating Mutation  Gene Function   \n",
       "35          2649      KRAS  Activating Mutation  Gene Function   \n",
       "35          2649      KRAS  Activating Mutation  Gene Function   \n",
       "36          2657     ERBB2  Activating Mutation  Gene Function   \n",
       "...          ...       ...                  ...            ...   \n",
       "1133         312      EGFR                 VIII  Gene Function   \n",
       "1133         312      EGFR                 VIII  Gene Function   \n",
       "1133         312      EGFR                 VIII  Gene Function   \n",
       "1133         312      EGFR                 VIII  Gene Function   \n",
       "1133         312      EGFR                 VIII  Gene Function   \n",
       "\n",
       "      variant_accepted evidence_ids evidence_status evidence_rating   \n",
       "34                True         2988      [accepted]             [4]  \\\n",
       "35               False         7180     [submitted]             [3]   \n",
       "35               False         7182     [submitted]             [2]   \n",
       "35               False         7184     [submitted]             [1]   \n",
       "36               False         7201     [submitted]             [3]   \n",
       "...                ...          ...             ...             ...   \n",
       "1133              True          773      [accepted]             [2]   \n",
       "1133              True         8191     [submitted]             [4]   \n",
       "1133              True         4500      [accepted]             [3]   \n",
       "1133              True         8192     [submitted]             [3]   \n",
       "1133              True         1017      [accepted]             [1]   \n",
       "\n",
       "     evidence_level  \n",
       "34              [D]  \n",
       "35              [B]  \n",
       "35              [D]  \n",
       "35              [D]  \n",
       "36              [B]  \n",
       "...             ...  \n",
       "1133            [C]  \n",
       "1133            [D]  \n",
       "1133            [D]  \n",
       "1133            [D]  \n",
       "1133            [D]  \n",
       "\n",
       "[347 rows x 9 columns]"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_function_non_supp_var_df "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"Number of unique variant ID's in Gene Function Variant category: 87\""
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_gene_function_variant_ids = len(set(gene_function_non_supp_var_df.variant_id))\n",
    "f\"Number of unique variant ID's in Gene Function Variant category: {unique_gene_function_variant_ids}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Percentage of all CIViC Variant Items in Gene Function category: 2.479338842975207%'"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_function_percentage_civic_variant_items = unique_gene_function_variant_ids/total_number_variants*100\n",
    "f\"Percentage of all CIViC Variant Items in Gene Function category: {gene_function_percentage_civic_variant_items}%\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"Number of unique evidence ID's in Gene Function category: 343\""
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_gene_function_evidence_ids = len(set(gene_function_non_supp_var_df.evidence_ids))\n",
    "f\"Number of unique evidence ID's in Gene Function category: {unique_gene_function_evidence_ids}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"Number of evidence ID's in Gene Function category: 347\""
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_function_evidence_ids = len(gene_function_non_supp_var_df.evidence_ids)\n",
    "f\"Number of evidence ID's in Gene Function category: {gene_function_evidence_ids}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Percentage of all CIViC Evidence Items in Gene Function category:3.4451586982723983%'"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_function_percentage_civic_evidence_items = unique_gene_function_evidence_ids/total_number_evidences*100\n",
    "f\"Percentage of all CIViC Evidence Items in Gene Function category:{gene_function_percentage_civic_evidence_items}%\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Percentage of all Not Supported Evidence Items in Gene Function category:9.176029962546817%'"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_function_percentage_not_supported_evidence_items = unique_gene_function_evidence_ids/total_number_not_supported_variant_unique_evidence_items*100\n",
    "f\"Percentage of all Not Supported Evidence Items in Gene Function category:{gene_function_percentage_not_supported_evidence_items}%\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Rearrangements"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>variant_name</th>\n",
       "      <th>category</th>\n",
       "      <th>variant_accepted</th>\n",
       "      <th>evidence_ids</th>\n",
       "      <th>evidence_status</th>\n",
       "      <th>evidence_rating</th>\n",
       "      <th>evidence_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>2390</td>\n",
       "      <td>MAP2K1</td>\n",
       "      <td>56_61QKQKVG&gt;R</td>\n",
       "      <td>Rearrangements</td>\n",
       "      <td>False</td>\n",
       "      <td>6384</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>709</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>Alu insertion</td>\n",
       "      <td>Rearrangements</td>\n",
       "      <td>True</td>\n",
       "      <td>1759</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>165</th>\n",
       "      <td>943</td>\n",
       "      <td>PDGFRA</td>\n",
       "      <td>D842_H845DELDIMH</td>\n",
       "      <td>Rearrangements</td>\n",
       "      <td>True</td>\n",
       "      <td>2460</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196</th>\n",
       "      <td>3009</td>\n",
       "      <td>ABL1</td>\n",
       "      <td>Double Ph</td>\n",
       "      <td>Rearrangements</td>\n",
       "      <td>True</td>\n",
       "      <td>8093</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>207</th>\n",
       "      <td>1577</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>E709_T710&gt;D</td>\n",
       "      <td>Rearrangements</td>\n",
       "      <td>False</td>\n",
       "      <td>4663</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[None]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1197</th>\n",
       "      <td>3524</td>\n",
       "      <td>MECOM</td>\n",
       "      <td>t(3;21)(q26.2;q22)</td>\n",
       "      <td>Rearrangements</td>\n",
       "      <td>False</td>\n",
       "      <td>9719</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1198</th>\n",
       "      <td>3264</td>\n",
       "      <td>VHL</td>\n",
       "      <td>t(3;6)(p12.3;q24.3)</td>\n",
       "      <td>Rearrangements</td>\n",
       "      <td>False</td>\n",
       "      <td>9063</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1199</th>\n",
       "      <td>4008</td>\n",
       "      <td>VHL</td>\n",
       "      <td>t(3;6)(p13;q25.1)</td>\n",
       "      <td>Rearrangements</td>\n",
       "      <td>False</td>\n",
       "      <td>10357</td>\n",
       "      <td>[rejected]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1199</th>\n",
       "      <td>4008</td>\n",
       "      <td>VHL</td>\n",
       "      <td>t(3;6)(p13;q25.1)</td>\n",
       "      <td>Rearrangements</td>\n",
       "      <td>False</td>\n",
       "      <td>10350</td>\n",
       "      <td>[rejected]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1200</th>\n",
       "      <td>3351</td>\n",
       "      <td>VHL</td>\n",
       "      <td>t(3;8)(p13;q24.1)</td>\n",
       "      <td>Rearrangements</td>\n",
       "      <td>False</td>\n",
       "      <td>9340</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>138 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      variant_id gene_name         variant_name        category   \n",
       "10          2390    MAP2K1        56_61QKQKVG>R  Rearrangements  \\\n",
       "40           709     BRCA1        Alu insertion  Rearrangements   \n",
       "165          943    PDGFRA     D842_H845DELDIMH  Rearrangements   \n",
       "196         3009      ABL1            Double Ph  Rearrangements   \n",
       "207         1577      EGFR          E709_T710>D  Rearrangements   \n",
       "...          ...       ...                  ...             ...   \n",
       "1197        3524     MECOM   t(3;21)(q26.2;q22)  Rearrangements   \n",
       "1198        3264       VHL  t(3;6)(p12.3;q24.3)  Rearrangements   \n",
       "1199        4008       VHL    t(3;6)(p13;q25.1)  Rearrangements   \n",
       "1199        4008       VHL    t(3;6)(p13;q25.1)  Rearrangements   \n",
       "1200        3351       VHL    t(3;8)(p13;q24.1)  Rearrangements   \n",
       "\n",
       "      variant_accepted evidence_ids evidence_status evidence_rating   \n",
       "10               False         6384     [submitted]             [3]  \\\n",
       "40                True         1759      [accepted]             [3]   \n",
       "165               True         2460      [accepted]             [2]   \n",
       "196               True         8093      [accepted]             [4]   \n",
       "207              False         4663     [submitted]          [None]   \n",
       "...                ...          ...             ...             ...   \n",
       "1197             False         9719     [submitted]             [2]   \n",
       "1198             False         9063     [submitted]             [2]   \n",
       "1199             False        10357      [rejected]             [3]   \n",
       "1199             False        10350      [rejected]             [2]   \n",
       "1200             False         9340     [submitted]             [3]   \n",
       "\n",
       "     evidence_level  \n",
       "10              [D]  \n",
       "40              [C]  \n",
       "165             [D]  \n",
       "196             [C]  \n",
       "207             [C]  \n",
       "...             ...  \n",
       "1197            [C]  \n",
       "1198            [C]  \n",
       "1199            [C]  \n",
       "1199            [C]  \n",
       "1200            [C]  \n",
       "\n",
       "[138 rows x 9 columns]"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rearrangements_non_supp_var_df "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"Number of unique variant ID's in Rearrangements Variant category: 50\""
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_rearrangements_variant_ids = len(set(rearrangements_non_supp_var_df.variant_id))\n",
    "f\"Number of unique variant ID's in Rearrangements Variant category: {unique_rearrangements_variant_ids}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Percentage of all CIViC Variant Items in Rearrangements category: 1.4249073810202337%'"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rearrangements_percentage_civic_variant_items = unique_rearrangements_variant_ids/total_number_variants*100\n",
    "f\"Percentage of all CIViC Variant Items in Rearrangements category: {rearrangements_percentage_civic_variant_items}%\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"Number of unique evidence ID's in Rearrangements category: 136\""
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_rearrangements_evidence_ids = len(set(rearrangements_non_supp_var_df.evidence_ids))\n",
    "f\"Number of unique evidence ID's in Rearrangements category: {unique_rearrangements_evidence_ids}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"Number of evidence ID's in Rearrangements category: 138\""
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rearrangements_evidence_ids = len(rearrangements_non_supp_var_df.evidence_ids)\n",
    "f\"Number of evidence ID's in Rearrangements category: {rearrangements_evidence_ids}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Percentage of all CIViC Evidence Items in Rearrangements category:1.3660104459622338%'"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rearrangements_percentage_civic_evidence_items = unique_rearrangements_evidence_ids/total_number_evidences*100\n",
    "f\"Percentage of all CIViC Evidence Items in Rearrangements category:{rearrangements_percentage_civic_evidence_items}%\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Percentage of all Not Supported Evidence Items in Rearrangements category:3.6383092562867843%'"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rearrangements_percentage_not_supported_evidence_items = unique_rearrangements_evidence_ids/total_number_not_supported_variant_unique_evidence_items*100\n",
    "f\"Percentage of all Not Supported Evidence Items in Rearrangements category:{rearrangements_percentage_not_supported_evidence_items}%\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Copy Number"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>variant_name</th>\n",
       "      <th>category</th>\n",
       "      <th>variant_accepted</th>\n",
       "      <th>evidence_ids</th>\n",
       "      <th>evidence_status</th>\n",
       "      <th>evidence_rating</th>\n",
       "      <th>evidence_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>3195</td>\n",
       "      <td>VHL</td>\n",
       "      <td>A149fs (c.444dup)</td>\n",
       "      <td>Copy Number</td>\n",
       "      <td>False</td>\n",
       "      <td>9982</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>3195</td>\n",
       "      <td>VHL</td>\n",
       "      <td>A149fs (c.444dup)</td>\n",
       "      <td>Copy Number</td>\n",
       "      <td>False</td>\n",
       "      <td>8822</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>3195</td>\n",
       "      <td>VHL</td>\n",
       "      <td>A149fs (c.444dup)</td>\n",
       "      <td>Copy Number</td>\n",
       "      <td>False</td>\n",
       "      <td>6156</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>2144</td>\n",
       "      <td>VHL</td>\n",
       "      <td>A56fs (c.166dup)</td>\n",
       "      <td>Copy Number</td>\n",
       "      <td>True</td>\n",
       "      <td>5742</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>1579</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>A767_V769dupASV</td>\n",
       "      <td>Copy Number</td>\n",
       "      <td>True</td>\n",
       "      <td>4665</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[None]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1156</th>\n",
       "      <td>414</td>\n",
       "      <td>ERBB2</td>\n",
       "      <td>Y772_A775DUP</td>\n",
       "      <td>Copy Number</td>\n",
       "      <td>True</td>\n",
       "      <td>4428</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[None]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1156</th>\n",
       "      <td>414</td>\n",
       "      <td>ERBB2</td>\n",
       "      <td>Y772_A775DUP</td>\n",
       "      <td>Copy Number</td>\n",
       "      <td>True</td>\n",
       "      <td>1046</td>\n",
       "      <td>[rejected]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1156</th>\n",
       "      <td>414</td>\n",
       "      <td>ERBB2</td>\n",
       "      <td>Y772_A775DUP</td>\n",
       "      <td>Copy Number</td>\n",
       "      <td>True</td>\n",
       "      <td>1047</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1156</th>\n",
       "      <td>414</td>\n",
       "      <td>ERBB2</td>\n",
       "      <td>Y772_A775DUP</td>\n",
       "      <td>Copy Number</td>\n",
       "      <td>True</td>\n",
       "      <td>960</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1181</th>\n",
       "      <td>2411</td>\n",
       "      <td>SDHB</td>\n",
       "      <td>p.193_196dupSTSC (c.577_588dupAGCACCAGCTGC)</td>\n",
       "      <td>Copy Number</td>\n",
       "      <td>True</td>\n",
       "      <td>6442</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>69 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      variant_id gene_name                                 variant_name   \n",
       "15          3195       VHL                            A149fs (c.444dup)  \\\n",
       "15          3195       VHL                            A149fs (c.444dup)   \n",
       "15          3195       VHL                            A149fs (c.444dup)   \n",
       "16          2144       VHL                             A56fs (c.166dup)   \n",
       "17          1579      EGFR                              A767_V769dupASV   \n",
       "...          ...       ...                                          ...   \n",
       "1156         414     ERBB2                                 Y772_A775DUP   \n",
       "1156         414     ERBB2                                 Y772_A775DUP   \n",
       "1156         414     ERBB2                                 Y772_A775DUP   \n",
       "1156         414     ERBB2                                 Y772_A775DUP   \n",
       "1181        2411      SDHB  p.193_196dupSTSC (c.577_588dupAGCACCAGCTGC)   \n",
       "\n",
       "         category  variant_accepted evidence_ids evidence_status   \n",
       "15    Copy Number             False         9982     [submitted]  \\\n",
       "15    Copy Number             False         8822     [submitted]   \n",
       "15    Copy Number             False         6156     [submitted]   \n",
       "16    Copy Number              True         5742      [accepted]   \n",
       "17    Copy Number              True         4665      [accepted]   \n",
       "...           ...               ...          ...             ...   \n",
       "1156  Copy Number              True         4428     [submitted]   \n",
       "1156  Copy Number              True         1046      [rejected]   \n",
       "1156  Copy Number              True         1047      [accepted]   \n",
       "1156  Copy Number              True          960      [accepted]   \n",
       "1181  Copy Number              True         6442      [accepted]   \n",
       "\n",
       "     evidence_rating evidence_level  \n",
       "15               [3]            [C]  \n",
       "15               [2]            [C]  \n",
       "15               [1]            [C]  \n",
       "16               [3]            [C]  \n",
       "17            [None]            [C]  \n",
       "...              ...            ...  \n",
       "1156          [None]            [D]  \n",
       "1156             [4]            [D]  \n",
       "1156             [4]            [D]  \n",
       "1156             [3]            [D]  \n",
       "1181             [2]            [C]  \n",
       "\n",
       "[69 rows x 9 columns]"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "copy_number_non_supp_var_df "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"Number of unique variant ID's in Copy Number Variant category: 34\""
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_copy_number_variant_ids = len(set(copy_number_non_supp_var_df.variant_id))\n",
    "f\"Number of unique variant ID's in Copy Number Variant category: {unique_copy_number_variant_ids}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Percentage of all CIViC Variant Items in Copy Number category: 0.968937019093759%'"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "copy_number_percentage_civic_variant_items = unique_copy_number_variant_ids/total_number_variants*100\n",
    "f\"Percentage of all CIViC Variant Items in Copy Number category: {copy_number_percentage_civic_variant_items}%\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"Number of unique evidence ID's in Copy Number category: 69\""
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_copy_number_evidence_ids = len(set(copy_number_non_supp_var_df.evidence_ids))\n",
    "f\"Number of unique evidence ID's in Copy Number category: {unique_copy_number_evidence_ids}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"Number of evidence ID's in Copy Number category: 69\""
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "copy_number_evidence_ids = len(copy_number_non_supp_var_df.evidence_ids)\n",
    "f\"Number of evidence ID's in Copy Number category: {copy_number_evidence_ids}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Percentage of all CIViC Evidence Items in Copy Number category:0.6930494174367215%'"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "copy_number_percentage_civic_evidence_items = unique_copy_number_evidence_ids/total_number_evidences*100\n",
    "f\"Percentage of all CIViC Evidence Items in Copy Number category:{copy_number_percentage_civic_evidence_items}%\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Percentage of all Not Supported Evidence Items in Copy Number category:1.8459069020866776%'"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "copy_number_percentage_not_supported_evidence_items = unique_copy_number_evidence_ids/total_number_not_supported_variant_unique_evidence_items*100\n",
    "f\"Percentage of all Not Supported Evidence Items in Copy Number category:{copy_number_percentage_not_supported_evidence_items}%\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Other"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>variant_name</th>\n",
       "      <th>category</th>\n",
       "      <th>variant_accepted</th>\n",
       "      <th>evidence_ids</th>\n",
       "      <th>evidence_status</th>\n",
       "      <th>evidence_rating</th>\n",
       "      <th>evidence_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2438</td>\n",
       "      <td>VHL</td>\n",
       "      <td>235 (CAG-TAG)</td>\n",
       "      <td>Other</td>\n",
       "      <td>False</td>\n",
       "      <td>6524</td>\n",
       "      <td>[rejected]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>265</td>\n",
       "      <td>TYMS</td>\n",
       "      <td>5' TANDEM REPEAT</td>\n",
       "      <td>Other</td>\n",
       "      <td>True</td>\n",
       "      <td>678</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>2211</td>\n",
       "      <td>BAP1</td>\n",
       "      <td>ALTERNATIVE TRANSCRIPT (ATI)</td>\n",
       "      <td>Other</td>\n",
       "      <td>True</td>\n",
       "      <td>5929</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>842</td>\n",
       "      <td>BRAF</td>\n",
       "      <td>APC</td>\n",
       "      <td>Other</td>\n",
       "      <td>False</td>\n",
       "      <td>1941</td>\n",
       "      <td>[rejected]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>3460</td>\n",
       "      <td>AR</td>\n",
       "      <td>AR alternative transcripts</td>\n",
       "      <td>Other</td>\n",
       "      <td>False</td>\n",
       "      <td>9578</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1201</th>\n",
       "      <td>3478</td>\n",
       "      <td>ESR2</td>\n",
       "      <td>underexpression beta-1</td>\n",
       "      <td>Other</td>\n",
       "      <td>False</td>\n",
       "      <td>9613</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1201</th>\n",
       "      <td>3478</td>\n",
       "      <td>ESR2</td>\n",
       "      <td>underexpression beta-1</td>\n",
       "      <td>Other</td>\n",
       "      <td>False</td>\n",
       "      <td>9618</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1201</th>\n",
       "      <td>3478</td>\n",
       "      <td>ESR2</td>\n",
       "      <td>underexpression beta-1</td>\n",
       "      <td>Other</td>\n",
       "      <td>False</td>\n",
       "      <td>9619</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1203</th>\n",
       "      <td>2422</td>\n",
       "      <td>NTRK3</td>\n",
       "      <td>~DEPRECATED-ETV6-NTRK3</td>\n",
       "      <td>Other</td>\n",
       "      <td>False</td>\n",
       "      <td>10692</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1203</th>\n",
       "      <td>2422</td>\n",
       "      <td>NTRK3</td>\n",
       "      <td>~DEPRECATED-ETV6-NTRK3</td>\n",
       "      <td>Other</td>\n",
       "      <td>False</td>\n",
       "      <td>10717</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>160 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      variant_id gene_name                  variant_name category   \n",
       "3           2438       VHL                 235 (CAG-TAG)    Other  \\\n",
       "9            265      TYMS              5' TANDEM REPEAT    Other   \n",
       "24          2211      BAP1  ALTERNATIVE TRANSCRIPT (ATI)    Other   \n",
       "25           842      BRAF                           APC    Other   \n",
       "27          3460        AR    AR alternative transcripts    Other   \n",
       "...          ...       ...                           ...      ...   \n",
       "1201        3478      ESR2        underexpression beta-1    Other   \n",
       "1201        3478      ESR2        underexpression beta-1    Other   \n",
       "1201        3478      ESR2        underexpression beta-1    Other   \n",
       "1203        2422     NTRK3        ~DEPRECATED-ETV6-NTRK3    Other   \n",
       "1203        2422     NTRK3        ~DEPRECATED-ETV6-NTRK3    Other   \n",
       "\n",
       "      variant_accepted evidence_ids evidence_status evidence_rating   \n",
       "3                False         6524      [rejected]             [3]  \\\n",
       "9                 True          678      [accepted]             [3]   \n",
       "24                True         5929      [accepted]             [3]   \n",
       "25               False         1941      [rejected]             [3]   \n",
       "27               False         9578     [submitted]             [4]   \n",
       "...                ...          ...             ...             ...   \n",
       "1201             False         9613     [submitted]             [4]   \n",
       "1201             False         9618     [submitted]             [4]   \n",
       "1201             False         9619     [submitted]             [4]   \n",
       "1203             False        10692     [submitted]             [3]   \n",
       "1203             False        10717     [submitted]             [3]   \n",
       "\n",
       "     evidence_level  \n",
       "3               [C]  \n",
       "9               [B]  \n",
       "24              [D]  \n",
       "25              [B]  \n",
       "27              [B]  \n",
       "...             ...  \n",
       "1201            [B]  \n",
       "1201            [B]  \n",
       "1201            [B]  \n",
       "1203            [C]  \n",
       "1203            [C]  \n",
       "\n",
       "[160 rows x 9 columns]"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "other_non_supp_var_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"Number of unique variant ID's in Other Variant category: 92\""
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_other_variant_ids = len(set(other_non_supp_var_df.variant_id))\n",
    "f\"Number of unique variant ID's in Other Variant category: {unique_other_variant_ids}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Percentage of all CIViC Variant Items in Other category: 2.62182958107723%'"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "other_percentage_civic_variant_items = unique_other_variant_ids/total_number_variants*100\n",
    "f\"Percentage of all CIViC Variant Items in Other category: {other_percentage_civic_variant_items}%\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Number of unique evidence IDs in Other category:160'"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_other_evidence_ids = len(set(other_non_supp_var_df.evidence_ids))\n",
    "f\"Number of unique evidence IDs in Other category:{unique_other_evidence_ids}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Number of evidence IDs in Other category:160'"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "other_evidence_ids = len(other_non_supp_var_df.evidence_ids)\n",
    "f\"Number of evidence IDs in Other category:{other_evidence_ids}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Percentage of all CIViC Evidence Items in Other category:1.6070711128967456%'"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "other_percentage_civic_evidence_items = unique_other_evidence_ids/total_number_evidences*100\n",
    "f\"Percentage of all CIViC Evidence Items in Other category:{other_percentage_civic_evidence_items}%\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Percentage of all Not Supported Evidence Items in Other category:4.280363830925628%'"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "other_percentage_not_supported_evidence_items = unique_other_evidence_ids/total_number_not_supported_variant_unique_evidence_items*100\n",
    "f\"Percentage of all Not Supported Evidence Items in Other category:{other_percentage_not_supported_evidence_items}%\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Genotypes Easy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>variant_name</th>\n",
       "      <th>category</th>\n",
       "      <th>variant_accepted</th>\n",
       "      <th>evidence_ids</th>\n",
       "      <th>evidence_status</th>\n",
       "      <th>evidence_rating</th>\n",
       "      <th>evidence_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1083</th>\n",
       "      <td>729</td>\n",
       "      <td>UGT1A1</td>\n",
       "      <td>UGT1A1*28</td>\n",
       "      <td>Genotypes Easy</td>\n",
       "      <td>True</td>\n",
       "      <td>1792</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>[A]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1084</th>\n",
       "      <td>732</td>\n",
       "      <td>UGT1A1</td>\n",
       "      <td>UGT1A1*60</td>\n",
       "      <td>Genotypes Easy</td>\n",
       "      <td>True</td>\n",
       "      <td>1795</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1141</th>\n",
       "      <td>426</td>\n",
       "      <td>BRAF</td>\n",
       "      <td>WILD TYPE</td>\n",
       "      <td>Genotypes Easy</td>\n",
       "      <td>True</td>\n",
       "      <td>995</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1142</th>\n",
       "      <td>2366</td>\n",
       "      <td>PTEN</td>\n",
       "      <td>WILD TYPE</td>\n",
       "      <td>Genotypes Easy</td>\n",
       "      <td>False</td>\n",
       "      <td>6272</td>\n",
       "      <td>[rejected]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1143</th>\n",
       "      <td>2651</td>\n",
       "      <td>KIT</td>\n",
       "      <td>WILDTYPE</td>\n",
       "      <td>Genotypes Easy</td>\n",
       "      <td>True</td>\n",
       "      <td>4144</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1145</th>\n",
       "      <td>369</td>\n",
       "      <td>TP53</td>\n",
       "      <td>Wildtype</td>\n",
       "      <td>Genotypes Easy</td>\n",
       "      <td>True</td>\n",
       "      <td>875</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1145</th>\n",
       "      <td>369</td>\n",
       "      <td>TP53</td>\n",
       "      <td>Wildtype</td>\n",
       "      <td>Genotypes Easy</td>\n",
       "      <td>True</td>\n",
       "      <td>906</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1145</th>\n",
       "      <td>369</td>\n",
       "      <td>TP53</td>\n",
       "      <td>Wildtype</td>\n",
       "      <td>Genotypes Easy</td>\n",
       "      <td>True</td>\n",
       "      <td>1149</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1145</th>\n",
       "      <td>369</td>\n",
       "      <td>TP53</td>\n",
       "      <td>Wildtype</td>\n",
       "      <td>Genotypes Easy</td>\n",
       "      <td>True</td>\n",
       "      <td>2965</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1145</th>\n",
       "      <td>369</td>\n",
       "      <td>TP53</td>\n",
       "      <td>Wildtype</td>\n",
       "      <td>Genotypes Easy</td>\n",
       "      <td>True</td>\n",
       "      <td>2963</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1145</th>\n",
       "      <td>369</td>\n",
       "      <td>TP53</td>\n",
       "      <td>Wildtype</td>\n",
       "      <td>Genotypes Easy</td>\n",
       "      <td>True</td>\n",
       "      <td>2964</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1146</th>\n",
       "      <td>2174</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>Wildtype</td>\n",
       "      <td>Genotypes Easy</td>\n",
       "      <td>True</td>\n",
       "      <td>2512</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1146</th>\n",
       "      <td>2174</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>Wildtype</td>\n",
       "      <td>Genotypes Easy</td>\n",
       "      <td>True</td>\n",
       "      <td>6185</td>\n",
       "      <td>[rejected]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1147</th>\n",
       "      <td>2590</td>\n",
       "      <td>PIK3CA</td>\n",
       "      <td>Wildtype</td>\n",
       "      <td>Genotypes Easy</td>\n",
       "      <td>True</td>\n",
       "      <td>2038</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1147</th>\n",
       "      <td>2590</td>\n",
       "      <td>PIK3CA</td>\n",
       "      <td>Wildtype</td>\n",
       "      <td>Genotypes Easy</td>\n",
       "      <td>True</td>\n",
       "      <td>7047</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1148</th>\n",
       "      <td>3200</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>Wildtype</td>\n",
       "      <td>Genotypes Easy</td>\n",
       "      <td>True</td>\n",
       "      <td>7052</td>\n",
       "      <td>[rejected]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1148</th>\n",
       "      <td>3200</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>Wildtype</td>\n",
       "      <td>Genotypes Easy</td>\n",
       "      <td>True</td>\n",
       "      <td>11057</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1148</th>\n",
       "      <td>3200</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>Wildtype</td>\n",
       "      <td>Genotypes Easy</td>\n",
       "      <td>True</td>\n",
       "      <td>8840</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1184</th>\n",
       "      <td>3512</td>\n",
       "      <td>BRAF</td>\n",
       "      <td>p61BRAF(V600E)</td>\n",
       "      <td>Genotypes Easy</td>\n",
       "      <td>False</td>\n",
       "      <td>9703</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1184</th>\n",
       "      <td>3512</td>\n",
       "      <td>BRAF</td>\n",
       "      <td>p61BRAF(V600E)</td>\n",
       "      <td>Genotypes Easy</td>\n",
       "      <td>False</td>\n",
       "      <td>9702</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[D]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      variant_id gene_name    variant_name        category  variant_accepted   \n",
       "1083         729    UGT1A1       UGT1A1*28  Genotypes Easy              True  \\\n",
       "1084         732    UGT1A1       UGT1A1*60  Genotypes Easy              True   \n",
       "1141         426      BRAF       WILD TYPE  Genotypes Easy              True   \n",
       "1142        2366      PTEN       WILD TYPE  Genotypes Easy             False   \n",
       "1143        2651       KIT        WILDTYPE  Genotypes Easy              True   \n",
       "1145         369      TP53        Wildtype  Genotypes Easy              True   \n",
       "1145         369      TP53        Wildtype  Genotypes Easy              True   \n",
       "1145         369      TP53        Wildtype  Genotypes Easy              True   \n",
       "1145         369      TP53        Wildtype  Genotypes Easy              True   \n",
       "1145         369      TP53        Wildtype  Genotypes Easy              True   \n",
       "1145         369      TP53        Wildtype  Genotypes Easy              True   \n",
       "1146        2174      EGFR        Wildtype  Genotypes Easy              True   \n",
       "1146        2174      EGFR        Wildtype  Genotypes Easy              True   \n",
       "1147        2590    PIK3CA        Wildtype  Genotypes Easy              True   \n",
       "1147        2590    PIK3CA        Wildtype  Genotypes Easy              True   \n",
       "1148        3200      KRAS        Wildtype  Genotypes Easy              True   \n",
       "1148        3200      KRAS        Wildtype  Genotypes Easy              True   \n",
       "1148        3200      KRAS        Wildtype  Genotypes Easy              True   \n",
       "1184        3512      BRAF  p61BRAF(V600E)  Genotypes Easy             False   \n",
       "1184        3512      BRAF  p61BRAF(V600E)  Genotypes Easy             False   \n",
       "\n",
       "     evidence_ids evidence_status evidence_rating evidence_level  \n",
       "1083         1792      [accepted]             [5]            [A]  \n",
       "1084         1795      [accepted]             [4]            [B]  \n",
       "1141          995      [accepted]             [2]            [B]  \n",
       "1142         6272      [rejected]             [2]            [D]  \n",
       "1143         4144      [accepted]             [2]            [B]  \n",
       "1145          875      [accepted]             [3]            [B]  \n",
       "1145          906      [accepted]             [3]            [B]  \n",
       "1145         1149      [accepted]             [3]            [B]  \n",
       "1145         2965      [accepted]             [3]            [B]  \n",
       "1145         2963      [accepted]             [4]            [D]  \n",
       "1145         2964      [accepted]             [4]            [D]  \n",
       "1146         2512      [accepted]             [4]            [B]  \n",
       "1146         6185      [rejected]             [3]            [B]  \n",
       "1147         2038      [accepted]             [3]            [B]  \n",
       "1147         7047     [submitted]             [2]            [B]  \n",
       "1148         7052      [rejected]             [5]            [B]  \n",
       "1148        11057     [submitted]             [3]            [B]  \n",
       "1148         8840      [accepted]             [2]            [D]  \n",
       "1184         9703     [submitted]             [2]            [C]  \n",
       "1184         9702     [submitted]             [4]            [D]  "
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotypes_easy_non_supp_var_df "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"Number of unique variant ID's in Genotypes Easy Variant category: 10\""
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_genotypes_easy_variant_ids = len(set(genotypes_easy_non_supp_var_df.variant_id))\n",
    "f\"Number of unique variant ID's in Genotypes Easy Variant category: {unique_genotypes_easy_variant_ids}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Percentage of all CIViC Variant Items in Genotypes Easy category: 0.28498147620404674%'"
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotypes_easy_percentage_civic_variant_items = unique_genotypes_easy_variant_ids/total_number_variants*100\n",
    "f\"Percentage of all CIViC Variant Items in Genotypes Easy category: {genotypes_easy_percentage_civic_variant_items}%\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"Number of unique evidence ID's in Genotypes Easy category:20\""
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_genotypes_easy_evidence_ids = len(set(genotypes_easy_non_supp_var_df.evidence_ids))\n",
    "f\"Number of unique evidence ID's in Genotypes Easy category:{unique_genotypes_easy_evidence_ids}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"Number of evidence ID's in Genotypes Easy category:20\""
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotypes_easy_evidence_ids = len(genotypes_easy_non_supp_var_df.evidence_ids)\n",
    "f\"Number of evidence ID's in Genotypes Easy category:{genotypes_easy_evidence_ids}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Percentage of all CIViC Evidence Items in Genotypes Easy category:0.2008838891120932%'"
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotypes_easy_percentage_civic_evidence_items = unique_genotypes_easy_evidence_ids/total_number_evidences*100\n",
    "f\"Percentage of all CIViC Evidence Items in Genotypes Easy category:{genotypes_easy_percentage_civic_evidence_items}%\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Percentage of all Not Supported Evidence Items in Genotypes Easy category:0.5350454788657035%'"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotypes_easy_percentage_not_supported_evidence_items = unique_genotypes_easy_evidence_ids/total_number_not_supported_variant_unique_evidence_items*100\n",
    "f\"Percentage of all Not Supported Evidence Items in Genotypes Easy category:{genotypes_easy_percentage_not_supported_evidence_items}%\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Genotypes Compound"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>variant_name</th>\n",
       "      <th>category</th>\n",
       "      <th>variant_accepted</th>\n",
       "      <th>evidence_ids</th>\n",
       "      <th>evidence_status</th>\n",
       "      <th>evidence_rating</th>\n",
       "      <th>evidence_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>821</td>\n",
       "      <td>HLA-C</td>\n",
       "      <td>COPY-NEUTRAL LOSS OF HETEROZYGOSITY</td>\n",
       "      <td>Genotypes Compound</td>\n",
       "      <td>True</td>\n",
       "      <td>1899</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>738</td>\n",
       "      <td>DPYD</td>\n",
       "      <td>DPYD*13 HOMOZYGOSITY</td>\n",
       "      <td>Genotypes Compound</td>\n",
       "      <td>True</td>\n",
       "      <td>1801</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>[A]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>176</th>\n",
       "      <td>737</td>\n",
       "      <td>DPYD</td>\n",
       "      <td>DPYD*2A HOMOZYGOSITY</td>\n",
       "      <td>Genotypes Compound</td>\n",
       "      <td>True</td>\n",
       "      <td>1800</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[5]</td>\n",
       "      <td>[A]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>548</th>\n",
       "      <td>302</td>\n",
       "      <td>PTCH1</td>\n",
       "      <td>LOH</td>\n",
       "      <td>Genotypes Compound</td>\n",
       "      <td>True</td>\n",
       "      <td>749</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>918</th>\n",
       "      <td>3235</td>\n",
       "      <td>PRKAR1A</td>\n",
       "      <td>PRKAR1A LOH, allelic imbalance 17q</td>\n",
       "      <td>Genotypes Compound</td>\n",
       "      <td>False</td>\n",
       "      <td>8936</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1044</th>\n",
       "      <td>3108</td>\n",
       "      <td>VHL</td>\n",
       "      <td>Single Allele Deletion</td>\n",
       "      <td>Genotypes Compound</td>\n",
       "      <td>False</td>\n",
       "      <td>8499</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1044</th>\n",
       "      <td>3108</td>\n",
       "      <td>VHL</td>\n",
       "      <td>Single Allele Deletion</td>\n",
       "      <td>Genotypes Compound</td>\n",
       "      <td>False</td>\n",
       "      <td>8580</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      variant_id gene_name                         variant_name   \n",
       "149          821     HLA-C  COPY-NEUTRAL LOSS OF HETEROZYGOSITY  \\\n",
       "175          738      DPYD                 DPYD*13 HOMOZYGOSITY   \n",
       "176          737      DPYD                 DPYD*2A HOMOZYGOSITY   \n",
       "548          302     PTCH1                                  LOH   \n",
       "918         3235   PRKAR1A   PRKAR1A LOH, allelic imbalance 17q   \n",
       "1044        3108       VHL               Single Allele Deletion   \n",
       "1044        3108       VHL               Single Allele Deletion   \n",
       "\n",
       "                category  variant_accepted evidence_ids evidence_status   \n",
       "149   Genotypes Compound              True         1899      [accepted]  \\\n",
       "175   Genotypes Compound              True         1801      [accepted]   \n",
       "176   Genotypes Compound              True         1800      [accepted]   \n",
       "548   Genotypes Compound              True          749      [accepted]   \n",
       "918   Genotypes Compound             False         8936     [submitted]   \n",
       "1044  Genotypes Compound             False         8499     [submitted]   \n",
       "1044  Genotypes Compound             False         8580     [submitted]   \n",
       "\n",
       "     evidence_rating evidence_level  \n",
       "149              [3]            [C]  \n",
       "175              [5]            [A]  \n",
       "176              [5]            [A]  \n",
       "548              [2]            [B]  \n",
       "918              [3]            [B]  \n",
       "1044             [3]            [C]  \n",
       "1044             [3]            [C]  "
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotypes_compound_non_supp_var_df "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"Number of unique variant ID's in Genotypes Compound Variant category: 6\""
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_genotypes_compound_variant_ids = len(set(genotypes_compound_non_supp_var_df.variant_id))\n",
    "f\"Number of unique variant ID's in Genotypes Compound Variant category: {unique_genotypes_compound_variant_ids}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Percentage of all CIViC Variant Items in Genotypes Compound category: 0.17098888572242804%'"
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotypes_compound_percentage_civic_variant_items = unique_genotypes_compound_variant_ids/total_number_variants*100\n",
    "f\"Percentage of all CIViC Variant Items in Genotypes Compound category: {genotypes_compound_percentage_civic_variant_items}%\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"Number of unique evidence ID's in Genptypes Compound category: 7\""
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_genotypes_compound_evidence_ids = len(set(genotypes_compound_non_supp_var_df.evidence_ids))\n",
    "f\"Number of unique evidence ID's in Genptypes Compound category: {unique_genotypes_compound_evidence_ids}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"Number of evidence ID's in Genptypes Compound category: 7\""
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotypes_compound_evidence_ids = len(genotypes_compound_non_supp_var_df.evidence_ids)\n",
    "f\"Number of evidence ID's in Genptypes Compound category: {genotypes_compound_evidence_ids}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Percentage of all CIViC Evidence Items in Genotypes Compound category:0.07030936118923262%'"
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotypes_compound_percentage_civic_evidence_items = unique_genotypes_compound_evidence_ids/total_number_evidences*100\n",
    "f\"Percentage of all CIViC Evidence Items in Genotypes Compound category:{genotypes_compound_percentage_civic_evidence_items}%\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Percentage of all Not Supported Evidence Items in Genotypes Compound category:0.18726591760299627%'"
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotypes_compound_percentage_not_supported_evidence_items = unique_genotypes_compound_evidence_ids/total_number_not_supported_variant_unique_evidence_items*100\n",
    "f\"Percentage of all Not Supported Evidence Items in Genotypes Compound category:{genotypes_compound_percentage_not_supported_evidence_items}%\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Region Defined Variant"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>variant_name</th>\n",
       "      <th>category</th>\n",
       "      <th>variant_accepted</th>\n",
       "      <th>evidence_ids</th>\n",
       "      <th>evidence_status</th>\n",
       "      <th>evidence_rating</th>\n",
       "      <th>evidence_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4188</td>\n",
       "      <td>VHL</td>\n",
       "      <td>10kb Deletion</td>\n",
       "      <td>Region Defined Variant</td>\n",
       "      <td>False</td>\n",
       "      <td>10678</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4186</td>\n",
       "      <td>VHL</td>\n",
       "      <td>2kb Deletion</td>\n",
       "      <td>Region Defined Variant</td>\n",
       "      <td>False</td>\n",
       "      <td>10791</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4186</td>\n",
       "      <td>VHL</td>\n",
       "      <td>2kb Deletion</td>\n",
       "      <td>Region Defined Variant</td>\n",
       "      <td>False</td>\n",
       "      <td>10676</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>823</td>\n",
       "      <td>EPCAM</td>\n",
       "      <td>3' Exon Deletion</td>\n",
       "      <td>Region Defined Variant</td>\n",
       "      <td>True</td>\n",
       "      <td>1901</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>253</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>3' UTR MUTATION</td>\n",
       "      <td>Region Defined Variant</td>\n",
       "      <td>False</td>\n",
       "      <td>663</td>\n",
       "      <td>[rejected]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1030</th>\n",
       "      <td>536</td>\n",
       "      <td>SULT1E1</td>\n",
       "      <td>SNP</td>\n",
       "      <td>Region Defined Variant</td>\n",
       "      <td>False</td>\n",
       "      <td>1312</td>\n",
       "      <td>[rejected]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1048</th>\n",
       "      <td>491</td>\n",
       "      <td>HSPH1</td>\n",
       "      <td>T17 DELETION</td>\n",
       "      <td>Region Defined Variant</td>\n",
       "      <td>True</td>\n",
       "      <td>1164</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[B]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1169</th>\n",
       "      <td>2349</td>\n",
       "      <td>IL7R</td>\n",
       "      <td>exon 6 mutations</td>\n",
       "      <td>Region Defined Variant</td>\n",
       "      <td>False</td>\n",
       "      <td>6228</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1178</th>\n",
       "      <td>2845</td>\n",
       "      <td>HOXD8</td>\n",
       "      <td>mutation</td>\n",
       "      <td>Region Defined Variant</td>\n",
       "      <td>False</td>\n",
       "      <td>7670</td>\n",
       "      <td>[submitted]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1179</th>\n",
       "      <td>4436</td>\n",
       "      <td>CHEK2</td>\n",
       "      <td>mutation</td>\n",
       "      <td>Region Defined Variant</td>\n",
       "      <td>True</td>\n",
       "      <td>11205</td>\n",
       "      <td>[accepted]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[A]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>896 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      variant_id gene_name      variant_name                category   \n",
       "1           4188       VHL     10kb Deletion  Region Defined Variant  \\\n",
       "4           4186       VHL      2kb Deletion  Region Defined Variant   \n",
       "4           4186       VHL      2kb Deletion  Region Defined Variant   \n",
       "5            823     EPCAM  3' Exon Deletion  Region Defined Variant   \n",
       "6            253      EGFR   3' UTR MUTATION  Region Defined Variant   \n",
       "...          ...       ...               ...                     ...   \n",
       "1030         536   SULT1E1               SNP  Region Defined Variant   \n",
       "1048         491     HSPH1      T17 DELETION  Region Defined Variant   \n",
       "1169        2349      IL7R  exon 6 mutations  Region Defined Variant   \n",
       "1178        2845     HOXD8          mutation  Region Defined Variant   \n",
       "1179        4436     CHEK2          mutation  Region Defined Variant   \n",
       "\n",
       "      variant_accepted evidence_ids evidence_status evidence_rating   \n",
       "1                False        10678     [submitted]             [3]  \\\n",
       "4                False        10791     [submitted]             [3]   \n",
       "4                False        10676     [submitted]             [2]   \n",
       "5                 True         1901      [accepted]             [4]   \n",
       "6                False          663      [rejected]             [4]   \n",
       "...                ...          ...             ...             ...   \n",
       "1030             False         1312      [rejected]             [2]   \n",
       "1048              True         1164      [accepted]             [2]   \n",
       "1169             False         6228     [submitted]             [4]   \n",
       "1178             False         7670     [submitted]             [2]   \n",
       "1179              True        11205      [accepted]             [3]   \n",
       "\n",
       "     evidence_level  \n",
       "1               [C]  \n",
       "4               [C]  \n",
       "4               [C]  \n",
       "5               [B]  \n",
       "6               [B]  \n",
       "...             ...  \n",
       "1030            [B]  \n",
       "1048            [B]  \n",
       "1169            [C]  \n",
       "1178            [C]  \n",
       "1179            [A]  \n",
       "\n",
       "[896 rows x 9 columns]"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "region_defined_variant_non_supp_var_df "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"Number of unique variant ID's in Region Defined Variant category: 190\""
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_region_defined_variant_ids = len(set(region_defined_variant_non_supp_var_df.variant_id))\n",
    "f\"Number of unique variant ID's in Region Defined Variant category: {unique_region_defined_variant_ids}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Percentage of all CIViC Variant Items in Region Defined category: 5.414648047876888%'"
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "region_defined_percentage_civic_variant_items = unique_region_defined_variant_ids/total_number_variants*100\n",
    "f\"Percentage of all CIViC Variant Items in Region Defined category: {region_defined_percentage_civic_variant_items}%\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"Number of unique evidence ID's in Region Defined Variant category: 894\""
      ]
     },
     "execution_count": 119,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_region_defined_evidence_ids = len(set(region_defined_variant_non_supp_var_df.evidence_ids))\n",
    "f\"Number of unique evidence ID's in Region Defined Variant category: {unique_region_defined_evidence_ids}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"Number of evidence ID's in Region Defined Variant category: 896\""
      ]
     },
     "execution_count": 120,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "region_defined_evidence_ids = len(region_defined_variant_non_supp_var_df.evidence_ids)\n",
    "f\"Number of evidence ID's in Region Defined Variant category: {region_defined_evidence_ids}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Percentage of all CIViC Evidence Items in Region Defined category:8.979509843310566%'"
      ]
     },
     "execution_count": 121,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "region_defined_percentage_civic_evidence_items = unique_region_defined_evidence_ids/total_number_evidences*100\n",
    "f\"Percentage of all CIViC Evidence Items in Region Defined category:{region_defined_percentage_civic_evidence_items}%\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Percentage of all Not Supported Evidence Items in Region Defined category:23.91653290529695%'"
      ]
     },
     "execution_count": 122,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "region_defined_percentage_not_supported_evidence_items = unique_region_defined_evidence_ids/total_number_not_supported_variant_unique_evidence_items*100\n",
    "f\"Percentage of all Not Supported Evidence Items in Region Defined category:{region_defined_percentage_not_supported_evidence_items}%\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Statistical Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_variant_dict = {'Variant Category':['Normalized', 'Unable to Normalize', 'Not Supported'],\n",
    "        # 'Total Number of Evidence Items': [total_number_not_normalized_variant_unique_evidence_items, total_number_not_normalized_variant_unique_evidence_items, total_number_not_supported_variant_unique_evidence_items],\n",
    "        'Percentage of all CIViC Evidence Items': [normalized_percentage_evidence_items, not_normalized_percentage_evidence_items, not_supported_variant_percentage_evidence_items],\n",
    "        'Percentage of all CIViC Variant Items': [percentage_normalized_variants, percentage_not_normalized_variants, percentage_not_supported_variants]}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Variant Category</th>\n",
       "      <th>Percentage of all CIViC Evidence Items</th>\n",
       "      <th>Percentage of all CIViC Variant Items</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Normalized</td>\n",
       "      <td>58.708317</td>\n",
       "      <td>52.949558</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Unable to Normalize</td>\n",
       "      <td>0.803536</td>\n",
       "      <td>1.795383</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Not Supported</td>\n",
       "      <td>37.545199</td>\n",
       "      <td>34.311770</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Variant Category  Percentage of all CIViC Evidence Items   \n",
       "0           Normalized                               58.708317  \\\n",
       "1  Unable to Normalize                                0.803536   \n",
       "2        Not Supported                               37.545199   \n",
       "\n",
       "   Percentage of all CIViC Variant Items  \n",
       "0                              52.949558  \n",
       "1                               1.795383  \n",
       "2                              34.311770  "
      ]
     },
     "execution_count": 124,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variant_statistics_df = pd.DataFrame(all_variant_dict)\n",
    "variant_statistics_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [],
   "source": [
    "not_supported_variant_dict = {'Category': ['Expression','Epigenetic Modification', 'Fusion', 'Protein Consequence', 'Gene Function', 'Rearrangements', 'Copy Number', 'Other', 'Genotypes Easy', 'Genotypes Compound', 'Region Defined Variant' ],\n",
    "                                # 'Total Number of Evidence Items': [expression_evidence_ids, epigenetic_evidence_ids, fusion_evidence_ids, protein_evidece_ids, gene_function_evidence_ids, rearrangements_evidence_ids, copy_number_evidence_ids, other_evidence_ids, genotypes_easy_evidence_ids, genotypes_compound_evidence_ids, region_defined_evidence_ids],\n",
    "                                'Percentage of all Not Supported Evidence Items':[expression_percentage_not_supported_evidence_items, epigenetic_percentage_not_supported_evidence_items, fusion_percentage_not_supported_evidence_items, protein_evidece_percentage_not_supported_evidence_items, gene_function_percentage_not_supported_evidence_items, rearrangements_percentage_not_supported_evidence_items, copy_number_percentage_not_supported_evidence_items, other_percentage_not_supported_evidence_items, genotypes_easy_percentage_not_supported_evidence_items, genotypes_compound_percentage_not_supported_evidence_items, region_defined_percentage_not_supported_evidence_items],\n",
    "                                'Percentage of all CIViC Evidence Items':[expression_percentage_civic_evidence_items, epigenetic_percentage_civic_evidence_items, fusion_percentage_civic_evidence_items, protein_evidece_percentage_civic_evidence_items, gene_function_percentage_civic_evidence_items, rearrangements_percentage_civic_evidence_items, copy_number_percentage_civic_evidence_items, other_percentage_civic_evidence_items, genotypes_easy_percentage_civic_evidence_items, genotypes_compound_percentage_civic_evidence_items, region_defined_percentage_civic_evidence_items],\n",
    "                                'Percentage of all CIViC Varinat Items': [expression_percentage_civic_variant_items, epigenetic_modification_percentage_civic_variant_items, fusion_percentage_civic_variant_items, protein_consequence_percentage_civic_variant_items, gene_function_percentage_civic_variant_items, rearrangements_percentage_civic_variant_items, copy_number_percentage_civic_variant_items, other_percentage_civic_variant_items, genotypes_easy_percentage_civic_variant_items, genotypes_compound_percentage_civic_variant_items, region_defined_percentage_civic_variant_items]}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Category</th>\n",
       "      <th>Percentage of all Not Supported Evidence Items</th>\n",
       "      <th>Percentage of all CIViC Evidence Items</th>\n",
       "      <th>Percentage of all CIViC Varinat Items</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Expression</td>\n",
       "      <td>16.506153</td>\n",
       "      <td>6.197268</td>\n",
       "      <td>8.292961</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Epigenetic Modification</td>\n",
       "      <td>0.642055</td>\n",
       "      <td>0.241061</td>\n",
       "      <td>0.427472</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Fusion</td>\n",
       "      <td>31.728197</td>\n",
       "      <td>11.912415</td>\n",
       "      <td>8.577942</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Protein Consequence</td>\n",
       "      <td>7.891921</td>\n",
       "      <td>2.963037</td>\n",
       "      <td>3.647763</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Gene Function</td>\n",
       "      <td>9.176030</td>\n",
       "      <td>3.445159</td>\n",
       "      <td>2.479339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Rearrangements</td>\n",
       "      <td>3.638309</td>\n",
       "      <td>1.366010</td>\n",
       "      <td>1.424907</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Copy Number</td>\n",
       "      <td>1.845907</td>\n",
       "      <td>0.693049</td>\n",
       "      <td>0.968937</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Other</td>\n",
       "      <td>4.280364</td>\n",
       "      <td>1.607071</td>\n",
       "      <td>2.621830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Genotypes Easy</td>\n",
       "      <td>0.535045</td>\n",
       "      <td>0.200884</td>\n",
       "      <td>0.284981</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Genotypes Compound</td>\n",
       "      <td>0.187266</td>\n",
       "      <td>0.070309</td>\n",
       "      <td>0.170989</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Region Defined Variant</td>\n",
       "      <td>23.916533</td>\n",
       "      <td>8.979510</td>\n",
       "      <td>5.414648</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   Category  Percentage of all Not Supported Evidence Items   \n",
       "0                Expression                                       16.506153  \\\n",
       "1   Epigenetic Modification                                        0.642055   \n",
       "2                    Fusion                                       31.728197   \n",
       "3       Protein Consequence                                        7.891921   \n",
       "4             Gene Function                                        9.176030   \n",
       "5            Rearrangements                                        3.638309   \n",
       "6               Copy Number                                        1.845907   \n",
       "7                     Other                                        4.280364   \n",
       "8            Genotypes Easy                                        0.535045   \n",
       "9        Genotypes Compound                                        0.187266   \n",
       "10   Region Defined Variant                                       23.916533   \n",
       "\n",
       "    Percentage of all CIViC Evidence Items   \n",
       "0                                 6.197268  \\\n",
       "1                                 0.241061   \n",
       "2                                11.912415   \n",
       "3                                 2.963037   \n",
       "4                                 3.445159   \n",
       "5                                 1.366010   \n",
       "6                                 0.693049   \n",
       "7                                 1.607071   \n",
       "8                                 0.200884   \n",
       "9                                 0.070309   \n",
       "10                                8.979510   \n",
       "\n",
       "    Percentage of all CIViC Varinat Items  \n",
       "0                                8.292961  \n",
       "1                                0.427472  \n",
       "2                                8.577942  \n",
       "3                                3.647763  \n",
       "4                                2.479339  \n",
       "5                                1.424907  \n",
       "6                                0.968937  \n",
       "7                                2.621830  \n",
       "8                                0.284981  \n",
       "9                                0.170989  \n",
       "10                               5.414648  "
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "not_supported_variant_statistics_df = pd.DataFrame(not_supported_variant_dict)\n",
    "not_supported_variant_statistics_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Notes: \n",
    "    Kori's analysis has variants under the transcript ariant category, while I do not\n",
    "    Kori's analysis yields 45.26% of not supported variants, mine yeilds 34.4%\n",
    "    other categories in not supported variants match in variant percentages b/w Kori's and mine"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.10.9 64-bit ('3.10.9')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "1bd4a0c8bc911393f4a19827cd0d555013b2fcb9ea63c8ddc9b9ae4ac59a8883"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
